--- 
slug: CreatePolicy
eventname: CreatePolicy
title: CreatePolicy
sidebar_label: CreatePolicy
---
                       
### Event Information

None


### Examples

1. Overly permissive policies: One example of how security can be impacted with CreatePolicy in AWS IAM is by creating overly permissive policies. This means granting excessive permissions to a user or group, allowing them to perform actions that they should not have access to. This can lead to unauthorized access, data breaches, or accidental deletion of resources.

2. Misconfigured policies: Another example is when policies are misconfigured during the creation process. This can result in unintended access to sensitive resources or services. For instance, if a policy is mistakenly configured to allow public access to an S3 bucket, it can expose sensitive data to the public, leading to a potential data breach.

3. Lack of least privilege principle: The third example is when the principle of least privilege is not followed while creating policies. This means granting more permissions than necessary to perform a specific task. By not adhering to the principle of least privilege, the risk of unauthorized access or misuse of resources increases, potentially compromising the security of the system.

### Remediation

#### Using Console

1. Overly permissive policies:
- Identify the user or group with the overly permissive policy.
- Access the AWS IAM console.
- Navigate to the "Policies" section.
- Search for the policy associated with the user or group.
- Review the policy and identify the excessive permissions.
- Modify the policy to remove the unnecessary permissions.
- Save the changes to the policy.

2. Misconfigured policies:
- Identify the policy that is misconfigured.
- Access the AWS IAM console.
- Navigate to the "Policies" section.
- Search for the misconfigured policy.
- Review the policy and identify the incorrect configuration.
- Modify the policy to correct the configuration.
- Save the changes to the policy.

3. Lack of least privilege:
- Identify the user or group with the policy lacking least privilege.
- Access the AWS IAM console.
- Navigate to the "Policies" section.
- Search for the policy associated with the user or group.
- Review the policy and identify the permissions that exceed the principle of least privilege.
- Modify the policy to restrict the permissions to the minimum required for the intended tasks.
- Save the changes to the policy.

#### Using CLI

1. Overly permissive policies: To remediate overly permissive policies in AWS IAM using AWS CLI, you can follow these steps:

- Identify the policy that is granting excessive permissions to a user or group.
- Use the `aws iam list-attached-user-policies` or `aws iam list-attached-group-policies` command to list the policies attached to the user or group.
- Use the `aws iam detach-user-policy` or `aws iam detach-group-policy` command to detach the policy from the user or group.
- Review the policy and modify it to grant only the necessary permissions.
- Use the `aws iam create-policy` command to create a new policy with the updated permissions.
- Use the `aws iam attach-user-policy` or `aws iam attach-group-policy` command to attach the new policy to the user or group.

2. Misconfigured policies: To remediate misconfigured policies in AWS IAM using AWS CLI, you can follow these steps:

- Identify the misconfigured policy that is allowing unintended access to sensitive resources or data.
- Use the `aws iam list-attached-user-policies` or `aws iam list-attached-group-policies` command to list the policies attached to the user or group.
- Use the `aws iam detach-user-policy` or `aws iam detach-group-policy` command to detach the misconfigured policy from the user or group.
- Review the policy and modify it to restrict access to sensitive resources or data.
- Use the `aws iam create-policy` command to create a new policy with the updated configuration.
- Use the `aws iam attach-user-policy` or `aws iam attach-group-policy` command to attach the new policy to the user or group.

3. Lack of least privilege: To remediate the lack of least privilege in AWS IAM using AWS CLI, you can follow these steps:

- Identify the policy that is granting more permissions than necessary to a user or group.
- Use the `aws iam list-attached-user-policies` or `aws iam list-attached-group-policies` command to list the policies attached to the user or group.
- Use the `aws iam detach-user-policy` or `aws iam detach-group-policy` command to detach the policy from the user or group.
- Review the policy and modify it to grant only the minimum required permissions.
- Use the `aws iam create-policy` command to create a new policy with the updated permissions.
- Use the `aws iam attach-user-policy` or `aws iam attach-group-policy` command to attach the new policy to the user or group.

#### Using Python

1. Overly permissive policies: To remediate overly permissive policies in AWS IAM using Python, you can follow these steps:

- Identify the policies that are overly permissive and need to be remediated.
- Use the AWS SDK for Python (Boto3) to retrieve the policy document for each policy.
- Parse the policy document to identify the specific permissions that are overly permissive.
- Modify the policy document to remove the excessive permissions, ensuring that only the necessary permissions are granted.
- Use the `update_policy` method in Boto3 to update the policy with the modified document.

Here's an example Python script that demonstrates how to remediate an overly permissive policy by removing a specific permission:

```python
import boto3
import json

# Specify the IAM policy ARN that needs to be remediated
policy_arn = 'arn:aws:iam::123456789012:policy/overly-permissive-policy'

# Create an IAM client
iam_client = boto3.client('iam')

# Retrieve the policy document
response = iam_client.get_policy_version(
    PolicyArn=policy_arn,
    VersionId='v1'
)

# Parse the policy document
policy_document = json.loads(response['PolicyVersion']['Document'])

# Remove the specific permission that is overly permissive
policy_document['Statement'][0]['Action'].remove('s3:*')

# Update the policy with the modified document
response = iam_client.create_policy_version(
    PolicyArn=policy_arn,
    PolicyDocument=json.dumps(policy_document),
    SetAsDefault=True
)

# Delete the old policy version
iam_client.delete_policy_version(
    PolicyArn=policy_arn,
    VersionId=response['PolicyVersion']['VersionId']
)
```

2. Misconfigured policies: To remediate misconfigured policies in AWS IAM using Python, you can follow these steps:

- Identify the misconfigured policies that need to be remediated.
- Use the AWS SDK for Python (Boto3) to retrieve the policy document for each policy.
- Modify the policy document to correct the misconfiguration, ensuring that the intended access is granted and any unintended access is removed.
- Use the `update_policy` method in Boto3 to update the policy with the modified document.

Here's an example Python script that demonstrates how to remediate a misconfigured policy by removing public access to an S3 bucket:

```python
import boto3
import json

# Specify the IAM policy ARN that needs to be remediated
policy_arn = 'arn:aws:iam::123456789012:policy/misconfigured-policy'

# Create an IAM client
iam_client = boto3.client('iam')

# Retrieve the policy document
response = iam_client.get_policy_version(
    PolicyArn=policy_arn,
    VersionId='v1'
)

# Parse the policy document
policy_document = json.loads(response['PolicyVersion']['Document'])

# Remove public access to S3 bucket
policy_document['Statement'][0]['Resource'][0] = 'arn:aws:s3:::my-bucket/*'

# Update the policy with the modified document
response = iam_client.create_policy_version(
    PolicyArn=policy_arn,
    PolicyDocument=json.dumps(policy_document),
    SetAsDefault=True
)

# Delete the old policy version
iam_client.delete_policy_version(
    PolicyArn=policy_arn,
    VersionId=response['PolicyVersion']['VersionId']
)
```

3. Lack of least privilege: To remediate the lack of least privilege in AWS IAM using Python, you can follow these steps:

- Identify the policies that grant more permissions than necessary and need to be remediated.
- Use the AWS SDK for Python (Boto3) to retrieve the policy document for each policy.
- Modify the policy document to remove the excessive permissions, ensuring that only the minimum required permissions are granted.
- Use the `update_policy` method in Boto3 to update the policy with the modified document.

Here's an example Python script that demonstrates how to remediate a policy by removing excessive permissions:

```python
import boto3
import json

# Specify the IAM policy ARN that needs to be remediated
policy_arn = 'arn:aws:iam::123456789012:policy/lack-of-least-privilege-policy'

# Create an IAM client
iam_client = boto3.client('iam')

# Retrieve the policy document
response = iam_client.get_policy_version(
    PolicyArn=policy_arn,
    VersionId='v1'
)

# Parse the policy document
policy_document = json.loads(response['PolicyVersion']['Document'])

# Remove excessive permissions
policy_document['Statement'][0]['Action'] = ['s3:GetObject']

# Update the policy with the modified document
response = iam_client.create_policy_version(
    PolicyArn=policy_arn,
    PolicyDocument=json.dumps(policy_document),
    SetAsDefault=True
)

# Delete the old policy version
iam_client.delete_policy_version(
    PolicyArn=policy_arn,
    VersionId=response['PolicyVersion']['VersionId']
)
```

Note: These scripts are just examples and may need to be modified based on your specific requirements and policies.


 