---
slug: principals_with_inline_policies
title: Principals with inline policies
sidebar_label: Principals with inline policies
---

### More Info:

Using inline policies is a bad practice.

### Risk Level

High

### Address

Security

### Compliance Standards

CBP,AWSWAF,AWSSSB

### Triage and Remediation
<Tabs>


<Tab title='Prevention'>
### How to Prevent
<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
To prevent principals with inline policies in AWS IAM using the AWS Management Console, follow these steps:

1. **Review IAM Users and Roles:**
   - Navigate to the IAM Dashboard in the AWS Management Console.
   - Click on "Users" to review all IAM users.
   - Click on each user to check for any inline policies attached.
   - Similarly, click on "Roles" to review all IAM roles and check for any inline policies attached.

2. **Use IAM Policy Simulator:**
   - Use the IAM Policy Simulator to test the effects of policies attached to users, groups, and roles.
   - Navigate to the IAM Policy Simulator in the AWS Management Console.
   - Select the user, group, or role you want to test.
   - Simulate the policies to ensure they grant only the necessary permissions.

3. **Implement Managed Policies:**
   - Replace inline policies with AWS Managed Policies or create custom managed policies.
   - Navigate to the "Policies" section in the IAM Dashboard.
   - Create a new policy or use an existing AWS Managed Policy.
   - Attach the managed policy to the users, groups, or roles instead of using inline policies.

4. **Enable IAM Access Analyzer:**
   - Enable IAM Access Analyzer to continuously monitor and analyze permissions granted using policies.
   - Navigate to the IAM Dashboard and select "Access Analyzer."
   - Create an analyzer to start monitoring your account for any resource policies that grant access to external entities.
   - Review findings and adjust policies as necessary to ensure they follow the principle of least privilege.

By following these steps, you can prevent the use of inline policies and ensure that your IAM policies are managed and monitored effectively.
</Accordion>

<Accordion title='Using CLI'>
To prevent principals with inline policies in AWS IAM using the AWS CLI, you can follow these steps:

1. **List All Users and Roles with Inline Policies:**
   Use the following commands to list all users and roles that have inline policies attached. This will help you identify any existing inline policies that need to be reviewed or removed.

   ```sh
   aws iam list-users
   aws iam list-roles
   ```

2. **List Inline Policies for Each User and Role:**
   For each user and role identified in the previous step, list the inline policies attached to them. This will give you a detailed view of the inline policies in use.

   ```sh
   aws iam list-user-policies --user-name <user-name>
   aws iam list-role-policies --role-name <role-name>
   ```

3. **Remove Inline Policies:**
   To prevent the use of inline policies, you can remove any existing inline policies from users and roles. This step ensures that no inline policies are currently in use.

   ```sh
   aws iam delete-user-policy --user-name <user-name> --policy-name <policy-name>
   aws iam delete-role-policy --role-name <role-name> --policy-name <policy-name>
   ```

4. **Enforce the Use of Managed Policies:**
   Encourage the use of AWS managed policies or create and attach customer managed policies to users and roles instead of inline policies. This can be done by attaching managed policies to users and roles.

   ```sh
   aws iam attach-user-policy --user-name <user-name> --policy-arn <policy-arn>
   aws iam attach-role-policy --role-name <role-name> --policy-arn <policy-arn>
   ```

By following these steps, you can prevent the use of inline policies in AWS IAM and ensure that all policies are managed centrally, making it easier to maintain and audit permissions.
</Accordion>

<Accordion title='Using Python'>
To prevent principals with inline policies in AWS IAM using Python scripts, you can follow these steps:

### 1. **Set Up AWS SDK for Python (Boto3)**
First, ensure you have Boto3 installed and configured with the necessary permissions to manage IAM policies.

```bash
pip install boto3
```

### 2. **List All IAM Users and Roles**
You need to list all IAM users and roles to check for inline policies.

```python
import boto3

# Initialize a session using Amazon IAM
iam_client = boto3.client('iam')

# List all IAM users
users = iam_client.list_users()

# List all IAM roles
roles = iam_client.list_roles()
```

### 3. **Check for Inline Policies**
For each user and role, check if there are any inline policies attached.

```python
def check_inline_policies(entity_name, entity_type):
    if entity_type == 'user':
        policies = iam_client.list_user_policies(UserName=entity_name)
    elif entity_type == 'role':
        policies = iam_client.list_role_policies(RoleName=entity_name)
    
    return policies['PolicyNames']

# Check inline policies for users
for user in users['Users']:
    user_name = user['UserName']
    user_policies = check_inline_policies(user_name, 'user')
    if user_policies:
        print(f"User {user_name} has inline policies: {user_policies}")

# Check inline policies for roles
for role in roles['Roles']:
    role_name = role['RoleName']
    role_policies = check_inline_policies(role_name, 'role')
    if role_policies:
        print(f"Role {role_name} has inline policies: {role_policies}")
```

### 4. **Prevent Creation of Inline Policies**
To prevent the creation of inline policies, you can create an IAM policy that denies the `PutUserPolicy` and `PutRolePolicy` actions and attach it to the IAM users or roles that should not be able to create inline policies.

```python
# Define the policy document
policy_document = {
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Deny",
            "Action": [
                "iam:PutUserPolicy",
                "iam:PutRolePolicy"
            ],
            "Resource": "*"
        }
    ]
}

# Create the policy
response = iam_client.create_policy(
    PolicyName='DenyInlinePolicyCreation',
    PolicyDocument=json.dumps(policy_document)
)

# Attach the policy to a user or role
policy_arn = response['Policy']['Arn']
iam_client.attach_user_policy(
    UserName='YourUserName',
    PolicyArn=policy_arn
)

# Or attach the policy to a role
iam_client.attach_role_policy(
    RoleName='YourRoleName',
    PolicyArn=policy_arn
)
```

### Summary
1. **Set Up AWS SDK for Python (Boto3)**: Install and configure Boto3.
2. **List All IAM Users and Roles**: Retrieve all IAM users and roles.
3. **Check for Inline Policies**: Identify users and roles with inline policies.
4. **Prevent Creation of Inline Policies**: Create and attach a policy that denies the creation of inline policies.

By following these steps, you can effectively prevent principals from having inline policies in AWS IAM using Python scripts.
</Accordion>

</AccordionGroup>
</Tab>

<Tab title='Cause'>
### Check Cause
<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
1. Log in to the AWS Management Console and open the IAM console at https://console.aws.amazon.com/iam/.

2. In the navigation pane, choose "Users". This will display a list of all the IAM users in your AWS account.

3. Click on the name of the user you want to check. This will open the user's summary page.

4. In the user's summary page, look for the "Permissions" tab. Under this tab, you will see a section for "Inline Policies" if any are attached to the user. If there are inline policies attached, they will be listed here. If there are no inline policies, the section will say "There are no inline policies to show."
</Accordion>

<Accordion title='Using CLI'>
1. **Install and Configure AWS CLI**: Before you can start, make sure you have AWS CLI installed and configured on your system. You can install it using pip (Python package installer). The command to install AWS CLI is `pip install awscli`. After installing, you need to configure it using `aws configure` command. You will be asked to provide your AWS Access Key ID, Secret Access Key, Default region name, and Default output format.

2. **List IAM Users**: The first step to detect principals with inline policies is to list all the IAM users. You can do this using the following AWS CLI command: `aws iam list-users`. This command will return a list of all IAM users in your AWS account.

3. **List Inline Policies for Each User**: The next step is to list all the inline policies attached to each IAM user. You can do this using the following AWS CLI command: `aws iam list-user-policies --user-name <username>`. Replace `<username>` with the name of the IAM user. This command will return a list of all inline policies attached to the specified IAM user.

4. **Check the Policies**: The final step is to check the policies. If the `list-user-policies` command returns any policies, it means that the IAM user has inline policies attached. You can get the details of each policy using the following AWS CLI command: `aws iam get-user-policy --user-name <username> --policy-name <policyname>`. Replace `<username>` with the name of the IAM user and `<policyname>` with the name of the policy. This command will return the details of the specified policy.
</Accordion>

<Accordion title='Using Python'>
1. **Setup AWS SDK (Boto3) in Python Environment:**
   First, you need to set up AWS SDK (Boto3) in your Python environment. You can install it using pip:
   ```
   pip install boto3
   ```
   After installing boto3, configure your AWS credentials either by setting up environment variables or by using the AWS CLI.

2. **List all IAM Users:**
   Use the `list_users` method from IAM client to get all the IAM users. Here is a sample script:
   ```python
   import boto3

   # Create IAM client
   iam = boto3.client('iam')

   # List users
   response = iam.list_users()
   for user in response['Users']:
       print('UserName: {0}\nUserID: {1}\nARN: {2}\nCreatedOn: {3}\n'.format(
           user['UserName'], user['UserId'], user['Arn'], user['CreateDate']))
   ```
   This script will list all the IAM users along with their details.

3. **Check for Inline Policies:**
   For each user, you can check if they have any inline policies attached. Use the `list_user_policies` method for this. Here is a sample script:
   ```python
   import boto3

   # Create IAM client
   iam = boto3.client('iam')

   # List users
   response = iam.list_users()
   for user in response['Users']:
       # List inline policies
       inline_policies = iam.list_user_policies(UserName=user['UserName'])
       if inline_policies['PolicyNames']:
           print('User: {0}\nInline Policies: {1}\n'.format(
               user['UserName'], inline_policies['PolicyNames']))
   ```
   This script will list all the IAM users who have inline policies attached.

4. **Analyze the Inline Policies:**
   If you want to analyze the inline policies, you can use the `get_user_policy` method. Here is a sample script:
   ```python
   import boto3

   # Create IAM client
   iam = boto3.client('iam')

   # List users
   response = iam.list_users()
   for user in response['Users']:
       # List inline policies
       inline_policies = iam.list_user_policies(UserName=user['UserName'])
       for policy_name in inline_policies['PolicyNames']:
           # Get inline policy
           policy = iam.get_user_policy(UserName=user['UserName'], PolicyName=policy_name)
           print('User: {0}\nPolicy Name: {1}\nPolicy Document: {2}\n'.format(
               user['UserName'], policy_name, policy['PolicyDocument']))
   ```
   This script will list all the IAM users who have inline policies attached and print the policy documents.
</Accordion>

</AccordionGroup>
</Tab>

<Tab title='Remediation'>
### Remediation

<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
None

#
</Accordion>

<Accordion title='Using CLI'>
To remediate the misconfiguration of principals with inline policies in AWS IAM using AWS CLI, follow these steps:

1. Identify the principals with inline policies:
   - Run the following AWS CLI command to list all IAM users with inline policies:
     ```
     aws iam list-users --query 'Users[?not_null(UserPolicyList)].[UserName, UserPolicyList[?PolicyName].PolicyName]' --output table
     ```
   - Run the following AWS CLI command to list all IAM roles with inline policies:
     ```
     aws iam list-roles --query 'Roles[?not_null(RolePolicyList)].[RoleName, RolePolicyList[?PolicyName].PolicyName]' --output table
     ```

2. Create managed policies for the inline policies:
   - For each IAM user with inline policies, create a managed policy using the following AWS CLI command:
     ```
     aws iam create-policy --policy-name <policy-name> --policy-document file://<policy-document.json>
     ```
     Replace `<policy-name>` with a suitable name for the policy and `<policy-document.json>` with the JSON document containing the inline policy.
   - For each IAM role with inline policies, create a managed policy using the same AWS CLI command.

3. Attach the managed policies to the principals:
   - For each IAM user, attach the managed policy using the following AWS CLI command:
     ```
     aws iam attach-user-policy --user-name <user-name> --policy-arn <policy-arn>
     ```
     Replace `<user-name>` with the IAM user's name and `<policy-arn>` with the ARN of the managed policy.
   - For each IAM role, attach the managed policy using the same AWS CLI command.

4. Verify the remediation:
   - Run the following AWS CLI command to list all IAM users and their associated managed policies:
     ```
     aws iam list-users --query 'Users[].[UserName,join(`, `,UserPolicyList[?PolicyName].PolicyName)]' --output table
     ```
   - Run the following AWS CLI command to list all IAM roles and their associated managed policies:
     ```
     aws iam list-roles --query 'Roles[].[RoleName,join(`, `,RolePolicyList[?PolicyName].PolicyName)]' --output table
     ```

By following these steps, you will remediate the misconfiguration of principals with inline policies in AWS IAM using AWS CLI.
</Accordion>

<Accordion title='Using Python'>
To remediate the misconfiguration of principals with inline policies in AWS IAM using Python, follow these steps:

1. Identify the principals with inline policies:
   - Use the AWS Identity and Access Management (IAM) service to list all the IAM users, groups, and roles.
   - For each principal, check if they have any inline policies attached.

2. Create managed policies:
   - Analyze the inline policies and identify their permissions and resources.
   - Create managed policies in AWS IAM using the `boto3` Python library.
   - Assign the appropriate permissions and resources to the managed policies.

3. Detach inline policies:
   - For each principal with inline policies, use the `boto3` library to detach the inline policies.
   - Keep track of the detached inline policies to later attach them as managed policies.

4. Attach managed policies:
   - For each principal, attach the corresponding managed policies created in step 2 using the `boto3` library.
   - Ensure that the managed policies provide the necessary permissions and resources required by the principals.

5. Update your code or infrastructure-as-code templates:
   - If you have any code or infrastructure-as-code templates that reference the inline policies, update them to use the managed policies instead.
   - Replace the inline policy references with the corresponding managed policy ARNs.

6. Test and validate:
   - Test the remediated IAM configurations thoroughly to ensure that the intended permissions and resources are still accessible.
   - Validate the permissions and resources for each principal to ensure they align with the desired access levels.

7. Monitor and enforce best practices:
   - Regularly monitor your AWS IAM configurations to identify any new instances of principals with inline policies.
   - Enforce best practices by using managed policies instead of inline policies wherever possible.

By following these steps, you can successfully remediate the misconfiguration of principals with inline policies in AWS IAM using Python.
</Accordion>

</AccordionGroup>
</Tab>
</Tabs>

