--- 
slug: gcp_rt_kubernetes_cluster_abac_changes
eventname: google.container.v1.ClusterManager.SetLegacyAbac
title: SetLegacyAbac
sidebar_label: google.container.v1.ClusterManager.SetLegacyAbac
---
                       


### Event Information

#### Meaning

1. The google.container.v1.ClusterManager.SetLegacyAbac event in gcprealtime for Kubernetes Engine refers to a change in the Access Control configuration of a Kubernetes cluster in Google Cloud Platform (GCP).

2. This event indicates that the Legacy Attribute-Based Access Control (ABAC) feature has been enabled or disabled for the Kubernetes cluster.

3. Legacy ABAC is an older access control mechanism in Kubernetes that uses attributes defined in the API objects to control access. Enabling or disabling this feature can impact the way access control policies are enforced within the cluster.

#### Example

- The `google.container.v1.ClusterManager.SetLegacyAbac` event in GCP Realtime for Kubernetes Engine can impact security if the Legacy ABAC (Attribute-Based Access Control) feature is enabled or disabled incorrectly.
- For example, if Legacy ABAC is disabled without implementing an alternative access control mechanism, it can lead to a lack of proper authorization and potentially allow unauthorized access to resources within the Kubernetes cluster.
- Remediation:
  1. If Legacy ABAC is currently enabled and you want to disable it, ensure that you have an alternative access control mechanism in place, such as RBAC (Role-Based Access Control) or IAM (Identity and Access Management) policies.
  2. To disable Legacy ABAC and enable RBAC in a GKE cluster, you can use the following gcloud command:
     ```
     gcloud container clusters update [CLUSTER_NAME] --no-enable-legacy-authorization
     ```
  3. It is recommended to thoroughly test the new access control mechanism before disabling Legacy ABAC to ensure that all necessary permissions are properly configured.

 