--- 
slug: CreateKey
eventname: CreateKey
title: CreateKey
sidebar_label: CreateKey
---
                       
### Event Information

#### Meaning

- The CreateKey event in AWS Key Management Service (KMS) refers to the action of creating a new customer master key (CMK) in KMS.
- When this event occurs, it indicates that a new CMK has been created and is ready to be used for encrypting and decrypting data in AWS services and applications.
- This event is important for auditing and compliance purposes, as it allows organizations to track the creation of new keys and monitor the usage of encryption in their AWS environment.

### Remediation

#### Using Console

- Example of security impact: If the CreateKey operation in AWS Key Management Service (KMS) is misconfigured, it can lead to unauthorized access to sensitive data. For example, if the key policy allows overly permissive access permissions or if the key is not properly protected, an attacker may be able to gain access to encrypted data.

- Remediation steps using AWS console:
  1. Open the AWS Management Console and navigate to the AWS KMS service.
  2. Select the appropriate region where the KMS key is located.
  3. Click on "Keys" in the left navigation pane to view the list of keys.
  4. Identify the key that needs remediation and click on its name to open its details.
  5. Review the key policy to ensure that it follows the principle of least privilege and restricts access to only authorized entities.
  6. If necessary, modify the key policy to tighten the access permissions by removing unnecessary or overly permissive statements.
  7. Enable key rotation if it is not already enabled to ensure that the key is regularly rotated for enhanced security.
  8. Enable and configure key usage logging to monitor and detect any unauthorized access attempts.
  9. Regularly review and audit the key policies and access controls to ensure ongoing compliance and security.

Note: It is recommended to follow AWS security best practices and consult AWS documentation for detailed instructions on securing AWS KMS keys.

#### Using CLI

- Example: If security is impacted with CreateKey in AWS Key Management Service (KMS), it could be due to the lack of proper key policy or incorrect permissions assigned to the key. For instance, if the key policy allows unauthorized users or roles to access or manage the key, it can compromise the security of the key and the data it protects.

- Remediation: To remediate this security issue for AWS KMS using AWS CLI, you can follow these steps:

  1. Identify the key that needs to be remediated by listing all the keys in KMS:
     ```
     aws kms list-keys
     ```

  2. Get the key policy for the identified key:
     ```
     aws kms get-key-policy --key-id <key-id> --policy-name default
     ```

  3. Review the key policy and make necessary changes to restrict access to authorized users or roles. You can modify the key policy using the `put-key-policy` command:
     ```
     aws kms put-key-policy --key-id <key-id> --policy-name default --policy <new-policy-json>
     ```

     Note: Replace `<key-id>` with the actual key ID and `<new-policy-json>` with the updated key policy JSON.

     Ensure that the updated key policy follows the principle of least privilege and only allows necessary permissions to authorized entities.

By following these steps, you can remediate the security impact of CreateKey in AWS KMS and ensure that the key is properly secured.

#### Using Python

- Example of security impact: If the CreateKey operation in AWS Key Management Service (KMS) is not properly secured, it can lead to unauthorized users being able to create encryption keys, potentially compromising the confidentiality and integrity of sensitive data.

- Remediation for AWS KMS using Python:
  1. Implement proper access controls: Ensure that only authorized users or roles have the necessary permissions to create keys in KMS. This can be achieved by using AWS Identity and Access Management (IAM) policies to restrict access to the CreateKey operation.
  
  ```python
  import boto3

  # Create an IAM policy that allows only specific users or roles to create keys
  policy_document = {
      "Version": "2012-10-17",
      "Statement": [
          {
              "Effect": "Allow",
              "Action": "kms:CreateKey",
              "Resource": "*",
              "Condition": {
                  "StringEquals": {
                      "aws:username": "allowed_user"
                  }
              }
          }
      ]
  }

  # Attach the policy to the IAM user or role
  iam_client = boto3.client('iam')
  iam_client.put_user_policy(
      UserName='allowed_user',
      PolicyName='AllowCreateKey',
      PolicyDocument=json.dumps(policy_document)
  )
  ```

  2. Enable AWS CloudTrail logging: AWS CloudTrail can be used to monitor and log all API calls made to KMS, including the CreateKey operation. By enabling CloudTrail, you can have a detailed audit trail of key creation activities, which can help in identifying any unauthorized or suspicious activities.

  ```python
  import boto3

  # Enable CloudTrail logging for KMS
  cloudtrail_client = boto3.client('cloudtrail')
  cloudtrail_client.update_trail(
      Name='my-cloudtrail',
      KmsKeyId='kms-key-id',
      IsLogging=True
  )
  ```

  3. Regularly review and monitor key creation activities: Implement a process to regularly review and monitor the creation of keys in KMS. This can involve analyzing CloudTrail logs, setting up alerts or notifications for key creation events, and conducting periodic security assessments to ensure compliance with security policies and best practices.


 