---
slug: authorization_failures_alarm
title: Authorization Failures Alarm
sidebar_label: Authorization Failures Alarm
---

### More Info:

Any unauthorized API calls made within your AWS account should be monitored using CloudWatch alarms.

### Risk Level

Medium

### Address

Security

### Compliance Standards

CISAWS, CBP, SOC2, HIPAA, ISO27001, AWSWAF, HITRUST

### Remediation

#### Using Console

The "Authorization Failures" alarm in AWS CloudWatch indicates that there have been authorization failures for your AWS resources. To remediate this issue, you can follow the below steps:

1. Login to your AWS console.
2. Navigate to the CloudWatch service.
3. In the left-hand menu, click on "Alarms".
4. Find the "Authorization Failures" alarm and click on it.
5. In the alarm details page, click on the "Actions" dropdown and select "Disable Alarm".
6. Once the alarm is disabled, you can investigate the root cause of the authorization failures and take appropriate actions to remediate it. 

Some of the common causes of authorization failures include incorrect IAM policies, expired credentials, and incorrect resource permissions. You can review your IAM policies, check the status of your credentials, and ensure that your resources have the correct permissions to resolve the issue. Once you have resolved the issue, you can re-enable the alarm to monitor for any future authorization failures.

#### Using CLI

The Authorization Failures alarm in AWS indicates that there have been failed attempts to authenticate or authorize access to AWS resources. To remediate this issue, you can follow these steps using AWS CLI:

1. Open your terminal and install the AWS CLI (if not already installed) by following the instructions provided in the AWS documentation.

2. Once installed, configure the AWS CLI by running the following command:

   ```
   aws configure
   ```

   This command will prompt you to enter your Access Key ID, Secret Access Key, default region name, and default output format.

3. After configuring the AWS CLI, you can use the following command to check the authorization failures alarm:

   ```
   aws cloudwatch describe-alarms --alarm-names "AuthorizationFailures"
   ```

4. If the alarm is in the ALARM state, you can use the following command to disable it:

   ```
   aws cloudwatch disable-alarm-actions --alarm-names "AuthorizationFailures"
   ```

5. To investigate the root cause of the authorization failures, you can use AWS CloudTrail to view the API calls made to AWS services. You can also use AWS Identity and Access Management (IAM) to review the permissions granted to users and roles.

6. Once you have identified the root cause of the authorization failures, you can take appropriate actions such as revoking access keys, updating IAM policies, or disabling unused IAM users or roles.

7. Finally, you can re-enable the authorization failures alarm by using the following command:

   ```
   aws cloudwatch enable-alarm-actions --alarm-names "AuthorizationFailures"
   ```

By following these steps, you will be able to remediate the Authorization Failures alarm in AWS.

#### Using Python

The "Authorization Failures" alarm in AWS CloudWatch is triggered when there are failed attempts to access AWS resources due to insufficient permissions. To remediate this issue using Python, you can follow these steps:

1. Identify the AWS resource that is causing the authorization failures. This can be done by checking the CloudWatch logs or by using AWS Config.

2. Review the permissions assigned to the IAM user or role that is attempting to access the resource. You can use the AWS CLI or the AWS Management Console to view the permissions.

3. Add the necessary permissions to the IAM user or role. This can be done by creating a new IAM policy or modifying an existing one.

4. Test the permissions by attempting to access the resource again. If the authorization failures alarm is no longer triggered, then the issue has been remediated.

Here is some sample Python code that can be used to modify an existing IAM policy to grant additional permissions:

```
import boto3

# Create an IAM client
iam = boto3.client('iam')

# Get the existing policy
policy = iam.get_policy(PolicyArn='arn:aws:iam::123456789012:policy/MyPolicy')

# Add a new permission to the policy
statement = {
    'Effect': 'Allow',
    'Action': 's3:GetObject',
    'Resource': 'arn:aws:s3:::my-bucket/*'
}
policy['Policy']['Statement'].append(statement)

# Update the policy with the new permission
iam.create_policy_version(
    PolicyArn='arn:aws:iam::123456789012:policy/MyPolicy',
    PolicyDocument=str(policy['Policy']),
    SetAsDefault=True
)
```

Note that this is just an example and you will need to modify the code to suit your specific use case.

### Additional Reading:

- [https://docs.aws.amazon.com/awscloudtrail/latest/userguide/use-cloudformation-template-to-create-cloudwatch-alarms.html](https://docs.aws.amazon.com/awscloudtrail/latest/userguide/use-cloudformation-template-to-create-cloudwatch-alarms.html) 

