---
slug: aws_config_changes_alarm
title: AWS Config Changes Alarm
sidebar_label: AWS Config Changes Alarm
---

### More Info:

AWS Config configuration changes should be monitored using CloudWatch alarms.

### Risk Level

Medium

### Address

Security

### Compliance Standards

SOC2, AWSWAF, HITRUST, CISAWS, CBP, NISTCSF


### Triage and Remediation
<Tabs>


<Tab title='Prevention'>
### How to Prevent
<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
To prevent AWS Config Changes Alarm in CloudWatch using the AWS Management Console, follow these steps:

1. **Navigate to CloudWatch:**
   - Sign in to the AWS Management Console.
   - Open the CloudWatch console by searching for "CloudWatch" in the AWS services search bar and selecting it.

2. **Create a New Alarm:**
   - In the CloudWatch console, select "Alarms" from the left-hand navigation pane.
   - Click on the "Create alarm" button.

3. **Select Metric:**
   - Choose the "Select metric" option.
   - In the "Browse" tab, navigate to "AWS Config" and select the appropriate metric that tracks configuration changes (e.g., `ConfigRuleEvaluationFailed`).

4. **Configure Alarm Settings:**
   - Set the conditions for the alarm, such as the threshold value and the period for evaluation.
   - Configure the actions to be taken when the alarm state is triggered, such as sending a notification to an SNS topic or triggering an AWS Lambda function.
   - Review and create the alarm.

By following these steps, you can set up an alarm in CloudWatch to monitor AWS Config changes and take appropriate actions when changes are detected.
</Accordion>

<Accordion title='Using CLI'>
To prevent AWS Config changes from going unnoticed, you can set up an alarm in CloudWatch using the AWS CLI. Here are the steps to achieve this:

1. **Create a CloudWatch Alarm for AWS Config Changes:**
   Use the `put-metric-alarm` command to create an alarm that triggers when there are changes in AWS Config.

   ```sh
   aws cloudwatch put-metric-alarm \
       --alarm-name "AWSConfigChangesAlarm" \
       --alarm-description "Alarm for AWS Config changes" \
       --metric-name "ConfigRuleComplianceChange" \
       --namespace "AWS/Config" \
       --statistic "Sum" \
       --period 300 \
       --threshold 1 \
       --comparison-operator "GreaterThanOrEqualToThreshold" \
       --evaluation-periods 1 \
       --alarm-actions "arn:aws:sns:us-east-1:123456789012:MySNSTopic" \
       --ok-actions "arn:aws:sns:us-east-1:123456789012:MySNSTopic" \
       --dimensions Name=ConfigRuleName,Value=MyConfigRule
   ```

2. **Create an SNS Topic for Alarm Notifications:**
   Use the `create-topic` command to create an SNS topic that will be used to send notifications when the alarm is triggered.

   ```sh
   aws sns create-topic --name MySNSTopic
   ```

3. **Subscribe to the SNS Topic:**
   Use the `subscribe` command to subscribe an email endpoint to the SNS topic so that you receive notifications.

   ```sh
   aws sns subscribe \
       --topic-arn "arn:aws:sns:us-east-1:123456789012:MySNSTopic" \
       --protocol email \
       --notification-endpoint "your-email@example.com"
   ```

4. **Enable AWS Config Rules:**
   Ensure that AWS Config rules are enabled and configured to monitor the resources you are interested in. Use the `put-config-rule` command to create or update a Config rule.

   ```sh
   aws configservice put-config-rule \
       --config-rule '{"ConfigRuleName": "MyConfigRule", "Source": {"Owner": "AWS", "SourceIdentifier": "S3_BUCKET_VERSIONING_ENABLED"}}'
   ```

By following these steps, you will set up a CloudWatch alarm to monitor AWS Config changes and receive notifications via SNS, helping you to prevent misconfigurations from going unnoticed.
</Accordion>

<Accordion title='Using Python'>
To prevent AWS Config changes from going unnoticed, you can set up CloudWatch Alarms using Python scripts. Here are the steps to achieve this:

### Step 1: Install Boto3
Ensure you have the Boto3 library installed, which is the AWS SDK for Python.

```bash
pip install boto3
```

### Step 2: Create an IAM Role with Necessary Permissions
Ensure you have an IAM role with the necessary permissions to create CloudWatch Alarms and access AWS Config. Attach the following policies to your IAM role:
- `CloudWatchFullAccess`
- `AWSConfigUserAccess`

### Step 3: Write the Python Script
Below is a Python script to create a CloudWatch Alarm for AWS Config changes:

```python
import boto3

# Initialize a session using Amazon CloudWatch
cloudwatch = boto3.client('cloudwatch', region_name='us-west-2')

# Create CloudWatch Alarm
response = cloudwatch.put_metric_alarm(
    AlarmName='AWSConfigChangesAlarm',
    AlarmDescription='Alarm when AWS Config changes are detected',
    ActionsEnabled=True,
    AlarmActions=[
        'arn:aws:sns:us-west-2:123456789012:MySNSTopic'  # Replace with your SNS topic ARN
    ],
    MetricName='ConfigChanges',
    Namespace='AWS/Config',
    Statistic='Sum',
    Period=300,
    EvaluationPeriods=1,
    Threshold=1,
    ComparisonOperator='GreaterThanOrEqualToThreshold',
    TreatMissingData='notBreaching'
)

print("CloudWatch Alarm created successfully")
```

### Step 4: Run the Script
Execute the script to create the CloudWatch Alarm.

```bash
python create_cloudwatch_alarm.py
```

### Summary
1. **Install Boto3**: Ensure you have the Boto3 library installed.
2. **Create IAM Role**: Ensure the IAM role has `CloudWatchFullAccess` and `AWSConfigUserAccess` policies.
3. **Write Python Script**: Use the provided script to create a CloudWatch Alarm for AWS Config changes.
4. **Run the Script**: Execute the script to set up the alarm.

By following these steps, you can effectively monitor AWS Config changes using CloudWatch Alarms, ensuring that any configuration changes are promptly detected and addressed.
</Accordion>

</AccordionGroup>
</Tab>

<Tab title='Cause'>
### Check Cause
<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
1. Log in to the AWS Management Console and navigate to the CloudWatch service. 

2. In the CloudWatch dashboard, click on "Alarms" in the left-hand navigation pane. This will display a list of all the alarms that have been configured.

3. Look for the AWS Config Changes Alarm in the list. If the alarm is in the "ALARM" state, it means that the changes in AWS Config have triggered the alarm. 

4. To get more details about the alarm, click on the alarm name. This will open a new page where you can see the metrics and conditions that are associated with the alarm. You can also see the history of the alarm, which includes the times when the alarm state changed and the reason for the change.
</Accordion>

<Accordion title='Using CLI'>
1. Install and configure AWS CLI: Before you can start using AWS CLI, you need to install it on your local machine. You can download it from the official AWS website. After installation, you need to configure it with your AWS account credentials. You can do this by running the command `aws configure` and then entering your AWS Access Key ID, Secret Access Key, Default region name, and Default output format when prompted.

2. List all CloudWatch alarms: You can use the `describe-alarms` command to list all the CloudWatch alarms in your AWS account. The command is as follows:

   ```
   aws cloudwatch describe-alarms
   ```

   This command will return a list of all the CloudWatch alarms along with their details.

3. Filter AWS Config Changes Alarm: If you want to check a specific alarm, for example, AWS Config Changes Alarm, you can filter the results using the `--alarm-names` option followed by the name of the alarm. The command is as follows:

   ```
   aws cloudwatch describe-alarms --alarm-names "AWS Config Changes Alarm"
   ```

   This command will return the details of the AWS Config Changes Alarm.

4. Check the state of the alarm: The state of the alarm can be checked in the output of the above command. Look for the `StateValue` field in the output. If the state is `OK`, it means that no misconfigurations are detected. If the state is `ALARM`, it means that a misconfiguration is detected.
</Accordion>

<Accordion title='Using Python'>
1. **Install AWS SDK for Python (Boto3):** To interact with AWS services, you need to install Boto3. You can install it using pip:
   ```
   pip install boto3
   ```

2. **Configure AWS Credentials:** Before you can interact with AWS services, you need to set up authentication credentials for your AWS account using the AWS CLI. If you have the AWS CLI installed, then you can use it to configure your credentials file:
   ```
   aws configure
   ```
   Alternatively, you can create the credential file yourself. By default, its location is at `~/.aws/credentials`. At a minimum, the credentials file should specify the access key and secret access key. To specify these for the `default` profile, you can use the following format:
   ```
   [default]
   aws_access_key_id = YOUR_ACCESS_KEY
   aws_secret_access_key = YOUR_SECRET_KEY
   ```

3. **Create a Python Script to Check AWS Config Changes Alarm in Cloud Watch:**
   ```python
   import boto3

   def check_config_changes_alarm():
       client = boto3.client('cloudwatch')
       response = client.describe_alarms()
       alarms = response['MetricAlarms']
       for alarm in alarms:
           if alarm['AlarmName'] == 'ConfigChangesAlarm':
               return alarm['StateValue']

   alarm_state = check_config_changes_alarm()
   print(f'ConfigChangesAlarm state: {alarm_state}')
   ```
   This script creates a CloudWatch client, retrieves all the alarms, and checks if there is an alarm named 'ConfigChangesAlarm'. If it exists, it prints the state of the alarm.

4. **Run the Python Script:** Finally, run the python script. If the 'ConfigChangesAlarm' is properly configured and no changes have been detected, the script should print 'OK'. If changes have been detected, it will print 'ALARM'. If the alarm is not configured correctly, it will print 'INSUFFICIENT_DATA'.
</Accordion>

</AccordionGroup>
</Tab>

<Tab title='Remediation'>
### Remediation

<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
AWS Config Changes Alarm is triggered when there are changes made to the AWS resources that are being monitored by AWS Config. To remediate this issue, follow these steps:

1. Login to the AWS Management Console and navigate to the AWS Config service.

2. Click on the "Rules" option in the left-hand menu.

3. Locate the rule that is triggering the alarm and click on the rule name.

4. Review the rule details and make sure it is configured correctly. If not, update the rule configuration as needed.

5. Click on the "Remediation actions" tab and select the option to "Automatically remediate noncompliant resources".

6. Configure the remediation action based on the specific rule and resource that is triggering the alarm. For example, if the rule is related to security group changes, you may want to automatically revert the security group to its previous state.

7. Save the remediation action and test it by making a change that triggers the alarm. The remediation action should automatically revert the change and resolve the alarm.

8. Monitor the AWS Config Changes Alarm to ensure that it does not trigger again for the same issue. If it does, review the rule configuration and remediation action to ensure they are still valid.

#
</Accordion>

<Accordion title='Using CLI'>
The AWS Config Changes Alarm is an AWS Config rule that checks whether an alarm exists for any changes to the AWS resources in your account. If no alarm exists, the rule is non-compliant.

To remediate this misconfiguration, you can follow the below steps using the AWS CLI:

1. List all the rules in your AWS Config:

```
aws configservice describe-config-rules
```

2. Check if the "AWS Config Changes Alarm" rule is non-compliant:

```
aws configservice get-compliance-details-by-config-rule --config-rule-name AWS-Config-Changes-Alarm
```

3. If the rule is non-compliant, create an alarm for AWS Config changes using the AWS CLI:

```
aws cloudwatch put-metric-alarm --alarm-name AWS-Config-Changes-Alarm --alarm-description "Alarm for AWS Config Changes" --metric-name ConfigurationChanges --namespace AWS/Config --statistic Sum --period 300 --threshold 1 --comparison-operator GreaterThanOrEqualToThreshold --evaluation-periods 1 --alarm-actions <your-SNS-topic-ARN>
```

Note: Replace `<your-SNS-topic-ARN>` with the ARN of the SNS topic you want to receive the alarm notifications.

4. Verify if the alarm has been created successfully:

```
aws cloudwatch describe-alarms --alarm-names AWS-Config-Changes-Alarm
```

5. Finally, check if the "AWS Config Changes Alarm" rule is compliant:

```
aws configservice get-compliance-details-by-config-rule --config-rule-name AWS-Config-Changes-Alarm
```

If the rule is compliant, then the misconfiguration has been remediated successfully.
</Accordion>

<Accordion title='Using Python'>
The AWS Config Changes Alarm is triggered when there are changes made to the AWS resources that are being monitored by AWS Config. To remediate this issue, you can create a Lambda function in Python that will automatically respond to the alarm.

Here are the step-by-step instructions to remediate the AWS Config Changes Alarm using Python:

1. Create an AWS Lambda function: Go to the AWS Management Console, navigate to the Lambda service, and click on the "Create function" button. Choose "Author from scratch", enter a name for the function, and select "Python 3.x" as the runtime.

2. Add the necessary permissions: In the "Permissions" section of the Lambda function, add the "AWS Config" trigger to the function. This will allow the function to be triggered when the AWS Config Changes Alarm is triggered.

3. Write the Python code: In the "Function code" section of the Lambda function, write the Python code to remediate the issue. Here is an example code that will stop and start an EC2 instance when the AWS Config Changes Alarm is triggered:

```
import boto3

def lambda_handler(event, context):
    ec2 = boto3.resource('ec2')
    instance_id = event['detail']['configurationItem']['resourceId']
    instance = ec2.Instance(instance_id)
    if event['detail']['configurationItemDiff']['changeType'] == 'CREATE':
        instance.stop()
        instance.wait_until_stopped()
        instance.start()
        instance.wait_until_running()
    elif event['detail']['configurationItemDiff']['changeType'] == 'DELETE':
        instance.start()
        instance.wait_until_running()
```

4. Test the function: Once you have written the code, test the function by triggering the AWS Config Changes Alarm. You should see the Lambda function being triggered and the EC2 instance being stopped and started automatically.

5. Deploy the function: Once you have tested the function, deploy it to your AWS account by clicking on the "Deploy" button in the Lambda console. Make sure to set the appropriate permissions and triggers for the function.

By following these steps, you can easily remediate the AWS Config Changes Alarm using Python and automate the response to this issue.
</Accordion>

</AccordionGroup>
</Tab>
</Tabs>
### Additional Reading:

- [https://docs.aws.amazon.com/config/latest/developerguide/monitor-config-with-cloudwatchevents.html](https://docs.aws.amazon.com/config/latest/developerguide/monitor-config-with-cloudwatchevents.html) 

