---
slug: cloudwatch_log_groups_not_encrypted
title: CloudWatch Log Groups Should Be Encrypted
sidebar_label: CloudWatch Log Groups Should Be Encrypted
---

### More Info:

Cloudwatch loggroups should be encrypted

### Risk Level

High

### Address

Security

### Compliance Standards

HIPAA,PCIDSS,GDPR,CISAWS,CBP,NIST,SOC2,AWSWAF,SEBI,RBI_UCB

### Remediation
### Using Console:

1. **Create an AWS KMS Key:**
   - Go to the AWS KMS console.
   - Click on "Create key" and follow the prompts to create a symmetric key.
   - Make note of the Key ID and ARN provided.

2. **Set Permissions on the KMS Key:**
   - After creating the key, navigate to it in the AWS KMS console.
   - Select "Key policy" from the left-hand menu.
   - Edit the key policy to allow CloudWatch Logs service principal to use the key.
   - Add permissions as described in the documentation under "Step 2: Set permissions on the KMS key".

3. **Associate the KMS Key with a Log Group:**
   - Navigate to the CloudWatch Logs console.
   - Find the log group you want to encrypt or create a new log group.
   - During log group creation or by selecting an existing log group, associate the KMS key with it.

4. **Disassociate Key from a Log Group (if needed):**
   - If you need to disassociate the KMS key from a log group, navigate to the CloudWatch Logs console, select the log group, and disassociate the key.

Enabling encryption from console has some limitations, you cannot associate KMS key with log group from Console, so the above steps are invalid as of now.

### Using CLI:

1. **Create an AWS KMS Key:**
   ```bash
   aws kms create-key
   ```

2. **Set Permissions on the KMS Key:**
   - Retrieve the default policy for the key:
     ```bash
     aws kms get-key-policy --key-id <key-id> --policy-name default --output text > ./policy.json
     ```
   - Edit `policy.json` to include the necessary permissions as described in the documentation.
   - Apply the updated policy to the key:
     ```bash
     aws kms put-key-policy --key-id <key-id> --policy-name default --policy file://policy.json
     ```

3. **Associate the KMS Key with a Log Group:**
   - To associate the key during log group creation:
     ```bash
     aws logs create-log-group --log-group-name <log-group-name> --kms-key-id <key-arn>
     ```
   - To associate the key with an existing log group:
     ```bash
     aws logs associate-kms-key --log-group-name <log-group-name> --kms-key-id <key-arn>
     ```

4. **Disassociate Key from a Log Group (if needed):**
   ```bash
   aws logs disassociate-kms-key --log-group-name <log-group-name>
   ```

### Using Python:

You can use the `boto3` library in Python to achieve the same tasks programmatically.

1. **Create an AWS KMS Key:**
   ```python
   import boto3

   kms = boto3.client('kms')
   response = kms.create_key()
   ```

2. **Set Permissions on the KMS Key:**
   - Retrieve, edit, and apply the policy similar to the AWS CLI method.

3. **Associate the KMS Key with a Log Group:**
   ```python
   import boto3

   logs = boto3.client('logs')
   logs.create_log_group(logGroupName='<log-group-name>', kmsKeyId='<key-arn>')
   ```

4. **Disassociate Key from a Log Group (if needed):**
   ```python
   import boto3

   logs = boto3.client('logs')
   logs.disassociate_kms_key(logGroupName='<log-group-name>')
   ```

Ensure you replace placeholders such as `<key-id>`, `<key-arn>`, and `<log-group-name>` with actual values in the commands or code snippets. Also, review and adjust the permissions and conditions according to your specific requirements.