---
slug: custom_schema_registry_policy_attached
title: Resource Policy Attachment In Custom Schema Registry
sidebar_label: Resource Policy Attachment In Custom Schema Registry
---

### More Info:

This rule checks if custom Amazon EventBridge schema registries have a resource policy attached. It ensures that proper access controls are in place for custom schema registries to prevent unauthorized access and ensure compliance with security best practices.

### Risk Level

Medium

### Address

Security

### Compliance Standards

CBP



<Tabs><Tab title='Cause'>
### Check Cause

#### Using Console

1. Log in to the AWS Management Console and navigate to the AWS Glue service.
2. In the AWS Glue console, click on "Data Catalog" in the left navigation pane.
3. Click on "Databases" and select the database that contains the table with the custom schema registry.
4. Click on "Tables" and select the table with the custom schema registry.
5. In the table details page, click on "View details" under the "Schema" section. Here, you can check the resource policy attachment in the custom schema registry. If the policy is not attached, it indicates a misconfiguration.

#### Using CLI

1. First, you need to install and configure AWS CLI on your local machine. You can do this by following the instructions provided by AWS. Make sure you have the necessary permissions to access the resources.

2. Once the AWS CLI is set up, you can use the `get-resource-policies` command to retrieve the resource policies attached to your schema registry. The command is as follows:

   ```
   aws schemas list-resource-policies --registry-name <your-registry-name>
   ```
   Replace `<your-registry-name>` with the name of your schema registry.

3. The output of the command will be a JSON object that contains the details of the resource policies attached to your schema registry. If the `Policies` array in the output is empty, it means that there are no resource policies attached to your schema registry.

4. To check the details of a specific resource policy, you can use the `describe-resource-policy` command as follows:

   ```
   aws schemas describe-resource-policy --registry-name <your-registry-name> --policy-name <your-policy-name>
   ```
   Replace `<your-registry-name>` with the name of your schema registry and `<your-policy-name>` with the name of the resource policy you want to check.

Remember to replace the placeholders with your actual values.

#### Using Python

1. Install the necessary Python libraries: Before you start, you need to install the AWS SDK for Python (Boto3) in your local environment. You can do this using pip:

   ```
   pip install boto3
   ```

2. Set up AWS credentials: You need to configure your AWS credentials. You can do this by creating the files ~/.aws/credentials and ~/.aws/config:

   ~/.aws/credentials:
   ```
   [default]
   aws_access_key_id = YOUR_ACCESS_KEY
   aws_secret_access_key = YOUR_SECRET_KEY
   ```

   ~/.aws/config:
   ```
   [default]
   region=us-east-1
   ```

3. Write a Python script to check the Resource Policy Attachment in Custom Schema Registry:

   ```python
   import boto3

   # Create CloudWatch client
   cloudwatch = boto3.client('cloudwatch')

   # List metrics through the pagination interface
   paginator = cloudwatch.get_paginator('list_metrics')

   for response in paginator.paginate(Dimensions=[{'Name': 'RegistryName'}],
                                      MetricName='ResourcePolicyAttachment',
                                      Namespace='AWS/Schemas'):
       print(response['Metrics'])
   ```

4. Run the Python script: Execute the script in your local environment. If there are any misconfigurations related to Resource Policy Attachment in Custom Schema Registry, they will be printed out. If the script doesn't return any metrics, it means there are no misconfigurations.

</Tab>


<Tab title='Remediation'>
### Remediation

#### Using Console

To remediate the misconfiguration of a resource policy attachment in a custom schema registry for AWS CloudWatch using the AWS Management Console, follow these step-by-step instructions:

1. **Access AWS Console**: Go to the AWS Management Console at https://aws.amazon.com and log in using your AWS account credentials.

2. **Navigate to CloudWatch**: From the AWS Management Console dashboard, search for "CloudWatch" in the search bar at the top and select CloudWatch from the search results.

3. **Select Custom Schema Registry**: In the CloudWatch console, navigate to the left-hand side menu and select "Custom Schema Registry" under the "Schemas" section.

4. **Identify Misconfigured Resource Policy**: Look for the custom schema registry resource that has a misconfigured resource policy attachment. This misconfiguration could be due to incorrect permissions or a policy that is not compliant with best practices.

5. **Edit Resource Policy**: Select the custom schema registry resource that has the misconfigured resource policy attachment. Look for an option to edit the resource policy or access control settings.

6. **Update Resource Policy**: Update the resource policy to ensure that it follows the principle of least privilege. Make sure that only the necessary permissions are granted to the required entities and that the policy is compliant with AWS best practices.

7. **Review Changes**: Before saving the updated resource policy, review the changes to ensure that the permissions are set correctly and that there are no unnecessary or overly permissive permissions granted.

8. **Save Changes**: Once you have reviewed and confirmed the updated resource policy, save the changes to apply the remediation for the misconfigured resource policy attachment in the custom schema registry for AWS CloudWatch.

By following these steps and ensuring that the resource policy for the custom schema registry in AWS CloudWatch is correctly configured and compliant with best practices, you can effectively remediate the misconfiguration of a resource policy attachment in the AWS CloudWatch service using the AWS Management Console.

#### Using CLI

To remediate the misconfiguration of resource policy attachment in a custom schema registry for AWS CloudWatch using AWS CLI, you can follow these steps:

1. **Identify the Custom Schema Registry**: First, identify the custom schema registry that is misconfigured in AWS CloudWatch. You can use the AWS CLI command `aws glue list-schemas` to list all the schemas and identify the custom schema registry.

2. **Check Resource Policy Attachment**: Use the AWS CLI command `aws glue get-resource-policy --resource-arn <schema-registry-arn>` to check the resource policy attached to the custom schema registry. Look for any misconfigurations in the policy that allow unauthorized access.

3. **Update Resource Policy**: If you find any misconfigurations in the resource policy, you can update it using the AWS CLI command `aws glue put-resource-policy --policy-in-json file://policy.json --resource-arn <schema-registry-arn>`. Ensure that the `policy.json` file contains the updated and secure resource policy.

4. **Verify Policy Attachment**: After updating the resource policy, verify the policy attachment using the AWS CLI command `aws glue get-resource-policy --resource-arn <schema-registry-arn>` to confirm that the policy is correctly attached to the custom schema registry.

5. **Monitor Access**: Regularly monitor the access to the custom schema registry in AWS CloudWatch and set up appropriate alarms and notifications to detect any unauthorized access attempts.

By following these steps and ensuring that the resource policy attached to the custom schema registry is secure and properly configured, you can remediate the misconfiguration of resource policy attachment in AWS CloudWatch using AWS CLI.

#### Using Python

To remediate the misconfiguration of a resource policy attachment in custom schema registry for AWS CloudWatch using Python, you can follow these steps:

1. **Identify the Misconfiguration**: First, identify the custom schema registry resource that has an incorrect or missing resource policy attached.

2. **Create an Appropriate Resource Policy**: You can create a new resource policy that allows the necessary permissions for the custom schema registry resource. Here is an example of a minimal resource policy that allows full access to the custom schema registry:
   ```json
   {
       "Version": "2012-10-17",
       "Statement": [
           {
               "Effect": "Allow",
               "Principal": {
                   "Service": "firehose.amazonaws.com"
               },
               "Action": "registry:*",
               "Resource": "arn:aws:schemas:us-east-1:123456789012:registry/MyCustomSchemaRegistry"
           }
       ]
   }
   ```

3. **Attach the Resource Policy**: Use the AWS SDK for Python (Boto3) to attach the newly created resource policy to the custom schema registry resource. Here is an example Python code snippet to attach the resource policy:
   ```python
   import boto3

   client = boto3.client('schemas')

   response = client.put_resource_policy(
       RegistryName='MyCustomSchemaRegistry',
       Policy='{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"firehose.amazonaws.com"},"Action":"registry:*","Resource":"arn:aws:schemas:us-east-1:123456789012:registry/MyCustomSchemaRegistry"}]'
   )

   print(response)
   ```

4. **Verify the Resource Policy**: After attaching the resource policy, verify that the custom schema registry resource now has the correct resource policy attached by listing the resource policies for the registry.

By following these steps and customizing the resource policy and Python code as per your specific requirements, you can successfully remediate the misconfiguration of a resource policy attachment in a custom schema registry for AWS CloudWatch using Python.



</Tab>
</Tabs>