---
slug: custom_schema_registry_policy_attached
title: Resource Policy Attachment In Custom Schema Registry
sidebar_label: Resource Policy Attachment In Custom Schema Registry
---

### More Info:

This rule checks if custom Amazon EventBridge schema registries have a resource policy attached. It ensures that proper access controls are in place for custom schema registries to prevent unauthorized access and ensure compliance with security best practices.

### Risk Level

Medium

### Address

Security

### Compliance Standards

CBP


### Triage and Remediation
<Tabs>

<Tab title='Cause'>
### Check Cause
<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
1. Log in to the AWS Management Console and navigate to the AWS Glue service.
2. In the AWS Glue service, click on "Registries" in the left-hand navigation pane. This will display a list of all the registries in your AWS account.
3. Click on the name of the registry you want to check. This will open the details page for that registry.
4. On the details page, look for the "Resource Policy" section. If a policy is attached, it will be displayed here. If no policy is attached, the section will be blank.
</Accordion>

<Accordion title='Using CLI'>
1. First, you need to install and configure AWS CLI on your local machine. You can do this by following the instructions provided by AWS. Make sure you have the necessary permissions to access the resources.

2. Once the AWS CLI is set up, you can use the `get-resource-policies` command to retrieve the resource policies attached to your schema registry. The command would look something like this:

   ```
   aws schemas get-resource-policies --registry-name <your-registry-name>
   ```
   Replace `<your-registry-name>` with the name of your schema registry.

3. The output of this command will be a JSON object that contains all the resource policies attached to your schema registry. You can inspect this JSON object to check for any misconfigurations. 

4. If you want to automate this process, you can write a Python script that uses the `boto3` library to interact with AWS. The script would use the `get_resource_policies` method of the `Schemas` client to retrieve the resource policies. Here is a sample script:

   ```python
   import boto3

   # Create a client
   client = boto3.client('schemas')

   # Get the resource policies
   response = client.get_resource_policies(RegistryName='<your-registry-name>')

   # Print the resource policies
   print(response['ResourcePolicies'])
   ```
   Replace `<your-registry-name>` with the name of your schema registry. This script will print the resource policies attached to your schema registry.
</Accordion>

<Accordion title='Using Python'>
1. Install the necessary Python libraries: Before you can start writing Python scripts to interact with AWS, you need to install the necessary libraries. The primary library you need is boto3, which is the Amazon Web Services (AWS) Software Development Kit (SDK) for Python. You can install it using pip:

   ```python
   pip install boto3
   ```

2. Set up AWS credentials: Boto3 needs your AWS credentials (access key and secret access key) to interact with AWS services. You can set them in your environment variables:

   ```python
   import os
   os.environ["AWS_ACCESS_KEY_ID"] = "your_access_key"
   os.environ["AWS_SECRET_ACCESS_KEY"] = "your_secret_key"
   ```

3. Write a Python script to list all the policies attached to your AWS resources. Here's a simple script that lists all the policies attached to your S3 buckets:

   ```python
   import boto3

   # Create a session using your AWS credentials
   session = boto3.Session(
       aws_access_key_id='YOUR_ACCESS_KEY',
       aws_secret_access_key='YOUR_SECRET_KEY',
   )

   # Create an S3 resource object using the session
   s3 = session.resource('s3')

   # List all the buckets
   for bucket in s3.buckets.all():
       # Get the bucket policy
       policy = s3.BucketPolicy(bucket.name)
       try:
           # Print the policy
           print(policy.policy)
       except Exception as e:
           print(f"No policy attached to {bucket.name}")
   ```

4. Analyze the output: The script will print the policies attached to each of your S3 buckets. You can analyze these policies to check if they are misconfigured. For example, you might want to check if any policies are granting public access to your buckets. If the script prints "No policy attached to [bucket name]", that means there is no policy attached to that bucket, which might be a misconfiguration if you expect a policy to be attached.
</Accordion>

</AccordionGroup>
</Tab>

<Tab title='Remediation'>
### Remediation

<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
To remediate the misconfiguration of a resource policy attachment in a custom schema registry for AWS CloudWatch using the AWS Management Console, follow these step-by-step instructions:

1. **Access AWS Console**: Go to the AWS Management Console at https://aws.amazon.com and log in using your AWS account credentials.

2. **Navigate to CloudWatch**: From the AWS Management Console dashboard, search for "CloudWatch" in the search bar at the top and select CloudWatch from the search results.

3. **Select Custom Schema Registry**: In the CloudWatch console, navigate to the left-hand side menu and select "Custom Schema Registry" under the "Schemas" section.

4. **Identify Misconfigured Resource Policy**: Look for the custom schema registry resource that has a misconfigured resource policy attachment. This misconfiguration could be due to incorrect permissions or a policy that is not compliant with best practices.

5. **Edit Resource Policy**: Select the custom schema registry resource that has the misconfigured resource policy attachment. Look for an option to edit the resource policy or access control settings.

6. **Update Resource Policy**: Update the resource policy to ensure that it follows the principle of least privilege. Make sure that only the necessary permissions are granted to the required entities and that the policy is compliant with AWS best practices.

7. **Review Changes**: Before saving the updated resource policy, review the changes to ensure that the permissions are set correctly and that there are no unnecessary or overly permissive permissions granted.

8. **Save Changes**: Once you have reviewed and confirmed the updated resource policy, save the changes to apply the remediation for the misconfigured resource policy attachment in the custom schema registry for AWS CloudWatch.

By following these steps and ensuring that the resource policy for the custom schema registry in AWS CloudWatch is correctly configured and compliant with best practices, you can effectively remediate the misconfiguration of a resource policy attachment in the AWS CloudWatch service using the AWS Management Console.

#
</Accordion>

<Accordion title='Using CLI'>
To remediate the misconfiguration of resource policy attachment in a custom schema registry for AWS CloudWatch using AWS CLI, you can follow these steps:

1. **Identify the Custom Schema Registry**: First, identify the custom schema registry that is misconfigured in AWS CloudWatch. You can use the AWS CLI command `aws glue list-schemas` to list all the schemas and identify the custom schema registry.

2. **Check Resource Policy Attachment**: Use the AWS CLI command `aws glue get-resource-policy --resource-arn <schema-registry-arn>` to check the resource policy attached to the custom schema registry. Look for any misconfigurations in the policy that allow unauthorized access.

3. **Update Resource Policy**: If you find any misconfigurations in the resource policy, you can update it using the AWS CLI command `aws glue put-resource-policy --policy-in-json file://policy.json --resource-arn <schema-registry-arn>`. Ensure that the `policy.json` file contains the updated and secure resource policy.

4. **Verify Policy Attachment**: After updating the resource policy, verify the policy attachment using the AWS CLI command `aws glue get-resource-policy --resource-arn <schema-registry-arn>` to confirm that the policy is correctly attached to the custom schema registry.

5. **Monitor Access**: Regularly monitor the access to the custom schema registry in AWS CloudWatch and set up appropriate alarms and notifications to detect any unauthorized access attempts.

By following these steps and ensuring that the resource policy attached to the custom schema registry is secure and properly configured, you can remediate the misconfiguration of resource policy attachment in AWS CloudWatch using AWS CLI.
</Accordion>

<Accordion title='Using Python'>
To remediate the misconfiguration of a resource policy attachment in custom schema registry for AWS CloudWatch using Python, you can follow these steps:

1. **Identify the Misconfiguration**: First, identify the custom schema registry resource that has an incorrect or missing resource policy attached.

2. **Create an Appropriate Resource Policy**: You can create a new resource policy that allows the necessary permissions for the custom schema registry resource. Here is an example of a minimal resource policy that allows full access to the custom schema registry:
   ```json
   {
       "Version": "2012-10-17",
       "Statement": [
           {
               "Effect": "Allow",
               "Principal": {
                   "Service": "firehose.amazonaws.com"
               },
               "Action": "registry:*",
               "Resource": "arn:aws:schemas:us-east-1:123456789012:registry/MyCustomSchemaRegistry"
           }
       ]
   }
   ```

3. **Attach the Resource Policy**: Use the AWS SDK for Python (Boto3) to attach the newly created resource policy to the custom schema registry resource. Here is an example Python code snippet to attach the resource policy:
   ```python
   import boto3

   client = boto3.client('schemas')

   response = client.put_resource_policy(
       RegistryName='MyCustomSchemaRegistry',
       Policy='{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"firehose.amazonaws.com"},"Action":"registry:*","Resource":"arn:aws:schemas:us-east-1:123456789012:registry/MyCustomSchemaRegistry"}]'
   )

   print(response)
   ```

4. **Verify the Resource Policy**: After attaching the resource policy, verify that the custom schema registry resource now has the correct resource policy attached by listing the resource policies for the registry.

By following these steps and customizing the resource policy and Python code as per your specific requirements, you can successfully remediate the misconfiguration of a resource policy attachment in a custom schema registry for AWS CloudWatch using Python.
</Accordion>

</AccordionGroup>
</Tab>
</Tabs>

