---
slug: custom_schema_registry_policy_attached
title: Resource Policy Attachment In Custom Schema Registry
sidebar_label: Resource Policy Attachment In Custom Schema Registry
---

### More Info:

This rule checks if custom Amazon EventBridge schema registries have a resource policy attached. It ensures that proper access controls are in place for custom schema registries to prevent unauthorized access and ensure compliance with security best practices.

### Risk Level

Medium

### Address

Security

### Compliance Standards

CBP


### Triage and Remediation
<Tabs>
<Tab title='Remediation'>
### Remediation

<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
To remediate the misconfiguration of a resource policy attachment in a custom schema registry for AWS CloudWatch using the AWS Management Console, follow these step-by-step instructions:

1. **Access AWS Console**: Go to the AWS Management Console at https://aws.amazon.com and log in using your AWS account credentials.

2. **Navigate to CloudWatch**: From the AWS Management Console dashboard, search for "CloudWatch" in the search bar at the top and select CloudWatch from the search results.

3. **Select Custom Schema Registry**: In the CloudWatch console, navigate to the left-hand side menu and select "Custom Schema Registry" under the "Schemas" section.

4. **Identify Misconfigured Resource Policy**: Look for the custom schema registry resource that has a misconfigured resource policy attachment. This misconfiguration could be due to incorrect permissions or a policy that is not compliant with best practices.

5. **Edit Resource Policy**: Select the custom schema registry resource that has the misconfigured resource policy attachment. Look for an option to edit the resource policy or access control settings.

6. **Update Resource Policy**: Update the resource policy to ensure that it follows the principle of least privilege. Make sure that only the necessary permissions are granted to the required entities and that the policy is compliant with AWS best practices.

7. **Review Changes**: Before saving the updated resource policy, review the changes to ensure that the permissions are set correctly and that there are no unnecessary or overly permissive permissions granted.

8. **Save Changes**: Once you have reviewed and confirmed the updated resource policy, save the changes to apply the remediation for the misconfigured resource policy attachment in the custom schema registry for AWS CloudWatch.

By following these steps and ensuring that the resource policy for the custom schema registry in AWS CloudWatch is correctly configured and compliant with best practices, you can effectively remediate the misconfiguration of a resource policy attachment in the AWS CloudWatch service using the AWS Management Console.

#
</Accordion>

<Accordion title='Using CLI'>
To remediate the misconfiguration of resource policy attachment in a custom schema registry for AWS CloudWatch using AWS CLI, you can follow these steps:

1. **Identify the Custom Schema Registry**: First, identify the custom schema registry that is misconfigured in AWS CloudWatch. You can use the AWS CLI command `aws glue list-schemas` to list all the schemas and identify the custom schema registry.

2. **Check Resource Policy Attachment**: Use the AWS CLI command `aws glue get-resource-policy --resource-arn <schema-registry-arn>` to check the resource policy attached to the custom schema registry. Look for any misconfigurations in the policy that allow unauthorized access.

3. **Update Resource Policy**: If you find any misconfigurations in the resource policy, you can update it using the AWS CLI command `aws glue put-resource-policy --policy-in-json file://policy.json --resource-arn <schema-registry-arn>`. Ensure that the `policy.json` file contains the updated and secure resource policy.

4. **Verify Policy Attachment**: After updating the resource policy, verify the policy attachment using the AWS CLI command `aws glue get-resource-policy --resource-arn <schema-registry-arn>` to confirm that the policy is correctly attached to the custom schema registry.

5. **Monitor Access**: Regularly monitor the access to the custom schema registry in AWS CloudWatch and set up appropriate alarms and notifications to detect any unauthorized access attempts.

By following these steps and ensuring that the resource policy attached to the custom schema registry is secure and properly configured, you can remediate the misconfiguration of resource policy attachment in AWS CloudWatch using AWS CLI.
</Accordion>

<Accordion title='Using Python'>
To remediate the misconfiguration of a resource policy attachment in custom schema registry for AWS CloudWatch using Python, you can follow these steps:

1. **Identify the Misconfiguration**: First, identify the custom schema registry resource that has an incorrect or missing resource policy attached.

2. **Create an Appropriate Resource Policy**: You can create a new resource policy that allows the necessary permissions for the custom schema registry resource. Here is an example of a minimal resource policy that allows full access to the custom schema registry:
   ```json
   {
       "Version": "2012-10-17",
       "Statement": [
           {
               "Effect": "Allow",
               "Principal": {
                   "Service": "firehose.amazonaws.com"
               },
               "Action": "registry:*",
               "Resource": "arn:aws:schemas:us-east-1:123456789012:registry/MyCustomSchemaRegistry"
           }
       ]
   }
   ```

3. **Attach the Resource Policy**: Use the AWS SDK for Python (Boto3) to attach the newly created resource policy to the custom schema registry resource. Here is an example Python code snippet to attach the resource policy:
   ```python
   import boto3

   client = boto3.client('schemas')

   response = client.put_resource_policy(
       RegistryName='MyCustomSchemaRegistry',
       Policy='{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"firehose.amazonaws.com"},"Action":"registry:*","Resource":"arn:aws:schemas:us-east-1:123456789012:registry/MyCustomSchemaRegistry"}]'
   )

   print(response)
   ```

4. **Verify the Resource Policy**: After attaching the resource policy, verify that the custom schema registry resource now has the correct resource policy attached by listing the resource policies for the registry.

By following these steps and customizing the resource policy and Python code as per your specific requirements, you can successfully remediate the misconfiguration of a resource policy attachment in a custom schema registry for AWS CloudWatch using Python.
</Accordion>

</AccordionGroup>
</Tab>
</Tabs>

