---
slug: wafv2_rulegroup_not_empty_rule
title: WAF V2 Rule Groups Should Not Be Empty
sidebar_label: WAF V2 Rule Groups Should Not Be Empty
---

### More Info:

WAF rule groups should not be empty

### Risk Level

High

### Address

Security

### Compliance Standards

CBP



### Triage and Remediation
<Tabs>


<Tab title='Prevention'>
### How to Prevent
<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
To prevent WAF V2 Rule Groups from being empty in AWS using the AWS Management Console, follow these steps:

1. **Navigate to the WAF & Shield Console:**
   - Open the AWS Management Console.
   - In the search bar, type "WAF" and select "WAF & Shield" from the dropdown.

2. **Select the Appropriate Web ACL:**
   - In the WAF & Shield dashboard, click on "Web ACLs" in the left-hand menu.
   - Select the Web ACL that you want to configure.

3. **Add Rule Groups:**
   - In the Web ACL details page, scroll down to the "Rules" section.
   - Click on "Add rules" and then select "Rule group" from the options.
   - Choose an existing rule group or create a new one by clicking "Create rule group."

4. **Configure and Save:**
   - If creating a new rule group, define the rules and conditions as per your security requirements.
   - Once the rule group is added, ensure it contains at least one rule.
   - Click "Save" to apply the changes to the Web ACL.

By following these steps, you can ensure that your WAF V2 Rule Groups are not empty, thereby maintaining the effectiveness of your web application firewall.
</Accordion>

<Accordion title='Using CLI'>
To prevent WAF V2 Rule Groups from being empty in AWS using the AWS CLI, you can follow these steps:

1. **Create a Web ACL:**
   First, create a Web ACL if you don't already have one. This is necessary to associate rule groups with it.
   ```sh
   aws wafv2 create-web-acl --name my-web-acl --scope REGIONAL --default-action Block={} --visibility-config SampledRequestsEnabled=true,CloudWatchMetricsEnabled=true,MetricName=myWebACLMetric --region us-east-1
   ```

2. **Create a Rule Group:**
   Create a rule group that will contain the rules you want to apply. Ensure that the rule group is not empty by specifying at least one rule.
   ```sh
   aws wafv2 create-rule-group --name my-rule-group --scope REGIONAL --capacity 10 --visibility-config SampledRequestsEnabled=true,CloudWatchMetricsEnabled=true,MetricName=myRuleGroupMetric --rules '[
     {
       "Name": "SampleRule",
       "Priority": 1,
       "Statement": {
         "ByteMatchStatement": {
           "SearchString": "bad-bot",
           "FieldToMatch": {
             "UriPath": {}
           },
           "TextTransformations": [
             {
               "Priority": 0,
               "Type": "NONE"
             }
           ],
           "PositionalConstraint": "CONTAINS"
         }
       },
       "Action": {
         "Block": {}
       },
       "VisibilityConfig": {
         "SampledRequestsEnabled": true,
         "CloudWatchMetricsEnabled": true,
         "MetricName": "SampleRuleMetric"
       }
     }
   ]' --region us-east-1
   ```

3. **Associate Rule Group with Web ACL:**
   Associate the created rule group with the Web ACL to ensure it is applied.
   ```sh
   aws wafv2 update-web-acl --name my-web-acl --scope REGIONAL --default-action Block={} --visibility-config SampledRequestsEnabled=true,CloudWatchMetricsEnabled=true,MetricName=myWebACLMetric --rules '[
     {
       "Name": "my-rule-group",
       "Priority": 1,
       "Statement": {
         "RuleGroupReferenceStatement": {
           "ARN": "arn:aws:wafv2:us-east-1:123456789012:regional/rulegroup/my-rule-group/abcd1234-efgh-5678-ijkl-9012mnopqrst"
         }
       },
       "OverrideAction": {
         "None": {}
       },
       "VisibilityConfig": {
         "SampledRequestsEnabled": true,
         "CloudWatchMetricsEnabled": true,
         "MetricName": "myRuleGroupMetric"
       }
     }
   ]' --region us-east-1
   ```

4. **Verify Configuration:**
   Verify that the rule group is associated with the Web ACL and is not empty.
   ```sh
   aws wafv2 get-web-acl --name my-web-acl --scope REGIONAL --region us-east-1
   ```

By following these steps, you ensure that your WAF V2 Rule Groups are not empty and are properly configured to protect your resources.
</Accordion>

<Accordion title='Using Python'>
To prevent WAF V2 Rule Groups from being empty in AWS using Python scripts, you can use the Boto3 library, which is the AWS SDK for Python. Here are the steps to ensure that WAF V2 Rule Groups are not empty:

### Step 1: Install Boto3
First, ensure that you have Boto3 installed. You can install it using pip if you haven't already:
```bash
pip install boto3
```

### Step 2: Initialize Boto3 Client
Initialize the Boto3 client for WAFV2. This will allow you to interact with the AWS WAFV2 service.
```python
import boto3

client = boto3.client('wafv2', region_name='us-east-1')  # Specify your region
```

### Step 3: Create a Rule Group with Rules
Create a WAF V2 Rule Group and ensure it contains at least one rule. Here is an example of how to create a rule group with a sample rule:
```python
response = client.create_rule_group(
    Name='MyRuleGroup',
    Scope='REGIONAL',  # or 'CLOUDFRONT' for CloudFront distributions
    Capacity=100,
    Rules=[
        {
            'Name': 'SampleRule',
            'Priority': 1,
            'Statement': {
                'ByteMatchStatement': {
                    'SearchString': b'example',
                    'FieldToMatch': {
                        'UriPath': {}
                    },
                    'TextTransformations': [
                        {
                            'Priority': 0,
                            'Type': 'NONE'
                        }
                    ],
                    'PositionalConstraint': 'CONTAINS'
                }
            },
            'Action': {
                'Block': {}
            },
            'VisibilityConfig': {
                'SampledRequestsEnabled': True,
                'CloudWatchMetricsEnabled': True,
                'MetricName': 'SampleRuleMetric'
            }
        }
    ],
    VisibilityConfig={
        'SampledRequestsEnabled': True,
        'CloudWatchMetricsEnabled': True,
        'MetricName': 'MyRuleGroupMetric'
    }
)
print(response)
```

### Step 4: Validate Rule Group
After creating the rule group, you can validate that it contains rules by describing the rule group:
```python
rule_group_arn = response['Summary']['ARN']

rule_group = client.get_rule_group(
    Name='MyRuleGroup',
    Scope='REGIONAL',  # or 'CLOUDFRONT'
    Id=rule_group_arn.split('/')[-1]
)

if not rule_group['RuleGroup']['Rules']:
    print("Rule Group is empty!")
else:
    print("Rule Group contains rules.")
```

By following these steps, you can ensure that your WAF V2 Rule Groups are not empty, thus preventing misconfigurations.
</Accordion>

</AccordionGroup>
</Tab>

<Tab title='Cause'>
### Check Cause
<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
1. Log in to the AWS Management Console and navigate to the AWS WAF & Shield service.

2. In the navigation pane, choose "WebACLs". This will display a list of all your WebACLs.

3. Click on the name of the WebACL that you want to check. This will open the details page for that WebACL.

4. In the "Rules" section, check if there are any rule groups that are empty. If a rule group is empty, it means that it does not contain any rules. This is a misconfiguration because an empty rule group does not provide any protection.
</Accordion>

<Accordion title='Using CLI'>
1. Install and configure AWS CLI: Before you can start using AWS CLI, you need to install it on your local machine and configure it with your AWS account credentials. You can do this by running the following commands:

   Installation:
   ```
   pip install awscli
   ```
   Configuration:
   ```
   aws configure
   ```
   You will be prompted to enter your AWS Access Key ID, Secret Access Key, Default region name, and Default output format.

2. List all WAFv2 rule groups: Use the `list-rule-groups` command to get a list of all the rule groups in your AWS WAFv2. Replace `--scope REGIONAL` with `--scope CLOUDFRONT` if you want to list rule groups for CloudFront.

   ```
   aws wafv2 list-rule-groups --scope REGIONAL
   ```
   This command will return a list of rule group IDs and ARNs.

3. Get and inspect each rule group: For each rule group ID returned by the previous command, use the `get-rule-group` command to get the details of the rule group. Replace `<rule-group-id>` and `<rule-group-arn>` with the actual ID and ARN of the rule group.

   ```
   aws wafv2 get-rule-group --scope REGIONAL --id <rule-group-id> --arn <rule-group-arn>
   ```
   This command will return the details of the rule group, including its rules.

4. Check if the rule group is empty: Inspect the output of the previous command. If the "Rules" array is empty, then the rule group is empty. If it's not empty, then the rule group contains one or more rules.
</Accordion>

<Accordion title='Using Python'>
To check if WAF V2 Rule Groups are not empty in AWS CloudWatch, you can use the AWS SDK for Python (Boto3). Here are the steps:

1. **Setup AWS SDK for Python (Boto3):**
   First, you need to install and configure Boto3. You can install it using pip:

   ```
   pip install boto3
   ```

   Then, configure your AWS credentials. You can do this by setting the following environment variables:

   ```
   AWS_ACCESS_KEY_ID = 'your_access_key'
   AWS_SECRET_ACCESS_KEY = 'your_secret_key'
   ```

2. **Create a Python Script to List WAFv2 Rule Groups:**
   Use the `list_rule_groups` method of the WAFv2 client to get a list of all rule groups. Here's a sample script:

   ```python
   import boto3

   client = boto3.client('wafv2')

   response = client.list_rule_groups(
       Scope='REGIONAL'  # or 'CLOUDFRONT' depending on where your rule groups are
   )

   rule_groups = response['RuleGroups']
   ```

3. **Check if Rule Groups are Empty:**
   For each rule group, use the `get_rule_group` method to get the rules in the group. If the 'Rules' list is empty, then the rule group is empty.

   ```python
   for rule_group in rule_groups:
       rule_group_response = client.get_rule_group(
           Name=rule_group['Name'],
           Id=rule_group['Id'],
           Scope='REGIONAL'  # or 'CLOUDFRONT'
       )

       if not rule_group_response['RuleGroup']['Rules']:
           print(f"Rule group {rule_group['Name']} is empty.")
   ```

4. **Handle Pagination:**
   The `list_rule_groups` method returns a maximum of 20 rule groups at a time. If you have more rule groups, you need to handle pagination by using the 'NextMarker' field in the response.

   ```python
   while 'NextMarker' in response:
       response = client.list_rule_groups(
           Scope='REGIONAL',  # or 'CLOUDFRONT'
           NextMarker=response['NextMarker']
       )

       rule_groups += response['RuleGroups']
   ```

This script will print the names of all rule groups that are empty. You can modify it to suit your needs, for example by sending an alert or logging the result to a file.
</Accordion>

</AccordionGroup>
</Tab>
<Tab title='Remediation'>
### Remediation

<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
To remediate the misconfiguration of having empty WAF V2 Rule Groups in AWS CloudWatch using the AWS console, follow these steps:

1. **Login to AWS Console**: Go to the AWS Management Console (https://aws.amazon.com/) and login using your credentials.

2. **Navigate to AWS WAF**: Click on the "Services" dropdown menu at the top of the page, search for "WAF & Shield" under the Security, Identity, & Compliance section, and click on it.

3. **Select the WAF Web ACL**: In the AWS WAF & Shield dashboard, click on "Web ACLs" from the left-hand menu.

4. **Choose the Web ACL**: Select the Web ACL that you want to check for empty WAF V2 Rule Groups.

5. **Review Rule Groups**: In the selected Web ACL, review the configured Rule Groups to identify any that are empty.

6. **Edit the Web ACL**: Click on the Web ACL that contains the empty WAF V2 Rule Groups to edit it.

7. **Remove Empty Rule Groups**: In the Web ACL editor, locate the empty Rule Groups and remove them by clicking on the delete or remove option next to each empty Rule Group.

8. **Add Rules to Rule Groups**: If necessary, add appropriate rules to the Rule Groups to ensure that they are not empty. You can create custom rules or use managed rule groups provided by AWS.

9. **Save Changes**: After removing the empty Rule Groups and adding necessary rules, save the changes to the Web ACL.

10. **Review and Deploy**: Review the updated Web ACL configuration to ensure that there are no more empty Rule Groups. Once you are satisfied with the changes, deploy the updated Web ACL to apply the changes.

By following these steps, you can remediate the misconfiguration of having empty WAF V2 Rule Groups in AWS CloudWatch using the AWS console.

</Accordion>

<Accordion title='Using CLI'>
To remediate the misconfiguration of having empty WAF V2 Rule Groups in AWS CloudWatch using AWS CLI, you can follow these steps:

1. **Identify the Empty WAF V2 Rule Groups**: First, you need to identify the empty WAF V2 Rule Groups in your AWS account. You can do this by running the following AWS CLI command:
   
   ```bash
   aws wafv2 list-rule-groups --scope CLOUDFRONT(or REGIONAL) --query "RuleGroups[?RuleGroupArn == null].Name"
   ```
   
   This command will list the names of all the empty WAF V2 Rule Groups in your account.

2. **Update the Empty Rule Groups**: To update the empty WAF V2 Rule Groups, you can either add rules to them or delete them based on your requirements. 

   - To add rules to a specific empty WAF V2 Rule Group, you can use the following AWS CLI command:
   
     ```bash
     aws wafv2 update-rule-group --scope CLOUDFRONT(or REGIONAL) --name <RuleGroupName> --rules-action ALLOW|BLOCK --rules file://rules.json
     ```
     
     Replace `<RuleGroupName>` with the name of the empty WAF V2 Rule Group and provide the necessary rules in the `rules.json` file.

   - To delete a specific empty WAF V2 Rule Group, you can use the following AWS CLI command:
   
     ```bash
     aws wafv2 delete-rule-group --scope CLOUDFRONT(or REGIONAL) --name <RuleGroupName>
     ```
     
     Replace `<RuleGroupName>` with the name of the empty WAF V2 Rule Group that you want to delete.

3. **Verify the Changes**: After updating or deleting the empty WAF V2 Rule Groups, you can verify the changes by listing all the WAF V2 Rule Groups in your account using the following AWS CLI command:
   
   ```bash
   aws wafv2 list-rule-groups --scope CLOUDFRONT(or REGIONAL) --query "RuleGroups[].Name"
   ```

   This command will list all the WAF V2 Rule Groups in your account.

By following these steps, you can remediate the misconfiguration of having empty WAF V2 Rule Groups in AWS CloudWatch using AWS CLI.

</Accordion>

<Accordion title='Using Python'>
To remediate the misconfiguration of having empty WAF V2 Rule Groups in AWS CloudWatch using Python, you can follow these steps:

1. Install the Boto3 library:
```bash
pip install boto3
```

2. Use the following Python script to check for and remediate empty WAF V2 Rule Groups in AWS CloudWatch:

```python
import boto3

def remediate_empty_rule_groups():
    # Initialize the AWS CloudWatch client
    client = boto3.client('wafv2')

    # Get all the available rule groups
    response = client.list_rule_groups(Scope='CLOUDFRONT(or REGIONAL)')

    for rule_group in response['RuleGroups']:
        rule_group_id = rule_group['Id']
        
        # Get the rules in each rule group
        rules_response = client.get_rule_group(RuleGroupId=rule_group_id)
        
        # Check if the rule group is empty
        if not rules_response['RuleGroup']['AvailableRules']:
            # Remediate by adding a sample rule
            sample_rule = {
                'Name': 'SampleRule',
                'Priority': 1,
                'Statement': {
                    'ByteMatchStatement': {
                        'SearchString': 'sample',
                        'FieldToMatch': {
                            'UriPath': {}
                        }
                    }
                },
                'Action': {
                    'Block': {}
                },
                'VisibilityConfig': {
                    'SampledRequestsEnabled': True,
                    'CloudWatchMetricsEnabled': True,
                    'MetricName': 'SampleRuleMetric'
                }
            }
            
            # Add the sample rule to the rule group
            client.put_managed_rule_group(
                RuleGroupName=rule_group_id,
                Updates=[
                    {
                        'Action': 'INSERT',
                        'ManagedRuleGroupStatement': {
                            'VendorName': 'AWS',
                            'Name': 'AWSManagedRulesCommonRuleSet',
                            'ExcludedRules': []
                        }
                    },
                    {
                        'Action': 'INSERT',
                        'RuleGroupReferenceStatement': {
                            'ARN': 'arn:aws:wafv2:us-west-2:123456789012:regional/managedrulegroup/AWSManagedRulesCommonRuleSet',
                            'ExcludedRules': []
                        }
                    },
                    {
                        'Action': 'INSERT',
                        'RuleStatement': {
                            'ManagedRuleGroupStatement': {
                                'VendorName': 'AWS',
                                'Name': 'AWSManagedRulesKnownBadInputsRuleSet',
                                'ExcludedRules': []
                            }
                        }
                    },
                    {
                        'Action': 'INSERT',
                        'RuleStatement': {
                            'ManagedRuleGroupStatement': {
                                'VendorName': 'AWS',
                                'Name': 'AWSManagedRulesPHPRuleSet',
                                'ExcludedRules': []
                            }
                        }
                    },
                    {
                        'Action': 'INSERT',
                        'RuleStatement': {
                            'ManagedRuleGroupStatement': {
                                'VendorName': 'AWS',
                                'Name': 'AWSManagedRulesSQLiRuleSet',
                                'ExcludedRules': []
                            }
                        }
                    },
                    {
                        'Action': 'INSERT',
                        'RuleStatement': {
                            'ManagedRuleGroupStatement': {
                                'VendorName': 'AWS',
                                'Name': 'AWSManagedRulesLinuxRuleSet',
                                'ExcludedRules': []
                            }
                        }
                    },
                    {
                        'Action': 'INSERT',
                        'RuleStatement': {
                            'ManagedRuleGroupStatement': {
                                'VendorName': 'AWS',
                                'Name': 'AWSManagedRulesAdminProtectionRuleSet',
                                'ExcludedRules': []
                            }
                        }
                    },
                    {
                        'Action': 'INSERT',
                        'RuleStatement': {
                            'ManagedRuleGroupStatement': {
                                'VendorName': 'AWS',
                                'Name': 'AWSManagedRulesBotControlRuleSet',
                                'ExcludedRules': []
                            }
                        }
                    },
                    {
                        'Action': 'INSERT',
                        'RuleStatement': {
                            'RuleGroupReferenceStatement': {
                                'ARN': rule_group_id,
                                'ExcludedRules': []
                            }
                        }
                    }
                ]
            )
            print(f"Remediated empty rule group: {rule_group_id} by adding a sample rule")
        else:
            print(f"Rule group: {rule_group_id} is not empty")

if __name__ == '__main__':
    remediate_empty_rule_groups()
```

3. Run this Python script to check for empty WAF V2 Rule Groups in AWS CloudWatch and remediate them by adding a sample rule to each empty rule group.

This script will add a sample rule to each empty WAF V2 Rule Group in AWS CloudWatch. You can customize the sample rule based on your requirements.
</Accordion>

</AccordionGroup>
</Tab>
</Tabs>
