---
slug: sql_trace_flag
title: SQL Server Trace Flag Should Be Off
sidebar_label: SQL Server Trace Flag Should Be Off
---

### More Info:

Microsoft SQL Trace Flags are frequently used to diagnose performance issues or to debug stored procedures or complex computer systems, but they may also be recommended by Microsoft Support to address behavior that is negatively impacting a specific workload. All documented trace flags and those recommended by Microsoft Support are fully supported in a production environment when used as directed. 3625(trace log) Limits the amount of information returned to users who are not members of the sysadmin fixed server role, by masking the parameters of some error messages using '******'. Setting this in a Google Cloud flag for the instance allows for security through obscurity and prevents the disclosure of sensitive information, hence this is recommended to set this flag globally to off to prevent the flag having been left on, or turned on by bad actors. This recommendation is applicable to SQL Server database instances.

### Risk Level

Low

### Address

Reliability, Security

### Compliance Standards

CISGCP, CBP



<Tabs><Tab title='Cause'>
### Check Cause

#### Using Console

1. Log in to the Google Cloud Console: Open your web browser, navigate to the Google Cloud Console (console.cloud.google.com), and sign in with your Google account.

2. Navigate to SQL Instances: From the left-hand side menu, select "SQL" under the "Storage" section. This will display a list of all SQL instances in your project.

3. Select the SQL Server Instance: Click on the name of the SQL Server instance you want to check. This will open the instance details page.

4. Check the Trace Flag: In the instance details page, navigate to the "Flags" section. If the Trace Flag is set to "ON", it means it's misconfigured. The correct configuration should be "OFF". If the flag is not listed, it means it's not set and is in the default state, which is "OFF".

#### Using CLI

1. First, you need to install and configure the Google Cloud SDK (gcloud CLI) on your local machine. You can download it from the official Google Cloud website and follow the instructions to install it. Once installed, authenticate your Google Cloud account by running the following command in your terminal:

   ```
   gcloud auth login
   ```

2. After successful authentication, you can list all the SQL instances in your GCP project by running the following command:

   ```
   gcloud sql instances list
   ```

3. Choose the SQL instance that you want to check and get the details of that instance by running the following command:

   ```
   gcloud sql instances describe [INSTANCE_NAME]
   ```

   Replace `[INSTANCE_NAME]` with the name of your SQL instance.

4. In the output of the above command, look for the `databaseFlags` field. If the `trace flag` is set to `ON`, then it is a misconfiguration. The `trace flag` should be `OFF` for security reasons. If the `trace flag` is not listed in the `databaseFlags` field, then it is not set and is considered `OFF` by default.

#### Using Python

To check if the SQL Server Trace Flag is off in SQL using Python scripts, you can follow these steps:

1. **Install Required Libraries**: First, you need to install the required libraries. You can use the `pyodbc` library to connect Python with SQL Server. You can install it using pip:
```python
pip install pyodbc
```

2. **Establish Connection**: After installing the library, you need to establish a connection with the SQL Server. Here is a sample script to connect to SQL Server:
```python
import pyodbc 
server = 'your_server' 
database = 'your_database' 
username = 'your_username' 
password = 'your_password' 
cnxn = pyodbc.connect('DRIVER={ODBC Driver 17 for SQL Server};SERVER='+server+';DATABASE='+database+';UID='+username+';PWD='+ password)
cursor = cnxn.cursor()
```
Replace 'your_server', 'your_database', 'your_username', and 'your_password' with your actual SQL Server details.

3. **Execute Query**: Now, you can execute the SQL query to check the status of the Trace Flag. The query to check the status of a specific Trace Flag (for example, Trace Flag 1222) is `DBCC TRACESTATUS (1222)`. Here is how you can execute it in Python:
```python
cursor.execute("DBCC TRACESTATUS (1222)")
rows = cursor.fetchall()
for row in rows:
    print(row)
```
This will print the status of Trace Flag 1222. If the status is 1, it means the Trace Flag is on. If the status is 0, it means the Trace Flag is off.

4. **Check Status**: Finally, you can add a condition in your Python script to check if the Trace Flag is off:
```python
if rows[0][1] == 0:
    print("Trace Flag is off")
else:
    print("Trace Flag is on")
```
This will print "Trace Flag is off" if the Trace Flag is off, and "Trace Flag is on" if the Trace Flag is on.

</Tab>


<Tab title='Remediation'>
### Remediation

#### Using Console

To remediate the SQL Server Trace Flag misconfiguration on GCP using the GCP Console, follow these steps:

1. Log in to the GCP Console and navigate to the Cloud SQL instances page.
2. Select the instance that you want to remediate.
3. Click on the "Edit" button at the top of the page.
4. Scroll down to the "Flags" section.
5. Look for the trace flag that needs to be turned off and click on the "X" icon to remove it.
6. Click on the "Save" button at the bottom of the page to apply the changes.

Once the changes are saved, the SQL Server Trace Flag will be turned off, and the misconfiguration will be remediated.

#### Using CLI

To remediate the SQL Server Trace Flag misconfiguration on GCP using GCP CLI, follow these steps:

1. Open the Cloud Shell in the GCP Console.

2. Run the following command to authenticate and set the default project:

```
gcloud auth login
gcloud config set project [PROJECT_ID]
```

3. Run the following command to list all the instances in your project:

```
gcloud compute instances list
```

4. Identify the instance that has the SQL Server Trace Flag misconfiguration.

5. SSH into the instance by running the following command:

```
gcloud compute ssh [INSTANCE_NAME]
```

6. Once you are logged in to the instance, connect to the SQL Server instance and run the following command to turn off the Trace Flag:

```
DBCC TRACEOFF (trace#)
```

Note: Replace "trace#" with the number of the trace flag that needs to be turned off.

7. Verify that the Trace Flag has been turned off by running the following command:

```
DBCC TRACESTATUS
```

8. Exit the SQL Server instance and the SSH session by typing "exit" in the terminal.

9. Verify that the misconfiguration has been remediated by running a security scan or by checking the compliance status of the instance.

Congratulations! You have successfully remediated the SQL Server Trace Flag misconfiguration on GCP using GCP CLI.

#### Using Python

To remediate the SQL Server Trace Flag misconfiguration in GCP using Python, you can follow the below steps:

Step 1: Connect to the SQL Server instance using the pyodbc library.

```
import pyodbc
server = '<your_server_name>.database.windows.net'
database = '<your_database_name>'
username = '<your_username>'
password = '<your_password>'
driver= '{ODBC Driver 17 for SQL Server}'
cnxn = pyodbc.connect('DRIVER='+driver+';SERVER='+server+';PORT=1433;DATABASE='+database+';UID='+username+';PWD='+ password)
```

Step 2: Check the current status of the Trace Flag using the below query:

```
cursor = cnxn.cursor()
cursor.execute("DBCC TRACESTATUS(-1)")
row = cursor.fetchone()
print("Trace Flag Status:", row[1])
```

Step 3: If the Trace Flag status is '1', then it is enabled. To disable it, execute the below query:

```
cursor.execute("DBCC TRACEOFF(1222,-1)")
cnxn.commit()
```

Step 4: Finally, verify that the Trace Flag is disabled by running the query in Step 2 again.

```
cursor.execute("DBCC TRACESTATUS(-1)")
row = cursor.fetchone()
print("Trace Flag Status:", row[1])
```

This should remediate the SQL Server Trace Flag misconfiguration in GCP using Python.


### Additional Reading:

- [https://cloud.google.com/sql/docs/sqlserver/flags](https://cloud.google.com/sql/docs/sqlserver/flags) 


</Tab>
</Tabs>