--- 
slug: gcp_rt_bigquery_dataset_changes
eventname: google.cloud.bigquery.v2.DatasetService.UpdateDataset
title: UpdateDataset
sidebar_label: google.cloud.bigquery.v2.DatasetService.UpdateDataset
---
                       
### Event Information

#### Meaning

- The google.cloud.bigquery.v2.DatasetService.UpdateDataset event in gcprealtime for BigQuery indicates that a dataset in BigQuery has been updated.
- This event is triggered when there are changes made to the metadata or configuration of a dataset, such as modifying the dataset's description, default table expiration time, or access controls.
- It provides real-time information about dataset modifications, allowing users to track and monitor changes made to their BigQuery datasets.

#### Example

1. Example of security impact: If an unauthorized user gains access to the `google.cloud.bigquery.v2.DatasetService.UpdateDataset` event in GCP Realtime for BigQuery, they could potentially modify the dataset configuration, access controls, or even delete the dataset. This could lead to unauthorized access to sensitive data, data loss, or disruption of data analytics processes.

Remediation:
- Enable audit logging for BigQuery datasets to track any changes made to the dataset configuration. This can help in identifying unauthorized modifications.
- Implement strict access controls and permissions for datasets, ensuring that only authorized users have the necessary privileges to modify dataset configurations.
- Regularly review and monitor dataset access logs and permissions to detect any suspicious activities or unauthorized changes.

Example CLI command to enable audit logging for a BigQuery dataset:
```
gcloud logging sinks create dataset-audit-log bigquery.googleapis.com/projects/[PROJECT_ID]/datasets/[DATASET_ID] --log-filter='protoPayload.methodName="google.cloud.bigquery.v2.DatasetService.UpdateDataset"'
```

2. Example of security impact: If the `google.cloud.bigquery.v2.DatasetService.UpdateDataset` event is triggered with incorrect or malicious parameters, it could lead to unintended changes in the dataset configuration. For example, an attacker could modify the dataset to grant excessive permissions to unauthorized users, potentially exposing sensitive data.

Remediation:
- Implement strict input validation and parameter sanitization in any applications or scripts that interact with the BigQuery API to prevent unauthorized or malicious modifications.
- Regularly review and monitor dataset configurations to ensure they align with the intended access controls and permissions.
- Implement automated checks or alerts to detect any unexpected or suspicious changes to dataset configurations.

Example CLI command to update dataset access controls using the `bq` command-line tool:
```
bq update --access=PROJECT:DATASET:ROLE [PROJECT_ID]:[DATASET_ID]
```

3. Example of security impact: If the `google.cloud.bigquery.v2.DatasetService.UpdateDataset` event is triggered by a compromised or unauthorized service account, it could result in unauthorized modifications to the dataset configuration. This could lead to data breaches, data loss, or disruption of data analytics processes.

Remediation:
- Regularly review and monitor the permissions assigned to service accounts and ensure they are limited to the necessary privileges.
- Implement strong access controls and multi-factor authentication for service accounts to prevent unauthorized access.
- Regularly rotate the credentials of service accounts to minimize the impact of compromised credentials.

Example CLI command to rotate the credentials of a service account:
```
gcloud iam service-accounts keys create --iam-account=[SERVICE_ACCOUNT_EMAIL] [OUTPUT_FILE]
```


 