

<Tabs><Tab title='Cause'>
### Check Cause

#### Using Console

1. Login to the Google Cloud Platform Console: Open your Google Cloud Platform console and sign in with your GCP account credentials.

2. Navigate to Dataproc Clusters: From the left-hand side menu, navigate to "Dataproc" under the "Big Data" section. Then click on "Clusters" to view all the existing Dataproc clusters.

3. Check Cluster Details: Click on the name of the cluster you want to inspect. This will open the cluster details page. 

4. Verify Encryption Settings: In the cluster details page, look for the "Security" section. Under this section, check if "Customer-managed key (CMEK)" is selected under "Encryption keys". If it is not, then the Dataproc cluster is not encrypted using CMEK, indicating a misconfiguration.

#### Using CLI

1. First, you need to install and configure the Google Cloud SDK (gcloud CLI) on your local machine. You can download it from the official Google Cloud website and follow the instructions to install and authenticate it.

2. Once the SDK is installed and configured, you can list all the Dataproc clusters in your project by running the following command in your terminal:

   ```
   gcloud dataproc clusters list --region=<region>
   ```

   Replace `<region>` with the region where your clusters are located. This command will return a list of all the clusters in the specified region.

3. For each cluster, you can check the encryption settings by describing the cluster with the following command:

   ```
   gcloud dataproc clusters describe <cluster-name> --region=<region>
   ```

   Replace `<cluster-name>` with the name of the cluster you want to check. This command will return a detailed description of the cluster, including its encryption settings.

4. In the output of the previous command, look for the `gceClusterConfig` field. If the `kmsKeyName` field is present and set to a valid KMS key, then the cluster is encrypted using CMEK. If the `kmsKeyName` field is not present or is set to an invalid key, then the cluster is not encrypted using CMEK.

#### Using Python

To check if Dataproc Clusters are encrypted using Customer-Managed Encryption Keys (CMEK) in Google Cloud Platform (GCP) Identity and Access Management (IAM), you can use the Google Cloud SDK (gcloud) or the Google Cloud Client Libraries for Python. Here are the steps:

1. **Setup Google Cloud SDK and Authentication:**
   Before you can interact with GCP resources, you need to authenticate your SDK. You can do this by running the command `gcloud auth login` and following the prompts. This will authenticate your SDK for all GCP services.

2. **Install Google Cloud Client Libraries for Python:**
   You can install the Google Cloud Client Libraries for Python by running the command `pip install google-cloud-dataproc`. This will allow you to interact with the Dataproc service using Python.

3. **List all Dataproc Clusters:**
   You can list all Dataproc Clusters in your project by using the `clusters().list()` method from the Dataproc Client Library. Here is a sample Python script:

   ```python
   from google.cloud import dataproc_v1

   def list_clusters(project_id, region):
       client = dataproc_v1.ClusterControllerClient(client_options={"api_endpoint": f"{region}-dataproc.googleapis.com:443"})
       response = client.list_clusters(request={"project_id": project_id, "region": region})
       for cluster in response:
           print(cluster)
   ```

4. **Check for CMEK Encryption:**
   For each cluster, you can check if it is encrypted with CMEK by checking the `encryption_config` field. If the `encryption_config` field is not present or the `kms_key_name` field is not set, then the cluster is not encrypted with CMEK. Here is a sample Python script:

   ```python
   from google.cloud import dataproc_v1

   def check_cmek_encryption(project_id, region):
       client = dataproc_v1.ClusterControllerClient(client_options={"api_endpoint": f"{region}-dataproc.googleapis.com:443"})
       response = client.list_clusters(request={"project_id": project_id, "region": region})
       for cluster in response:
           encryption_config = cluster.config.encryption_config
           if encryption_config is None or not encryption_config.kms_key_name:
               print(f"Cluster {cluster.cluster_name} is not encrypted with CMEK")
   ```

Remember to replace `project_id` and `region` with your actual GCP project ID and region.

</Tab>

<Tab title='Remediation'>
### Remediation

#### Using Console

To remediate the misconfiguration "Ensure Dataproc Clusters Encrypted Using CMEK" for GCP using GCP console, you can follow the below steps:

1. Login to your GCP console.
2. Navigate to the Dataproc Clusters page.
3. Select the cluster for which you want to enable encryption.
4. Click on the "Edit" button at the top of the page.
5. Scroll down to the "Security" section.
6. Under "Encryption", select "Customer-managed key".
7. Choose the Cloud KMS key that you want to use for encryption.
8. Click on the "Save" button at the bottom of the page.

Once you have completed these steps, your Dataproc cluster will be encrypted using the specified Cloud KMS key. This will ensure that your data is secure and protected from unauthorized access.

#### Using CLI

To remediate the misconfiguration "Ensure Dataproc Clusters Encrypted Using CMEK" in GCP using GCP CLI, you can follow the following steps:

1. Open the Cloud Shell from the GCP console.
2. Run the command `gcloud dataproc clusters list --region=REGION` to list all the Dataproc clusters in the specified region.
3. For each cluster in the list, run the command `gcloud dataproc clusters describe CLUSTER_NAME --region=REGION` to get the details of the cluster.
4. Check if the `encryptionConfig` property is set to `CMEK` in the output of the above command. If it is not set to `CMEK`, then the cluster is not encrypted using CMEK.
5. To encrypt the cluster using CMEK, run the command `gcloud dataproc clusters update CLUSTER_NAME --region=REGION --update-encryption-config kmsKeyName=KEY_NAME` where `KEY_NAME` is the name of the KMS key to be used for encryption.
6. After the cluster is updated, run the command `gcloud dataproc clusters describe CLUSTER_NAME --region=REGION` again to verify that the `encryptionConfig` property is set to `CMEK`.

By following the above steps, you can remediate the misconfiguration "Ensure Dataproc Clusters Encrypted Using CMEK" for GCP using GCP CLI.

#### Using Python

To remediate the misconfiguration "Ensure Dataproc Clusters Encrypted Using CMEK" on GCP using Python, you can follow the below steps:

1. First, you need to create a Cloud KMS key ring and a key in the same project where your Dataproc cluster is running. You can use the following code to create a key ring and a key:

```python
from google.cloud import kms_v1

# Replace with your project ID
project_id = 'your-project-id'

# Create the client
client = kms_v1.KeyManagementServiceClient()

# Create the key ring
key_ring_id = 'dataproc-key-ring'
location_id = 'us-central1'
key_ring_parent = client.location_path(project_id, location_id)
key_ring = client.create_key_ring(key_ring_parent, key_ring_id)

# Create the key
key_id = 'dataproc-key'
purpose = kms_v1.enums.CryptoKey.CryptoKeyPurpose.ENCRYPT_DECRYPT
key = {'purpose': purpose}
key_parent = client.key_ring_path(project_id, location_id, key_ring_id)
key = client.create_crypto_key(key_parent, key_id, key)
```

2. Once you have created the key, you need to update your Dataproc cluster to use the newly created key for encryption. You can use the following code to update a cluster:

```python
from google.cloud import dataproc_v1 as dataproc

# Replace with your project ID and cluster name
project_id = 'your-project-id'
region = 'us-central1'
cluster_name = 'your-cluster-name'

# Create the client
client = dataproc.ClusterControllerClient()

# Get the cluster
cluster = client.get_cluster(project_id, region, cluster_name)

# Update the cluster config to use the new key
config = cluster.config
config.encryption_config.gce_pd_kms_key_name = 'projects/{}/locations/{}/keyRings/{}/cryptoKeys/{}'.format(project_id, location_id, key_ring_id, key_id)
cluster.config = config

# Update the cluster
update_mask = dataproc.types.FieldMask(paths=['config.encryption_config.gce_pd_kms_key_name'])
operation = client.update_cluster(project_id, region, cluster_name, cluster, update_mask)
```

3. Finally, you need to verify that the cluster is using the correct key for encryption. You can use the following code to get the encryption configuration of a cluster:

```python
# Get the cluster
cluster = client.get_cluster(project_id, region, cluster_name)

# Print the encryption config
print(cluster.config.encryption_config)
```

If the output shows that the `gce_pd_kms_key_name` field is set to the correct key, then the cluster is encrypted using CMEK.


</Tab>
</Tabs>