

<Tabs><Tab title='Cause'>
### Check Cause

#### Using Console

1. Log in to the Azure portal and navigate to the "Storage Accounts" section.
2. Select the storage account you want to inspect.
3. In the storage account window, select "Access keys" under the "Settings" section.
4. Here, you can see the age of your access keys. If the keys have not been rotated in a long time, it's a misconfiguration. Azure doesn't provide the exact age of the keys, but you can manually track when you last rotated them.

#### Using CLI

1. Install Azure CLI: Before you can run any Azure CLI commands, you need to install the Azure CLI on your local machine. You can download it from the official Microsoft Azure website.

2. Login to Azure: After installing the Azure CLI, you need to login to your Azure account. You can do this by running the following command in your terminal or command prompt:
   ```
   az login
   ```
   This command will open a new browser window where you can enter your Azure credentials.

3. List all storage accounts: Once you are logged in, you can list all the storage accounts in your subscription by running the following command:
   ```
   az storage account list
   ```
   This command will return a JSON object with all the storage accounts in your subscription.

4. Check Access Keys: For each storage account, you can check the last time the access keys were rotated by running the following command:
   ```
   az storage account keys list --account-name <YourStorageAccountName>
   ```
   Replace `<YourStorageAccountName>` with the name of your storage account. This command will return a JSON object with the details of the access keys for the specified storage account. Look for the `creationTime` field in the JSON object to see when the access keys were last rotated.

#### Using Python

Detecting Access Keys Not Rotated in Storage can be done using Python scripts. Here are the steps:

1. AWS:
   - Use the AWS SDK for Python (Boto3) to interact with AWS services.
   - Use the IAM client from Boto3 to list all the access keys.
   - For each access key, use the `get_access_key_last_used` method to get the last used date.
   - If the last used date is more than 90 days, then the key is not rotated.
   - Python script example:
     ```python
     import boto3
     from datetime import datetime, timedelta

     iam = boto3.client('iam')
     response = iam.list_access_keys(UserName='username')

     for key in response['AccessKeyMetadata']:
         last_used_response = iam.get_access_key_last_used(AccessKeyId=key['AccessKeyId'])
         if 'LastUsedDate' in last_used_response['AccessKeyLastUsed']:
             last_used_date = last_used_response['AccessKeyLastUsed']['LastUsedDate']
             if last_used_date < datetime.now() - timedelta(days=90):
                 print(f"Access key {key['AccessKeyId']} not rotated")
     ```

2. Azure:
   - Use the Azure SDK for Python to interact with Azure services.
   - Use the `list_keys` method from the StorageAccountManagementClient to list all the keys.
   - For each key, check the `creation_time` property.
   - If the creation time is more than 90 days, then the key is not rotated.
   - Python script example:
     ```python
     from azure.mgmt.storage import StorageManagementClient
     from datetime import datetime, timedelta

     storage_client = StorageManagementClient(credentials, subscription_id)
     keys = storage_client.storage_accounts.list_keys(resource_group_name, account_name)

     for key in keys.keys:
         if key.creation_time < datetime.now() - timedelta(days=90):
             print(f"Key {key.key_name} not rotated")
     ```

3. GCP:
   - Use the Google Cloud SDK for Python to interact with GCP services.
   - Use the `list_service_account_keys` method from the IAM client to list all the keys.
   - For each key, check the `valid_after_time` property.
   - If the valid after time is more than 90 days, then the key is not rotated.
   - Python script example:
     ```python
     from google.cloud import iam
     from datetime import datetime, timedelta

     client = iam.ServiceAccountManagerClient()
     keys = client.list_service_account_keys(project_id, service_account_id)

     for key in keys.keys:
         if key.valid_after_time < datetime.now() - timedelta(days=90):
             print(f"Key {key.name} not rotated")
     ```

</Tab>

<Tab title='Remediation'>
### Remediation

#### Using Console

To remediate the issue of Access Keys not rotated in Azure, you can follow the below steps:

1. Login to the Azure portal with your credentials.
2. Navigate to 'Azure Active Directory' from the left-hand menu.
3. Click on 'Users' from the 'Manage' section.
4. Select the user whose access keys need to be rotated.
5. Click on 'Reset Password' from the top menu.
6. In the 'Reset Password' window, select 'Auto-generate password'.
7. Click on 'Reset'.
8. The new password will be displayed on the screen. Copy this password to a secure location.
9. Click on 'Show Password' to view the password again if required.
10. Click on 'Close' to exit the 'Reset Password' window.

Once you have reset the password, you will need to update the access keys for the user in all the applications and services that use them. This will ensure that the user has the latest access keys and the old keys are no longer in use.

#### Using CLI

Access keys not being rotated can pose a security risk as it increases the chances of unauthorized access to your Azure resources. To remediate this issue, you can follow the below steps using Azure CLI:

1. Login to your Azure account using Azure CLI by running the command `az login`.

2. Once you are logged in, run the following command to list all the storage accounts in your subscription: 

   `az storage account list --query [*].name --output tsv`

3. Select the storage account that you want to remediate and run the following command to list all the access keys associated with it: 

   `az storage account keys list --account-name <storage-account-name> --query [*].value --output tsv`

4. Once you have the access keys, create a new access key to replace the existing one by running the following command: 

   `az storage account keys renew --account-name <storage-account-name> --key primary`

   This will generate a new access key and invalidate the old one.

5. Repeat step 4 for the secondary key as well by running the following command: 

   `az storage account keys renew --account-name <storage-account-name> --key secondary`

6. Finally, verify that the keys have been rotated successfully by running the command in step 3 again. 

   Note: It is recommended to automate the access key rotation process using Azure Key Vault or Azure Functions to ensure that the keys are rotated on a regular basis.

#### Using Python

To remediate the issue of Access Keys not rotated in Azure using Python, you can follow the below steps:

Step 1: Install the Azure SDK for Python using pip.

```python
pip install azure
```

Step 2: Create a connection to Azure using the `ServicePrincipalCredentials` class. You will need to provide the `tenant_id`, `client_id`, and `client_secret`.

```python
from azure.common.credentials import ServicePrincipalCredentials

tenant_id = 'your_tenant_id'
client_id = 'your_client_id'
client_secret = 'your_client_secret'

credentials = ServicePrincipalCredentials(
    client_id=client_id,
    secret=client_secret,
    tenant=tenant_id
)
```

Step 3: Use the `azure.mgmt.authorization` package to get a list of all the users in your Azure subscription.

```python
from azure.mgmt.authorization import AuthorizationManagementClient

subscription_id = 'your_subscription_id'

authorization_client = AuthorizationManagementClient(
    credentials=credentials,
    subscription_id=subscription_id
)

users = authorization_client.role_assignments.list()
```

Step 4: For each user, check if the access key has been rotated in the last `n` days (where `n` is a value you define).

```python
from datetime import datetime, timedelta

n_days = 90

for user in users:
    if user.properties.created_time < datetime.now() - timedelta(days=n_days):
        # Access key has not been rotated in the last n_days days
        # Remediation steps go here
```

Step 5: Remediate the issue by rotating the access key for the user.

```python
from azure.graphrbac import GraphRbacManagementClient

graph_client = GraphRbacManagementClient(
    credentials=credentials,
    tenant_id=tenant_id
)

user_object_id = user.properties.principal_id

# Get the user's current access keys
user = graph_client.users.get(user_object_id)
access_keys = user.get_service_principals().current().get_password_credentials()

# Generate a new access key
new_access_key = generate_access_key()

# Update the user's access key
access_keys.password = new_access_key
graph_client.service_principals.update_password_credentials(user_object_id, access_keys)
```

Note: You will need to define the `generate_access_key()` function to generate a new access key. This function should return a string that meets the Azure password requirements.


</Tab>
</Tabs>