--- 
slug: CreateMountTarget
eventname: CreateMountTarget
title: CreateMountTarget
sidebar_label: CreateMountTarget
---
                       

### Event Information

#### Meaning

- The CreateMountTarget event in AWS Real-time for EFS refers to the creation of a mount target for an Amazon Elastic File System (EFS) in AWS. 
- A mount target is a network interface that allows EC2 instances to connect to an EFS file system securely. 
- This event indicates that a new mount target has been successfully created, enabling EC2 instances to access the EFS file system through the specified mount target.

#### Example

1. Possible Security Impact: Unauthorized Access to EFS File System
- If the CreateMountTarget event in AWS Real-time for EFS is not properly secured, it could potentially allow unauthorized access to the EFS file system. This can occur if the mount target is created with incorrect or weak security group settings, allowing unauthorized inbound traffic to the file system.

Remediation:
- Ensure that the mount target is created with the appropriate security group settings to restrict inbound traffic to only trusted sources.
- Use the following AWS CLI command to create a mount target with a specific security group:
```
aws efs create-mount-target --file-system-id <file-system-id> --subnet-id <subnet-id> --security-groups <security-group-ids>
```
- Replace `<file-system-id>` with the ID of the EFS file system, `<subnet-id>` with the ID of the subnet where the mount target will be created, and `<security-group-ids>` with the ID(s) of the desired security group(s).

2. Possible Security Impact: Exposure of Sensitive Data
- If the CreateMountTarget event in AWS Real-time for EFS is not properly secured, it could lead to the exposure of sensitive data stored in the EFS file system. This can occur if the mount target is created in a public subnet without any additional security measures, allowing unauthorized access to the file system.

Remediation:
- Ensure that the mount target is created in a private subnet or a subnet with restricted access.
- Use the following AWS CLI command to create a mount target in a specific subnet:
```
aws efs create-mount-target --file-system-id <file-system-id> --subnet-id <subnet-id>
```
- Replace `<file-system-id>` with the ID of the EFS file system and `<subnet-id>` with the ID of the desired subnet.

3. Possible Security Impact: Lack of Encryption for Data in Transit
- If the CreateMountTarget event in AWS Real-time for EFS is not properly secured, it could result in data being transmitted over the network without encryption. This can occur if the mount target is created without enabling encryption options, potentially exposing sensitive data to interception or unauthorized access.

Remediation:
- Enable encryption for data in transit by creating the mount target with the `--encrypted` flag.
- Use the following AWS CLI command to create an encrypted mount target:
```
aws efs create-mount-target --file-system-id <file-system-id> --subnet-id <subnet-id> --encrypted
```
- Replace `<file-system-id>` with the ID of the EFS file system and `<subnet-id>` with the ID of the desired subnet.

 