
### Event Information

- The "Add user" event in Azure Active Directory refers to the action of creating a new user account within the Azure AD tenant.
- This event signifies the addition of a new user to the Azure AD directory, allowing them to authenticate and access resources within the Azure environment.
- The "Add user" event is typically triggered when an administrator or user with appropriate permissions creates a new user account using the Azure AD management portal, PowerShell, or through API calls.


### Examples

1. Weak or easily guessable passwords: When adding a user in Azure Active Directory, it is important to ensure that strong passwords are set for each user. Weak or easily guessable passwords can make it easier for unauthorized individuals to gain access to the user's account, potentially compromising the security of the Azure environment.

2. Lack of multi-factor authentication (MFA): Multi-factor authentication adds an extra layer of security by requiring users to provide additional verification, such as a code sent to their mobile device, in addition to their password. Failing to enable MFA for user accounts in Azure Active Directory can increase the risk of unauthorized access if passwords are compromised.

3. Incorrect assignment of permissions: When adding a user in Azure Active Directory, it is important to carefully assign the appropriate permissions based on the user's role and responsibilities. Incorrectly assigning permissions can result in users having access to resources or actions they should not have, potentially leading to security breaches or unauthorized activities within the Azure environment.

### Remediation

#### Using Console

To remediate weak or easily guessable passwords, lack of multi-factor authentication (MFA), and incorrect assignment of permissions in Azure Active Directory, you can follow these steps:

1. Weak or easily guessable passwords:
   - In the Azure portal, navigate to Azure Active Directory.
   - Select "Users" and choose the user for whom you want to set a strong password.
   - Under the "Password reset" section, click on "Reset password".
   - Set a strong password that meets the Azure password complexity requirements.
   - Click on "Reset" to save the new password.

2. Lack of multi-factor authentication (MFA):
   - In the Azure portal, navigate to Azure Active Directory.
   - Select "Users" and choose the user for whom you want to enable MFA.
   - Under the "Security" section, click on "Multi-factor authentication".
   - In the MFA settings page, select the user and click on "Enable" to enable MFA for the user.
   - Follow the prompts to set up the preferred MFA method, such as using a mobile app or receiving a verification code via SMS.

3. Incorrect assignment of permissions:
   - In the Azure portal, navigate to the resource or resource group for which you want to manage permissions.
   - Select "Access control (IAM)" and click on "Add".
   - Choose the appropriate role for the user, such as "Contributor" or "Reader".
   - Search for and select the user or group that needs the correct permissions.
   - Click on "Save" to assign the selected role to the user or group.

By following these steps, you can remediate weak passwords, enable MFA, and ensure correct assignment of permissions in Azure Active Directory.

#### Using CLI

1. To remediate weak or easily guessable passwords in Azure Active Directory using Azure CLI, you can enforce password policies and set password complexity requirements. Use the following command:

```
az ad policy password update --id <policy-id> --password-validity-period <validity-period> --password-reuse-history-count <reuse-history-count> --password-reuse-prevention <reuse-prevention> --password-min-length <min-length> --password-max-length <max-length> --password-character-set <character-set>
```

Replace `<policy-id>` with the ID of the password policy you want to update, and adjust the other parameters according to your desired password requirements.

2. To remediate the lack of multi-factor authentication (MFA) in Azure Active Directory using Azure CLI, you can enable MFA for user accounts. Use the following command:

```
az ad user update --id <user-id> --force-change-password-next-login true --password-policies DisablePasswordExpiration --add authenticationMethods/password
```

Replace `<user-id>` with the ID of the user you want to enable MFA for. This command forces the user to change their password at the next login and adds password authentication as one of the authentication methods.

3. To remediate incorrect assignment of permissions in Azure Active Directory using Azure CLI, you can review and update the user's role assignments. Use the following command:

```
az role assignment create --assignee <user-id> --role <role-name> --scope <scope>
```

Replace `<user-id>` with the ID of the user you want to assign the role to, `<role-name>` with the name of the desired role, and `<scope>` with the scope of the assignment (e.g., resource group, subscription, etc.). This command creates a new role assignment for the user.

#### Using Python

To remediate weak or easily guessable passwords, you can use the Azure Active Directory PowerShell module to enforce password policies and set strong password requirements for users. Here's an example of a Python script that utilizes the Azure AD Graph API to achieve this:

```python
import requests
import json

# Define the Azure AD Graph API endpoint
graph_api_endpoint = 'https://graph.windows.net/{your-tenant-id}/users/{user-object-id}?api-version=1.6'

# Define the access token for authentication
access_token = 'your-access-token'

# Define the new password policy
new_password_policy = {
    "passwordPolicies": "DisablePasswordExpiration, DisableStrongPassword"
}

# Update the user's password policy
response = requests.patch(graph_api_endpoint, headers={'Authorization': 'Bearer ' + access_token}, data=json.dumps(new_password_policy))

# Check the response status code
if response.status_code == 204:
    print("Password policy updated successfully.")
else:
    print("Failed to update password policy.")
```

To enable multi-factor authentication (MFA) for user accounts in Azure Active Directory, you can use the Azure AD PowerShell module. Here's an example of a Python script that enables MFA for a specific user:

```python
import requests
import json

# Define the Azure AD Graph API endpoint
graph_api_endpoint = 'https://graph.windows.net/{your-tenant-id}/users/{user-object-id}?api-version=1.6'

# Define the access token for authentication
access_token = 'your-access-token'

# Define the new MFA settings
new_mfa_settings = {
    "strongAuthenticationMethods": [
        {
            "type": "OneWaySMS",
            "usageLocation": "US"
        }
    ]
}

# Update the user's MFA settings
response = requests.patch(graph_api_endpoint, headers={'Authorization': 'Bearer ' + access_token}, data=json.dumps(new_mfa_settings))

# Check the response status code
if response.status_code == 204:
    print("MFA enabled successfully.")
else:
    print("Failed to enable MFA.")
```

To correct the incorrect assignment of permissions in Azure Active Directory, you can use the Azure AD PowerShell module to review and modify the permissions assigned to users. Here's an example of a Python script that lists the permissions assigned to a specific user:

```python
import requests

# Define the Azure AD Graph API endpoint
graph_api_endpoint = 'https://graph.windows.net/{your-tenant-id}/users/{user-object-id}/appRoleAssignments?api-version=1.6'

# Define the access token for authentication
access_token = 'your-access-token'

# Get the user's app role assignments
response = requests.get(graph_api_endpoint, headers={'Authorization': 'Bearer ' + access_token})

# Check the response status code
if response.status_code == 200:
    app_role_assignments = response.json().get('value', [])
    for assignment in app_role_assignments:
        print("App Role: " + assignment['appRoleId'])
        print("Resource: " + assignment['resourceId'])
        print("Principal: " + assignment['principalId'])
        print("---")
else:
    print("Failed to retrieve app role assignments.")
```

You can modify the permissions assigned to users by making appropriate changes to the app role assignments using the Azure AD Graph API.

