--- 
slug: Add service principal credentials
eventname: Add service principal credentials
title: Add service principal credentials
sidebar_label: Add service principal credentials
---
                       
### Event Information

#### Meaning

- The "Add service principal credentials" event in Azure Active Directory refers to the action of adding credentials (such as passwords or certificates) to a service principal object in Azure AD.
- Service principals are used to authenticate and authorize applications and services to access resources in Azure. By adding credentials to a service principal, you are providing it with the necessary authentication information to access resources on behalf of the application or service it represents.
- This event is important for managing access control and security in Azure AD, as it allows you to grant specific permissions and control the level of access that a service principal has to Azure resources.

### Remediation

#### Using Console

- Example: If security is impacted with adding service principal credentials in Azure Active Directory (Azure AD), it could potentially lead to unauthorized access to resources or data within the Azure environment. For instance, if the service principal credentials are compromised or misconfigured, an attacker could gain unauthorized access to Azure resources and perform malicious activities.

- Remediation steps using Azure console:
  1. Sign in to the Azure portal (https://portal.azure.com) using your Azure AD credentials.
  2. Navigate to the Azure Active Directory service by searching for "Azure Active Directory" in the search bar.
  3. In the Azure Active Directory blade, select "App registrations" from the left-hand menu.
  4. Locate the specific service principal that needs remediation and click on it to open its details.
  5. Review the credentials associated with the service principal, including the client secret or certificate.
  6. If the credentials are compromised or misconfigured, click on "Certificates & secrets" or "Managed application in local directory" (depending on the type of credentials) to manage them.
  7. Remove any compromised or unnecessary credentials and generate new ones if required.
  8. Ensure that the remaining credentials are securely stored and properly managed.
  9. Regularly review and rotate the credentials to minimize the risk of unauthorized access.
  10. Monitor the Azure AD logs and enable Azure AD Identity Protection to detect and respond to any suspicious activities related to the service principal.

Note: The exact steps may vary slightly depending on the Azure portal version and layout. It is recommended to refer to the official Azure documentation for the most up-to-date instructions.

#### Using CLI

1. Impact on Security: Adding service principal credentials in Azure Active Directory (Azure AD) can potentially impact security if the credentials are not properly managed or if they fall into the wrong hands. This can lead to unauthorized access to Azure resources and sensitive data.

2. Example: Let's say an administrator accidentally adds service principal credentials with excessive permissions to Azure AD. This can result in a security breach if an attacker gains access to these credentials and abuses the elevated privileges to compromise resources or steal data.

3. Remediation using Azure CLI: To remediate this issue in Azure AD using Azure CLI, you can follow these steps:

   - Identify the service principal credentials that need to be removed or modified. You can use the `az ad sp list` command to list all the service principals in your Azure AD tenant.
   - Once you have identified the specific service principal, you can use the `az ad sp delete` command to delete it. For example, `az ad sp delete --id <service_principal_id>` will delete the service principal with the specified ID.
   - If you want to modify the permissions or roles associated with the service principal, you can use the `az ad sp update` command. For example, `az ad sp update --id <service_principal_id> --add <role>` will add a specific role to the service principal.

Note: It is important to carefully review and validate the changes before executing the commands to ensure that the correct service principal credentials are being modified or removed.

#### Using Python

1. Impact of adding service principal credentials in Azure Active Directory:
   - Unauthorized access: If the service principal credentials are compromised or misused, it can lead to unauthorized access to Azure resources, potentially exposing sensitive data or allowing malicious activities.

2. Remediation steps for Azure Active Directory using Python:
   - Rotate credentials regularly: Implement a process to regularly rotate the service principal credentials to minimize the impact of any potential compromise. This can be achieved using the Azure SDK for Python, specifically the `azure-identity` library, which provides functionality to manage service principal credentials.
   - Implement strong authentication methods: Enforce the use of strong authentication methods, such as multi-factor authentication (MFA), for service principal accounts to add an extra layer of security.
   - Monitor and audit service principal activities: Continuously monitor and audit the activities performed by service principals in Azure Active Directory. Utilize Azure Monitor and Azure Log Analytics to collect and analyze logs, and set up alerts for any suspicious or unauthorized activities.

Python script example for rotating service principal credentials using the `azure-identity` library:

```python
from azure.identity import DefaultAzureCredential
from azure.mgmt.authorization import AuthorizationManagementClient

# Create a DefaultAzureCredential object to authenticate
credential = DefaultAzureCredential()

# Create an instance of the AuthorizationManagementClient
authorization_client = AuthorizationManagementClient(credential, <subscription_id>)

# Get the service principal by its object id
service_principal = authorization_client.service_principals.get(<service_principal_object_id>)

# Generate new credentials for the service principal
new_credentials = authorization_client.service_principals.create_password_credentials(
    service_principal.id,
    password_credentials={
        "displayName": "New Credentials",
        "endDateTime": "<new_credentials_expiry_date>",
        "value": "<new_credentials_value>"
    }
)

# Delete the old credentials
authorization_client.service_principals.delete_password_credentials(
    service_principal.id,
    old_credentials_id="<old_credentials_id>"
)
```

Note: Replace `<subscription_id>`, `<service_principal_object_id>`, `<new_credentials_expiry_date>`, `<new_credentials_value>`, and `<old_credentials_id>` with the appropriate values in the above script.


 