--- 
slug: Remove policy from application
eventname: Remove policy from application
title: Remove policy from application
sidebar_label: Remove policy from application
---
                       

### Event Information

#### Meaning

- The "Remove policy from application" event in Azure Active Directory refers to the action of removing a policy that was previously assigned to an application.
- This event indicates that a policy, which defines certain access controls or security settings, has been revoked or unassigned from an application in Azure Active Directory.
- Removing a policy from an application can have implications on the access and permissions granted to users or groups associated with that application, and it is important to review and understand the impact of this event on the overall security posture of the Azure Active Directory environment.

#### Example

1. Example: If a malicious actor gains unauthorized access to the Azure Active Directory (AAD) and removes a security policy from an application, it can lead to a security breach. For instance, if the "Require MFA" policy is removed from an application, it allows users to authenticate without multi-factor authentication, increasing the risk of unauthorized access.

Remediation:
- Immediately revoke the compromised user's access and reset their credentials.
- Enable auditing and monitoring for AAD events to detect any suspicious activity.
- Regularly review and update security policies for applications in AAD.
- Implement conditional access policies to enforce multi-factor authentication for critical applications.
- Utilize Azure AD Privileged Identity Management (PIM) to limit the duration and scope of administrative access.

Azure CLI Command for enabling auditing and monitoring:
```
az monitor diagnostic-settings create --name <diagnostic-settings-name> --resource <aad-resource-id> --resource-type Microsoft.AzureActiveDirectory --logs '[{"category": "AuditLogs", "enabled": true}]'
```

2. Example: Another scenario where security can be impacted is if a user with elevated privileges mistakenly removes a critical security policy from an application. This can lead to unintended access or exposure of sensitive data.

Remediation:
- Regularly review and validate security policies for applications in AAD.
- Implement RBAC (Role-Based Access Control) to limit the number of users with administrative privileges.
- Enable Azure AD Privileged Identity Management (PIM) to enforce just-in-time access and approval workflows for administrative roles.
- Implement backup and restore mechanisms for critical security policies to quickly recover from accidental removal.

Azure CLI Command for enabling Azure AD Privileged Identity Management (PIM):
```
az ad pim enable --id <aad-resource-id>
```

3. Example: A misconfiguration in the Azure Active Directory can also lead to the removal of a security policy from an application. For instance, if incorrect permissions are assigned to a user or group, they may be able to modify or remove security policies unintentionally.

Remediation:
- Regularly review and validate permissions assigned to users and groups in Azure Active Directory.
- Implement least privilege principles and assign permissions based on the principle of least privilege.
- Utilize Azure AD Conditional Access policies to enforce additional security controls and restrictions.
- Enable Azure AD Privileged Identity Management (PIM) to monitor and manage administrative access to Azure AD resources.

Azure CLI Command for enabling Azure AD Conditional Access policies:
```
az ad conditional-access policy create --name <policy-name> --description <policy-description> --state enabled --grant-controls <grant-controls> --session-controls <session-controls> --users <user-ids> --applications <application-ids> --locations <location-ids>
```

 