--- 
slug: gcp_rt_vpc_subnetwork_changes
eventname: v1.compute.subnetworks.insert
title: v1.compute.subnetworks.insert
sidebar_label: v1.compute.subnetworks.insert
---
                       
### Event Information

- The v1.compute.subnetworks.insert event in GCP for Compute refers to the creation of a new subnetwork within a specific project and region.
- This event indicates that a user or automated process has initiated the creation of a subnetwork, which is a logical partition of a network that allows for more granular control over IP addressing and routing.
- The event can be triggered through the GCP Console, API, or command-line tools, and it signifies the start of the subnetwork creation process, including the allocation of IP ranges and configuration of routing settings.


### Examples

- Unauthorized access: If security is impacted with v1.compute.subnetworks.insert in GCP for Compute, it could potentially allow unauthorized users to create subnetworks, leading to a potential security breach. This could result in unauthorized access to sensitive data or resources within the network.
- Network misconfiguration: Improper use of v1.compute.subnetworks.insert API in GCP Compute can lead to network misconfiguration, which can impact security. For example, if the API is used to create subnetworks with incorrect firewall rules or routing configurations, it can expose resources to unauthorized access or disrupt network connectivity.
- Resource exhaustion: If the v1.compute.subnetworks.insert API is abused or misused, it can lead to resource exhaustion. For instance, an attacker could repeatedly create subnetworks using this API, consuming excessive resources and potentially impacting the availability and performance of the network. This can also result in increased costs for the organization.

### Remediation

#### Using Console

To remediate the issues mentioned in the previous response for GCP Compute using the GCP console, you can follow these step-by-step instructions:

1. Restricting SSH access:
   - Go to the GCP Console and navigate to the Compute Engine section.
   - Select the instance for which you want to restrict SSH access.
   - Click on the "Edit" button at the top of the page.
   - Scroll down to the "Firewalls" section and click on "Add firewall rule".
   - Provide a name for the firewall rule and set the "Targets" to "All instances in the network".
   - In the "Source IP ranges" field, enter the IP range from which you want to allow SSH access (e.g., your organization's IP range).
   - Set the "Protocols and ports" to allow SSH (port 22) traffic.
   - Click on the "Create" button to save the firewall rule.

2. Enabling VPC Flow Logs:
   - Go to the GCP Console and navigate to the VPC Network section.
   - Select the VPC network for which you want to enable flow logs.
   - Click on the "Edit" button at the top of the page.
   - Scroll down to the "Flow logs" section and click on "Add flow log".
   - Provide a name for the flow log and select the desired configuration options (e.g., sampling rate, metadata inclusion).
   - Choose the desired destination for the flow logs (e.g., Stackdriver Logging, Cloud Storage).
   - Click on the "Create" button to enable flow logs for the VPC network.

3. Implementing IAM best practices:
   - Go to the GCP Console and navigate to the IAM & Admin section.
   - Select the project for which you want to implement IAM best practices.
   - Click on the "IAM" tab to view the IAM roles and permissions.
   - Review the existing IAM roles and identify any unnecessary or overly permissive roles.
   - Remove any unnecessary roles and adjust the permissions of existing roles to follow the principle of least privilege.
   - Consider creating custom IAM roles with specific permissions tailored to the needs of different user groups.
   - Regularly review and audit the IAM roles and permissions to ensure they align with the organization's security requirements.

#### Using CLI

1. Enable VPC Flow Logs for GCP Compute instances:
   - Use the `gcloud compute instances update` command to enable VPC Flow Logs for a specific instance:
     ```
     gcloud compute instances update INSTANCE_NAME --enable-network-endpoint-logging
     ```

2. Restrict SSH access to GCP Compute instances:
   - Use the `gcloud compute firewall-rules update` command to update the firewall rule for SSH access:
     ```
     gcloud compute firewall-rules update FIREWALL_RULE_NAME --source-ranges=IP_RANGE --allow=tcp:22
     ```

3. Implement disk encryption for GCP Compute instances:
   - Use the `gcloud compute disks create` command to create an encrypted disk:
     ```
     gcloud compute disks create DISK_NAME --size=SIZE --type=DISK_TYPE --encryption-key=KEY_NAME
     ```

#### Using Python

To remediate the issues mentioned in the previous response for GCP Compute using Python, you can use the following approaches:

1. Enforce strong passwords:
   - Use the Google Cloud Identity and Access Management (IAM) API to create a custom password policy for your GCP project.
   - Write a Python script to programmatically enforce the password policy by setting the minimum password length, complexity requirements, and expiration period for user accounts.

```python
from googleapiclient import discovery

def set_password_policy(project_id, min_length, complexity, expiration):
    service = discovery.build('iam', 'v1')
    policy = {
        'updateMask': 'passwordPolicy',
        'passwordPolicy': {
            'minLength': min_length,
            'requireSymbols': complexity['symbols'],
            'requireNumbers': complexity['numbers'],
            'requireLowercase': complexity['lowercase'],
            'requireUppercase': complexity['uppercase'],
            'expireAfterDays': expiration
        }
    }
    request = service.projects().updateIAMPolicy(resource=project_id, body=policy)
    response = request.execute()
    print('Password policy updated successfully.')

# Usage example
set_password_policy('your-project-id', 10, {'symbols': True, 'numbers': True, 'lowercase': True, 'uppercase': True}, 90)
```

2. Enable disk encryption:
   - Use the Google Cloud Disk Encryption API to enable disk encryption for your GCP Compute instances.
   - Write a Python script to programmatically enable disk encryption for existing instances or configure it for new instances.

```python
from googleapiclient import discovery

def enable_disk_encryption(project_id, zone, instance_name):
    service = discovery.build('compute', 'v1')
    instance = service.instances().get(project=project_id, zone=zone, instance=instance_name).execute()
    instance['disks'][0]['diskEncryptionKey'] = {
        'rawKey': 'your-encryption-key'
    }
    request = service.instances().setDiskEncryptionKey(project=project_id, zone=zone, instance=instance_name, body=instance)
    response = request.execute()
    print('Disk encryption enabled successfully.')

# Usage example
enable_disk_encryption('your-project-id', 'us-central1-a', 'your-instance-name')
```

3. Implement network security groups:
   - Use the Google Cloud Firewall API to create network security groups (firewall rules) to control inbound and outbound traffic to your GCP Compute instances.
   - Write a Python script to programmatically create and manage firewall rules based on your desired network security policies.

```python
from googleapiclient import discovery

def create_firewall_rule(project_id, rule_name, source_ip_range, target_tags, allowed_ports):
    service = discovery.build('compute', 'v1')
    firewall_rule = {
        'name': rule_name,
        'network': 'global/networks/default',
        'sourceRanges': [source_ip_range],
        'targetTags': target_tags,
        'allowed': [{'IPProtocol': 'tcp', 'ports': allowed_ports}]
    }
    request = service.firewalls().insert(project=project_id, body=firewall_rule)
    response = request.execute()
    print('Firewall rule created successfully.')

# Usage example
create_firewall_rule('your-project-id', 'allow-http', '0.0.0.0/0', ['web-servers'], [80, 443])
```

Please note that the above scripts are just examples and may need to be modified based on your specific requirements and environment.


 