---
slug: redshift_reserved_node_recent_purchase
title: Redshift Reserved Node Recent Purchases Should Be Reviewed
sidebar_label: Redshift Reserved Node Recent Purchases Should Be Reviewed
---

### More Info:

Ensure Redshift Reserved Node purchases are regularly reviewed for cost optimization (informational).

### Risk Level

Low

### Address

Cost Optimisation

### Compliance Standards

CBP


### Remediation

#### Using Console

To remediate the misconfiguration "Redshift Reserved Node Recent Purchases Should Be Reviewed" for AWS using the AWS console, you can follow these steps:

1. Log in to the AWS Management Console and navigate to the Amazon Redshift console.
2. In the left navigation pane, choose "Reserved Nodes".
3. Review the list of recently purchased reserved nodes and identify any that are not needed or are not being used.
4. Select the unused reserved nodes and choose "Modify Reserved Nodes" from the "Actions" menu.
5. In the "Modify Reserved Nodes" dialog box, adjust the number of reserved nodes to match your current needs. You can also choose to exchange your reserved nodes for a different type or term.
6. Click "Modify" to save your changes.
7. Repeat this process for any other unused reserved nodes.

By following these steps, you can review and adjust your reserved node purchases to ensure that you are not paying for unused resources, and that you are maximizing the value of your reserved node investments.

#### Using CLI

The "Redshift Reserved Node Recent Purchases Should Be Reviewed" misconfiguration means that a new reserved node has been purchased for Amazon Redshift in the last 30 days and it is recommended to review this purchase to ensure it is still needed and the correct type and size.

To remediate this misconfiguration using AWS CLI, follow these steps:

1. Open your terminal and install AWS CLI if you haven't already done so.
2. Run the following command to list all the reserved nodes for Amazon Redshift:

   ```
   aws redshift describe-reserved-nodes
   ```

3. Identify the recently purchased reserved node(s) based on the "OfferingType" and "State" fields.
4. Review the recent purchase and determine if it is still needed and the correct type and size.
5. If the recent purchase is not needed or the wrong type/size, you can modify or delete it using the following commands:

   - To modify a reserved node:

     ```
     aws redshift modify-reserved-node --reserved-node-id <reserved-node-id> --node-type <new-node-type> --num-nodes <new-num-nodes>
     ```

     Replace `<reserved-node-id>` with the ID of the reserved node you want to modify and `<new-node-type>` and `<new-num-nodes>` with the new type and number of nodes you want to use.

   - To delete a reserved node:

     ```
     aws redshift delete-reserved-node --reserved-node-id <reserved-node-id>
     ```

     Replace `<reserved-node-id>` with the ID of the reserved node you want to delete.

6. After modifying or deleting the reserved node, run the first command again to confirm that the recent purchase has been remediated.

#### Using Python

To remediate the misconfiguration "Redshift Reserved Node Recent Purchases Should Be Reviewed" for AWS using Python, you can follow these steps:

1. Define the AWS credentials for your account using the `boto3` library:

```python
import boto3

# Define the AWS credentials
aws_access_key_id = 'your_access_key_id'
aws_secret_access_key = 'your_secret_access_key'
region_name = 'your_region_name'

# Create a Redshift client using the AWS credentials
client = boto3.client('redshift', 
                      aws_access_key_id=aws_access_key_id, 
                      aws_secret_access_key=aws_secret_access_key, 
                      region_name=region_name)
```

2. Use the `describe_reserved_nodes` method to get the list of all reserved nodes in your account:

```python
# Get the list of all reserved nodes in your account
reserved_nodes = client.describe_reserved_nodes()
```

3. Iterate through the list of reserved nodes and check if any of them have been purchased recently:

```python
# Define the threshold for recent purchases (in days)
recent_threshold = 30

# Iterate through the list of reserved nodes and check if any of them have been purchased recently
for node in reserved_nodes['ReservedNodes']:
    # Get the purchase date of the reserved node
    purchase_date = node['StartTime'].date()
    
    # Calculate the number of days since the purchase date
    days_since_purchase = (datetime.now().date() - purchase_date).days
    
    # Check if the reserved node has been purchased recently
    if days_since_purchase < recent_threshold:
        # If the reserved node has been purchased recently, print a warning message
        print(f"Reserved node {node['ReservedNodeId']} has been purchased recently ({days_since_purchase} days ago). Please review the recent purchases.")
```

4. If any reserved node has been purchased recently, review the recent purchases and make sure they are still necessary. If they are not necessary, you can modify or delete the reserved nodes using the `modify_reserved_nodes` or `delete_reserved_nodes` methods, respectively.

```python
# Modify a reserved node
client.modify_reserved_nodes(ReservedNodeIds=['your_reserved_node_id'], 
                             TargetNodeCount=2)

# Delete a reserved node
client.delete_reserved_nodes(ReservedNodeIds=['your_reserved_node_id'])
```

By following these steps, you can remediate the misconfiguration "Redshift Reserved Node Recent Purchases Should Be Reviewed" for AWS using Python.


### Additional Reading:

- [https://docs.aws.amazon.com/redshift/latest/mgmt/purchase-reserved-node-instance.html](https://docs.aws.amazon.com/redshift/latest/mgmt/purchase-reserved-node-instance.html) 

