---
slug: azure_audit_sqlserver_audit_action_groups_invalid
title: Ensure that AuditActionGroups in auditing is set properly
sidebar_label: Ensure that AuditActionGroups in auditing is set properly
---

### More Info:

Configure the 'AuditActionGroups' property to appropriate groups to capture all the critical activities on the SQL Server and all the SQL databases hosted on the SQL server.

### Risk Level

Medium

### Address

Reliability, Security

### Compliance Standards

HITRUST, SOC2, NISTCSF, PCIDSS


<Tabs><Tab title='Cause'>
### Check Cause

#### Using Console

1. Log in to the Azure portal: Open your web browser, navigate to the Azure portal (https://portal.azure.com/), and sign in with your Azure account credentials.

2. Navigate to SQL servers: On the left-hand side menu, click on "SQL servers" under the "Databases" section. This will open a list of all your SQL servers.

3. Select the SQL server: From the list of SQL servers, click on the name of the server for which you want to check the AuditActionGroups configuration.

4. Check Audit settings: On the SQL server page, click on "Auditing" under the "Security" section. Here, you can see the current AuditActionGroups settings. If the "AuditActionGroups" is not set properly, it means there is a misconfiguration. 

Remember, the AuditActionGroups should be set to capture all the necessary actions that you want to audit. If it's not set to capture certain actions, those actions won't be audited, which could lead to security risks.

#### Using CLI

1. Install Azure CLI: Before you can run any Azure CLI commands, you need to install the Azure CLI on your local machine. You can download it from the official Microsoft website. Once installed, open your command prompt or terminal.

2. Login to Azure: Use the following command to log in to your Azure account. You will be prompted to enter your username and password.
   ```
   az login
   ```
3. Select the Subscription: If you have multiple subscriptions, select the subscription where the SQL server is located using the following command. Replace 'my-subscription-id' with your actual subscription id.
   ```
   az account set --subscription 'my-subscription-id'
   ```
4. Check AuditActionGroups: Now, you can check the AuditActionGroups for your SQL server using the following command. Replace 'my-resource-group' with the name of your resource group and 'my-server' with the name of your SQL server.
   ```
   az sql server audit-policy show --name 'my-server' --resource-group 'my-resource-group' --query 'auditActionsAndGroups'
   ```
This command will return a list of AuditActionGroups. You should check if the required AuditActionGroups are present in this list.

#### Using Python

1. Install the necessary Python libraries: To interact with Azure resources, you need to install the Azure SDK for Python. You can do this using pip:

   ```bash
   pip install azure-mgmt-sql
   ```

2. Import the necessary modules and set up Azure credentials: You need to import the necessary modules and set up your Azure credentials. You can do this using the Azure Identity library and the Azure Management Client.

   ```python
   from azure.identity import DefaultAzureCredential
   from azure.mgmt.sql import SqlManagementClient

   credential = DefaultAzureCredential()
   subscription_id = 'your-subscription-id'  # replace with your Azure Subscription ID
   sql_client = SqlManagementClient(credential, subscription_id)
   ```

3. Get the list of SQL servers and databases: You can get the list of SQL servers and databases in your subscription using the `sql_client`.

   ```python
   for server in sql_client.servers.list():
       for database in sql_client.databases.list_by_server(server.resource_group, server.name):
           print(f"Server: {server.name}, Database: {database.name}")
   ```

4. Check the AuditActionGroups: You can check the AuditActionGroups for each database using the `sql_client`. If the AuditActionGroups is not set properly, print a message.

   ```python
   for server in sql_client.servers.list():
       for database in sql_client.databases.list_by_server(server.resource_group, server.name):
           audit_policy = sql_client.database_blob_auditing_policies.get(server.resource_group, server.name, database.name)
           if 'SUCCESSFUL_DATABASE_AUTHENTICATION_GROUP' not in audit_policy.audit_actions_and_groups:
               print(f"AuditActionGroups is not set properly for Server: {server.name}, Database: {database.name}")
   ```

This script will print a message for each database where the AuditActionGroups is not set properly. You can modify this script to suit your needs, for example by checking for other action groups or by taking some action when a misconfiguration is detected.

</Tab>


<Tab title='Remediation'>
### Remediation

#### Using Console

To remediate the misconfiguration related to AuditActionGroups in Azure, please follow the below steps:

1. Login to the Azure portal (https://portal.azure.com/).
2. Go to the Azure Active Directory service.
3. Select the "Audit logs" option under the Monitoring section.
4. In the Audit logs blade, click on the "Diagnostic settings" option.
5. Select the diagnostic setting that needs to be remediated.
6. In the "Diagnostic settings" blade, scroll down to the "Categories" section.
7. In the "Categories" section, ensure that the "AuditLogs" option is selected.
8. Under the "AuditLogs" option, select the "Select specific actions" radio button.
9. In the "Select specific actions" section, ensure that all the required AuditActionGroups are selected.
10. Click on the "Save" button to save the changes.

By following the above steps, the misconfiguration related to AuditActionGroups in Azure can be remediated.

#### Using CLI

To remediate the misconfiguration of AuditActionGroups in auditing for AZURE using AZURE CLI, follow these steps:

1. Open the AZURE CLI on your local machine or use the AZURE Cloud Shell.
2. Run the following command to get the current configuration of AuditActionGroups:

   ```
   az monitor activity-log list --query [].categories.actionGroups
   ```

3. Check the output of the above command to see if AuditActionGroups are set properly. If not, proceed to the next step.
4. Run the following command to set the AuditActionGroups:

   ```
   az monitor activity-log update --set categories.actionGroups=<comma separated list of action groups>
   ```

   Replace `<comma separated list of action groups>` with the appropriate list of action groups. For example, if you want to set the AuditActionGroups to "Write", "Delete", and "Action", the command would be:

   ```
   az monitor activity-log update --set categories.actionGroups=Write,Delete,Action
   ```

5. Verify the configuration by running the first command again:

   ```
   az monitor activity-log list --query [].categories.actionGroups
   ```

   The output should now show the updated list of AuditActionGroups.

By following these steps, you can remediate the misconfiguration of AuditActionGroups in auditing for AZURE using AZURE CLI.

#### Using Python

To remediate the misconfiguration of AuditActionGroups in Azure using Python, follow the below steps:

1. Import the necessary libraries:

```python
from azure.identity import DefaultAzureCredential
from azure.mgmt.monitor import MonitorManagementClient
```

2. Set the credentials:

```python
credential = DefaultAzureCredential()
subscription_id = '<Your Subscription ID>'
```

3. Initialize the MonitorManagementClient:

```python
monitor_client = MonitorManagementClient(credential, subscription_id)
```

4. Get the existing AuditActionGroups:

```python
audit_action_groups = monitor_client.activity_log_alerts.list_action_groups(resource_group_name='<Your Resource Group Name>', action_group_name='<Your Action Group Name>')
```

5. Update the AuditActionGroups:

```python
updated_audit_action_groups = [
    {
        "id": "/subscriptions/<Your Subscription ID>/resourceGroups/<Your Resource Group Name>/providers/microsoft.insights/actionGroups/<Your Action Group Name>",
        "action_group_type": "CustomEmail/SMS/Push/Voice",
        "short_name": "<Your Short Name>",
        "email_receivers": [
            {
                "name": "<Your Email Name>",
                "email_address": "<Your Email Address>",
                "use_common_alert_schema": True
            }
        ]
    }
]

monitor_client.action_groups.create_or_update(resource_group_name='<Your Resource Group Name>', action_group_name='<Your Action Group Name>', parameters=updated_audit_action_groups)
```

6. Verify the updated AuditActionGroups:

```python
updated_audit_action_groups = monitor_client.activity_log_alerts.list_action_groups(resource_group_name='<Your Resource Group Name>', action_group_name='<Your Action Group Name>')
```

By following these steps, you can remediate the misconfiguration of AuditActionGroups in Azure using Python.




</Tab>
</Tabs>