---
slug: acm_certificates_validity
title: AWS ACM Certificates Not Valid
sidebar_label: AWS ACM Certificates Not Valid
---

### More Info:

Ensure that all the requests made during SSL/TLS certificate issue or renewal process are validated. These requests are managed within your account by the Amazon Certificate Manager (ACM), an AWS service that lets you provision, deploy and maintain SSL/TLS certificates for use with other AWS resources such as ELB load balancers, CloudFront distributions or APIs via Amazon API Gateway.

### Risk Level

Medium

### Address

Security

### Compliance Standards

NIST


### Triage and Remediation
<Tabs>
<Tab title='Remediation'>
### Remediation

<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
Sure, here are the step-by-step instructions to remediate the AWS ACM Certificates Not Valid misconfiguration:

1. Firstly, log in to your AWS Management Console and navigate to the AWS Certificate Manager (ACM) service.

2. In the ACM dashboard, locate the certificate that is displaying the "Not Valid" status.

3. Click on the certificate name to view the details of the certificate.

4. In the certificate details page, check the "Status" field to see if it is showing "Pending Validation" or "Failed". If the status is "Pending Validation", then you need to complete the validation process. If the status is "Failed", then you need to identify the reason for the failure and take appropriate action to fix it.

5. If the certificate is still in the "Pending Validation" status, click on the "Domain" field to view the validation options.

6. Choose the appropriate validation method based on your domain registrar and follow the instructions provided by ACM to complete the validation process.

7. Once the validation process is completed successfully, the certificate status will change to "Issued" and the certificate will become valid.

8. If the certificate status is "Failed", check the "Reason" field to identify the reason for the failure.

9. Based on the reason for the failure, take appropriate action to fix the issue. For example, if the failure reason is "DNS validation failed", then you need to ensure that the DNS records for your domain are correctly configured.

10. After fixing the issue, click on the "Request Validation" button to initiate the validation process again.

11. Once the validation process is completed successfully, the certificate status will change to "Issued" and the certificate will become valid.

That's it! Following these steps should help you remediate the AWS ACM Certificates Not Valid misconfiguration.

#
</Accordion>

<Accordion title='Using CLI'>
If you are facing an issue with AWS ACM certificates not being valid, you can follow the below steps to remediate it:

1. Verify the certificate status by running the following command:
   ```
   aws acm describe-certificate --certificate-arn <certificate-arn>
   ```
   Replace `<certificate-arn>` with the ARN of the certificate you want to verify.

2. Check the `Status` field in the output of the above command. If the status is `ISSUED`, then the certificate is valid. If the status is `PENDING_VALIDATION`, then the certificate is still being validated. If the status is `EXPIRED`, then the certificate has expired.

3. If the certificate has expired, you can renew it by running the following command:
   ```
   aws acm renew-certificate --certificate-arn <certificate-arn>
   ```
   Replace `<certificate-arn>` with the ARN of the certificate you want to renew.

4. If the certificate is still being validated, you can check the validation status by running the following command:
   ```
   aws acm describe-certificate --certificate-arn <certificate-arn> --certificate-statuses PENDING_VALIDATION
   ```
   Replace `<certificate-arn>` with the ARN of the certificate you want to check.

5. If the validation is successful, the status will change to `ISSUED`. If the validation fails, you can check the validation error by running the following command:
   ```
   aws acm describe-certificate --certificate-arn <certificate-arn> --certificate-statuses FAILED
   ```
   Replace `<certificate-arn>` with the ARN of the certificate you want to check.

6. Resolve the validation error and repeat the validation process.

7. Once the certificate is valid, you can update your resources to use the new certificate.
</Accordion>

<Accordion title='Using Python'>
To remediate the AWS ACM Certificates not valid misconfiguration using Python, you can use the `boto3` library to interact with the AWS ACM service. Here are the step-by-step instructions:

1. Install the `boto3` library using the following command:

```
pip install boto3
```

2. Import the `boto3` library and create an ACM client object:

```python
import boto3

acm_client = boto3.client('acm')
```

3. List all the ACM certificates using the `list_certificates()` method:

```python
response = acm_client.list_certificates()
```

4. Loop through the list of certificates and check if the certificate is valid using the `describe_certificate()` method:

```python
for certificate in response['CertificateSummaryList']:
    certificate_arn = certificate['CertificateArn']
    certificate_details = acm_client.describe_certificate(CertificateArn=certificate_arn)
    if certificate_details['Certificate']['Status'] != 'ISSUED':
        # Certificate is not valid
        # Remediation steps go here
```

5. If a certificate is not valid, you can either renew it or delete it. To renew a certificate, use the `renew_certificate()` method:

```python
acm_client.renew_certificate(CertificateArn=certificate_arn)
```

6. To delete a certificate, use the `delete_certificate()` method:

```python
acm_client.delete_certificate(CertificateArn=certificate_arn)
```

7. Once you have remediated all the invalid certificates, you can verify that all certificates are valid by repeating step 3 and checking that all certificates have a status of `ISSUED`.

Note: Before deleting a certificate, make sure that it is not being used by any resources in your AWS account.
</Accordion>

</AccordionGroup>
</Tab>
</Tabs>
### Additional Reading:

- [https://docs.aws.amazon.com/acm/latest/userguide/acm-certificate.html](https://docs.aws.amazon.com/acm/latest/userguide/acm-certificate.html) 

