

<Tabs><Tab title='Cause'>
### Check Cause

#### Using Console

1. Log in to the AWS Management Console and navigate to the API Gateway service.
2. In the API Gateway dashboard, select the APIs section on the left-hand side.
3. For each API, select the "Stages" option. Here, you can see the details of each stage of your API.
4. In the stage editor panel, under the "Settings" tab, check the "Client Certificate for Endpoint Verification" section. If the certificate is due to expire in less than 30 days, it will be indicated here.

#### Using CLI

1. Install and configure AWS CLI: Before you can start, you need to install AWS CLI on your local machine and configure it with your AWS account credentials. You can do this by running the following commands:

   ```
   pip install awscli
   aws configure
   ```

   You will be prompted to provide your AWS Access Key ID, Secret Access Key, Default region name, and Default output format.

2. List all the API Gateways: Use the following command to list all the API Gateways in your AWS account:

   ```
   aws apigateway get-rest-apis
   ```

   This command will return a list of all the REST APIs in your account. Note down the "id" of the API Gateway you want to check.

3. Get the details of the API Gateway: Use the following command to get the details of the API Gateway:

   ```
   aws apigateway get-rest-api --rest-api-id <API_ID>
   ```

   Replace `<API_ID>` with the id of the API Gateway you noted down in the previous step. This command will return the details of the API Gateway including the id of the ACM certificate associated with it.

4. Check the ACM certificate: Use the following command to check the ACM certificate:

   ```
   aws acm describe-certificate --certificate-arn <CERTIFICATE_ARN>
   ```

   Replace `<CERTIFICATE_ARN>` with the ARN of the ACM certificate associated with the API Gateway. This command will return the details of the ACM certificate including its expiration date. If the expiration date is less than 30 days from the current date, then the ACM certificate is due for renewal.

#### Using Python

1. Install and configure AWS SDK for Python (Boto3) on your local machine. This will allow you to interact with AWS services using Python. You can install it using pip:

```python
pip install boto3
```

2. Import the necessary modules and create a session using your AWS credentials:

```python
import boto3
from botocore.exceptions import BotoCoreError, ClientError

session = boto3.Session(
    aws_access_key_id='YOUR_ACCESS_KEY',
    aws_secret_access_key='YOUR_SECRET_KEY',
    region_name='YOUR_REGION'
)
```

3. Create a client for the 'apigateway' service and list all the Rest APIs:

```python
client = session.client('apigateway')
response = client.get_rest_apis()

apis = response['items']
```

4. Iterate over the APIs and check the expiration date of the associated ACM certificate. If the certificate is set to expire in less than 30 days, print a warning message:

```python
import datetime

for api in apis:
    try:
        certificate_arn = api['clientCertificateId']
        acm_client = session.client('acm')
        certificate = acm_client.describe_certificate(
            CertificateArn=certificate_arn
        )
        expiration_date = certificate['Certificate']['NotAfter']
        if (expiration_date - datetime.datetime.now()).days < 30:
            print(f"Warning: Certificate for API {api['name']} is set to expire in less than 30 days.")
    except (BotoCoreError, ClientError) as error:
        print(f"Error checking certificate for API {api['name']}: {error}")
```

This script will check all the APIs in the specified region and print a warning for each one that has a certificate set to expire in less than 30 days.

</Tab>

<Tab title='Remediation'>
### Remediation

#### Using Console

Sure, here are the step by step instructions to remediate the misconfiguration of AWS ACM Certificates Renewal Under 30 Days:

1. Login to your AWS console.
2. Go to the AWS Certificate Manager service.
3. Select the certificate that needs to be renewed.
4. Click on the "Renew" button.
5. In the "Renewal Settings" page, select the validity period for the renewed certificate. Ensure that the validity period is greater than 30 days.
6. Click on the "Next" button.
7. Review the details of the renewed certificate and confirm that they are correct.
8. Click on the "Renew" button.

Once the certificate is renewed, you will need to update the certificate on your website or application to ensure that it is using the renewed certificate. 

By following these steps, you will have successfully remediated the misconfiguration of AWS ACM Certificates Renewal Under 30 Days.

#### Using CLI

The following are the step-by-step instructions to remediate the AWS ACM Certificates Renewal Under 30 Days misconfiguration using AWS CLI:

1. Open the AWS CLI on your local machine.

2. Run the following command to list all the certificates that are expiring in the next 30 days:

   ```
   aws acm list-certificates --query "CertificateSummaryList[?NotAfter<=\`$(date -v+30d +%Y-%m-%dT%H:%M:%SZ)\`]" --output table
   ```

   This command will display a list of all the certificates that are expiring in the next 30 days.

3. Identify the certificate that needs to be renewed and note down its ARN.

4. Run the following command to request a new certificate:

   ```
   aws acm request-certificate --domain-name <domain-name> --validation-method DNS --subject-alternative-names <domain-name1> <domain-name2> --idempotency-token <idempotency-token>
   ```

   Replace `<domain-name>` with the domain name for which you want to request a certificate. If you want to add additional domain names, specify them using the `--subject-alternative-names` option. The `--validation-method` option specifies the validation method for the certificate. In this case, we are using DNS validation.

5. After requesting the certificate, you need to validate it. Run the following command to get the CNAME record that you need to add to your DNS configuration:

   ```
   aws acm describe-certificate --certificate-arn <certificate-arn> --query "Certificate.DomainValidationOptions[0].ResourceRecord" --output text
   ```

   Replace `<certificate-arn>` with the ARN of the certificate that you requested in step 4.

6. Add the CNAME record to your DNS configuration.

7. Run the following command to wait until the certificate is issued:

   ```
   aws acm wait certificate-validated --certificate-arn <certificate-arn>
   ```

   Replace `<certificate-arn>` with the ARN of the certificate that you requested in step 4.

8. After the certificate is issued, you can update your application to use the new certificate.

#### Using Python

To remediate the AWS ACM Certificates Renewal Under 30 Days misconfiguration using Python, you can follow these steps:

1. Install the AWS SDK for Python (boto3) using pip:

```python
pip install boto3
```

2. Create an AWS IAM user with the necessary permissions to access and manage ACM certificates. The user should have the following IAM policies attached:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "ACMPermissions",
            "Effect": "Allow",
            "Action": [
                "acm:ListCertificates",
                "acm:DescribeCertificate",
                "acm:RenewCertificate"
            ],
            "Resource": "*"
        }
    ]
}
```

3. Configure the AWS credentials in your local environment. You can do this by exporting the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables or by using the AWS CLI configure command.

```python
import boto3
import datetime

# Create a client for ACM
acm_client = boto3.client('acm')

# Get a list of all certificates
certificates = acm_client.list_certificates()

# Loop through the certificates and check their expiration dates
for cert in certificates['CertificateSummaryList']:
    cert_arn = cert['CertificateArn']
    cert_info = acm_client.describe_certificate(CertificateArn=cert_arn)
    cert_expiration = cert_info['Certificate']['NotAfter']
    days_until_expiration = (cert_expiration - datetime.datetime.now(datetime.timezone.utc)).days

    # Renew the certificate if it expires in less than 30 days
    if days_until_expiration < 30:
        acm_client.renew_certificate(CertificateArn=cert_arn)
```

This Python code will check for all ACM certificates in your AWS account and renew any certificates that expire in less than 30 days. You can run this code on a regular basis (e.g., using a cron job) to ensure that your certificates are always up to date.


</Tab>
</Tabs>