---
slug: acm_certificates_renewal_30_days
title: AWS ACM Certificates Renewal Under 30 Days
sidebar_label: AWS ACM Certificates Renewal Under 30 Days
---

### More Info:

Ensure that your SSL/TLS certificates managed by AWS ACM are renewed 30 days before their validity period ends. Certificate Manager is the AWS service that lets you easily provision, manage, and deploy SSL/TLS certificates for use with other AWS resources such as Elastic Load Balancers, CloudFront distributions or APIs on Amazon API Gateway.

### Risk Level

Medium

### Address

Security

### Compliance Standards

NIST


### Triage and Remediation
<Tabs>
<Tab title='Remediation'>
### Remediation

<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
Sure, here are the step by step instructions to remediate the misconfiguration of AWS ACM Certificates Renewal Under 30 Days:

1. Login to your AWS console.
2. Go to the AWS Certificate Manager service.
3. Select the certificate that needs to be renewed.
4. Click on the "Renew" button.
5. In the "Renewal Settings" page, select the validity period for the renewed certificate. Ensure that the validity period is greater than 30 days.
6. Click on the "Next" button.
7. Review the details of the renewed certificate and confirm that they are correct.
8. Click on the "Renew" button.

Once the certificate is renewed, you will need to update the certificate on your website or application to ensure that it is using the renewed certificate. 

By following these steps, you will have successfully remediated the misconfiguration of AWS ACM Certificates Renewal Under 30 Days.

#
</Accordion>

<Accordion title='Using CLI'>
The following are the step-by-step instructions to remediate the AWS ACM Certificates Renewal Under 30 Days misconfiguration using AWS CLI:

1. Open the AWS CLI on your local machine.

2. Run the following command to list all the certificates that are expiring in the next 30 days:

   ```
   aws acm list-certificates --query "CertificateSummaryList[?NotAfter<=\`$(date -v+30d +%Y-%m-%dT%H:%M:%SZ)\`]" --output table
   ```

   This command will display a list of all the certificates that are expiring in the next 30 days.

3. Identify the certificate that needs to be renewed and note down its ARN.

4. Run the following command to request a new certificate:

   ```
   aws acm request-certificate --domain-name <domain-name> --validation-method DNS --subject-alternative-names <domain-name1> <domain-name2> --idempotency-token <idempotency-token>
   ```

   Replace `<domain-name>` with the domain name for which you want to request a certificate. If you want to add additional domain names, specify them using the `--subject-alternative-names` option. The `--validation-method` option specifies the validation method for the certificate. In this case, we are using DNS validation.

5. After requesting the certificate, you need to validate it. Run the following command to get the CNAME record that you need to add to your DNS configuration:

   ```
   aws acm describe-certificate --certificate-arn <certificate-arn> --query "Certificate.DomainValidationOptions[0].ResourceRecord" --output text
   ```

   Replace `<certificate-arn>` with the ARN of the certificate that you requested in step 4.

6. Add the CNAME record to your DNS configuration.

7. Run the following command to wait until the certificate is issued:

   ```
   aws acm wait certificate-validated --certificate-arn <certificate-arn>
   ```

   Replace `<certificate-arn>` with the ARN of the certificate that you requested in step 4.

8. After the certificate is issued, you can update your application to use the new certificate.
</Accordion>

<Accordion title='Using Python'>
To remediate the AWS ACM Certificates Renewal Under 30 Days misconfiguration using Python, you can follow these steps:

1. Install the AWS SDK for Python (boto3) using pip:

```python
pip install boto3
```

2. Create an AWS IAM user with the necessary permissions to access and manage ACM certificates. The user should have the following IAM policies attached:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "ACMPermissions",
            "Effect": "Allow",
            "Action": [
                "acm:ListCertificates",
                "acm:DescribeCertificate",
                "acm:RenewCertificate"
            ],
            "Resource": "*"
        }
    ]
}
```

3. Configure the AWS credentials in your local environment. You can do this by exporting the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables or by using the AWS CLI configure command.

```python
import boto3
import datetime

# Create a client for ACM
acm_client = boto3.client('acm')

# Get a list of all certificates
certificates = acm_client.list_certificates()

# Loop through the certificates and check their expiration dates
for cert in certificates['CertificateSummaryList']:
    cert_arn = cert['CertificateArn']
    cert_info = acm_client.describe_certificate(CertificateArn=cert_arn)
    cert_expiration = cert_info['Certificate']['NotAfter']
    days_until_expiration = (cert_expiration - datetime.datetime.now(datetime.timezone.utc)).days

    # Renew the certificate if it expires in less than 30 days
    if days_until_expiration < 30:
        acm_client.renew_certificate(CertificateArn=cert_arn)
```

This Python code will check for all ACM certificates in your AWS account and renew any certificates that expire in less than 30 days. You can run this code on a regular basis (e.g., using a cron job) to ensure that your certificates are always up to date.
</Accordion>

</AccordionGroup>
</Tab>
</Tabs>
### Additional Reading:

- [https://docs.aws.amazon.com/acm/latest/userguide/managed-renewal.html](https://docs.aws.amazon.com/acm/latest/userguide/managed-renewal.html) 

