---
slug: acm_certificates_renewal_7_days
title: AWS ACM Certificates Renewal Under 7 Days
sidebar_label: AWS ACM Certificates Renewal Under 7 Days
---

### More Info:

Ensure that your SSL/TLS certificates managed by AWS ACM are renewed 7 days before their validity period ends. Certificate Manager is the AWS service that lets you easily provision, manage, and deploy SSL/TLS certificates for use with other AWS resources such as Elastic Load Balancers, CloudFront distributions or APIs on Amazon API Gateway.

### Risk Level

High

### Address

Security

### Compliance Standards

NIST


### Triage and Remediation
<Tabs>
<Tab title='Remediation'>
### Remediation

<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
The following are the step-by-step instructions to remediate the AWS ACM Certificates Renewal Under 7 Days misconfiguration using the AWS console:

1. Log in to the AWS Management Console.
2. Navigate to the AWS Certificate Manager (ACM) console.
3. Click on the certificate that needs to be renewed.
4. Click on the "Renew" button.
5. In the Renew Certificate wizard, choose the validation method that you prefer.
6. Click on "Review" to review the certificate details.
7. Verify that the certificate details are correct and click on "Renew Certificate".
8. Wait for the certificate renewal process to complete.
9. Once the renewal process is complete, update the certificate in your applications or services that use it.

Note: It is recommended to set up an automated renewal process for certificates to avoid this misconfiguration in the future.

#
</Accordion>

<Accordion title='Using CLI'>
The following are the step by step instructions on how to remediate the AWS ACM Certificates Renewal Under 7 Days misconfiguration using AWS CLI:

1. Open your AWS CLI and run the following command to list all the certificates in the us-east-1 region:

```
aws acm list-certificates --region us-east-1
```

2. Identify the certificate that needs to be renewed and make a note of its ARN.

3. Run the following command to check the expiration date of the certificate:

```
aws acm describe-certificate --certificate-arn <certificate-arn> --region us-east-1
```

Replace `<certificate-arn>` with the ARN of the certificate that needs to be renewed.

4. If the certificate is expiring in less than 7 days, then it needs to be renewed. Run the following command to request a new certificate:

```
aws acm request-certificate --domain-name <domain-name> --validation-method DNS --region us-east-1
```

Replace `<domain-name>` with the domain name for which the certificate needs to be renewed.

5. Once the certificate is requested, it needs to be validated. Run the following command to get the CNAME record that needs to be added to the DNS configuration:

```
aws acm describe-certificate --certificate-arn <certificate-arn> --region us-east-1 --query "Certificate.DomainValidationOptions[0].ResourceRecord" --output text
```

6. Add the CNAME record to the DNS configuration of the domain name.

7. Run the following command to confirm that the certificate is validated:

```
aws acm describe-certificate --certificate-arn <certificate-arn> --region us-east-1 --query "Certificate.DomainValidationOptions[0].ValidationStatus" --output text
```

8. Once the certificate is validated, it can be used. Run the following command to update the certificate:

```
aws acm add-tags-to-certificate --certificate-arn <certificate-arn> --tags Key=Name,Value=<certificate-name> --region us-east-1
```

Replace `<certificate-arn>` with the ARN of the new certificate and `<certificate-name>` with a name for the new certificate.

9. Finally, update the certificate in the application or service that is using it with the new certificate ARN.

These steps should remediate the AWS ACM Certificates Renewal Under 7 Days misconfiguration.
</Accordion>

<Accordion title='Using Python'>
To remediate the AWS ACM Certificates Renewal Under 7 Days misconfiguration using Python, you can follow these steps:

1. Import the required libraries: boto3, datetime

```
import boto3
import datetime
```

2. Create an AWS ACM client object:

```
acm_client = boto3.client('acm')
```

3. Get a list of all the ACM certificates:

```
certificates = acm_client.list_certificates()
```

4. Loop through each certificate and check if it is expiring in less than 7 days:

```
for certificate in certificates['CertificateSummaryList']:
    certificate_details = acm_client.describe_certificate(CertificateArn=certificate['CertificateArn'])
    expiry_date = certificate_details['Certificate']['NotAfter']
    days_left = (expiry_date - datetime.datetime.now()).days
    if days_left < 7:
        # Renew the certificate
        acm_client.renew_certificate(CertificateArn=certificate['CertificateArn'])
```

5. If a certificate is expiring in less than 7 days, renew it using the `renew_certificate` method of the ACM client object.

6. Save the Python file and run it periodically to check for expiring certificates and renew them automatically.

Note: Make sure to set up AWS credentials in the environment where you are running the Python script.
</Accordion>

</AccordionGroup>
</Tab>
</Tabs>
### Additional Reading:

- [https://docs.aws.amazon.com/acm/latest/userguide/managed-renewal.html](https://docs.aws.amazon.com/acm/latest/userguide/managed-renewal.html) 

