---
slug: stack_failed_status
title: CloudFormation Stacks Should Not Have A Failed Status
sidebar_label: CloudFormation Stacks Should Not Have A Failed Status
---

### More Info:

None of your Amazon CloudFormation stacks should be in Failed mode for more than 6 hours. Any failed CloudFormation stacks that are not fixed on time can lead to application downtime, security issues or unexpected costs on your AWS bill.

### Risk Level

Informational

### Address

Operational Maturity

### Compliance Standards

CBP



### Triage and Remediation
<Tabs>


<Tab title='Prevention'>
### How to Prevent
<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
To prevent CloudFormation Stacks from having a failed status in AWS CloudFormation using the AWS Management Console, follow these steps:

1. **Template Validation**:
   - Before creating or updating a stack, use the AWS CloudFormation Designer or the "Validate Template" feature in the console to ensure that your CloudFormation template is syntactically correct and adheres to AWS resource specifications.

2. **Resource Limits and Quotas**:
   - Check AWS service limits and quotas for the resources you are deploying. Ensure that you are not exceeding any limits, such as the number of EC2 instances, IAM roles, or VPCs, which could cause stack creation to fail.

3. **IAM Permissions**:
   - Ensure that the IAM roles and users executing the CloudFormation stack have the necessary permissions to create, update, and delete the resources defined in the template. Use the IAM Policy Simulator to verify permissions.

4. **Parameter Validation**:
   - When defining parameters in your CloudFormation template, use constraints and default values to ensure that the input provided during stack creation is valid. This can prevent runtime errors due to invalid parameter values.

By following these steps, you can minimize the risk of encountering a failed status in your CloudFormation stacks.
</Accordion>

<Accordion title='Using CLI'>
To prevent CloudFormation stacks from having a failed status using AWS CLI, you can follow these steps:

1. **Validate the Template Before Deployment:**
   Use the `aws cloudformation validate-template` command to ensure that your CloudFormation template is syntactically correct and can be processed by CloudFormation.

   ```sh
   aws cloudformation validate-template --template-body file://template.json
   ```

2. **Use Change Sets for Safe Updates:**
   Create and review change sets before executing them to understand the impact of your changes. This helps in preventing unexpected failures.

   ```sh
   aws cloudformation create-change-set --stack-name my-stack --template-body file://template.json --change-set-name my-change-set
   aws cloudformation describe-change-set --change-set-name my-change-set --stack-name my-stack
   ```

3. **Enable Stack Policy:**
   Apply a stack policy to protect critical resources from being unintentionally updated or deleted during stack updates.

   ```sh
   aws cloudformation set-stack-policy --stack-name my-stack --stack-policy-body file://stack-policy.json
   ```

4. **Monitor Stack Events:**
   Continuously monitor stack events to catch and address issues early. Use the `describe-stack-events` command to get detailed information about the stack events.

   ```sh
   aws cloudformation describe-stack-events --stack-name my-stack
   ```

By following these steps, you can minimize the risk of your CloudFormation stacks ending up in a failed status.
</Accordion>

<Accordion title='Using Python'>
To prevent CloudFormation stacks from having a failed status using Python scripts, you can follow these steps:

1. **Validate CloudFormation Templates Before Deployment:**
   Use the `boto3` library to validate your CloudFormation templates before deploying them. This helps catch syntax errors and other issues early.

   ```python
   import boto3

   def validate_template(template_body):
       client = boto3.client('cloudformation')
       response = client.validate_template(TemplateBody=template_body)
       return response

   # Example usage
   template_body = """
   {
       "AWSTemplateFormatVersion": "2010-09-09",
       "Resources": {
           "MyBucket": {
               "Type": "AWS::S3::Bucket"
           }
       }
   }
   """
   validation_response = validate_template(template_body)
   print(validation_response)
   ```

2. **Use Change Sets for Safe Updates:**
   Create and execute change sets to preview the impact of changes before applying them. This helps avoid unexpected failures.

   ```python
   import boto3

   def create_change_set(stack_name, template_body, change_set_name):
       client = boto3.client('cloudformation')
       response = client.create_change_set(
           StackName=stack_name,
           TemplateBody=template_body,
           ChangeSetName=change_set_name,
           ChangeSetType='UPDATE'
       )
       return response

   # Example usage
   stack_name = 'my-stack'
   change_set_name = 'my-change-set'
   change_set_response = create_change_set(stack_name, template_body, change_set_name)
   print(change_set_response)
   ```

3. **Monitor Stack Events During Deployment:**
   Continuously monitor stack events to catch and handle errors as they occur during the stack creation or update process.

   ```python
   import boto3
   import time

   def monitor_stack_events(stack_name):
       client = boto3.client('cloudformation')
       while True:
           response = client.describe_stack_events(StackName=stack_name)
           for event in response['StackEvents']:
               print(f"Resource: {event['LogicalResourceId']}, Status: {event['ResourceStatus']}")
               if 'FAILED' in event['ResourceStatus']:
                   print(f"Error: {event['ResourceStatusReason']}")
                   return
           time.sleep(10)

   # Example usage
   monitor_stack_events(stack_name)
   ```

4. **Implement Rollback Triggers:**
   Use rollback triggers to automatically roll back stack operations if certain CloudWatch alarms are triggered.

   ```python
   import boto3

   def create_stack_with_rollback_triggers(stack_name, template_body, rollback_triggers):
       client = boto3.client('cloudformation')
       response = client.create_stack(
           StackName=stack_name,
           TemplateBody=template_body,
           RollbackConfiguration={
               'RollbackTriggers': rollback_triggers,
               'MonitoringTimeInMinutes': 10
           }
       )
       return response

   # Example usage
   rollback_triggers = [
       {
           'Arn': 'arn:aws:cloudwatch:region:account-id:alarm:alarm-name',
           'Type': 'AWS::CloudWatch::Alarm'
       }
   ]
   create_stack_response = create_stack_with_rollback_triggers(stack_name, template_body, rollback_triggers)
   print(create_stack_response)
   ```

By following these steps, you can significantly reduce the chances of encountering failed statuses in your CloudFormation stacks.
</Accordion>

</AccordionGroup>
</Tab>

<Tab title='Cause'>
### Check Cause
<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
1. Log in to the AWS Management Console and navigate to the CloudFormation service.
2. In the CloudFormation dashboard, under the "Stacks" section, you will see a list of all your CloudFormation stacks.
3. Check the "Status" column for each stack. If a stack has a status of "ROLLBACK_FAILED", "UPDATE_ROLLBACK_FAILED", "DELETE_FAILED", or any other status indicating a failure, it means that the stack has a misconfiguration.
4. For more details about the failure, select the stack with the failed status. In the stack details pane, click on the "Events" tab. This will show you a log of all events related to the stack, including any errors that led to the failed status.
</Accordion>

<Accordion title='Using CLI'>
1. Install and configure AWS CLI: Before you can run any AWS CLI commands, you need to install and configure AWS CLI on your local system. You can do this by following the instructions provided in the AWS CLI User Guide.

2. List all CloudFormation stacks: Use the following AWS CLI command to list all the CloudFormation stacks in your AWS account:

   ```
   aws cloudformation list-stacks --query 'StackSummaries[?StackStatus==`FAILED`].StackName' --output text
   ```
   This command will return the names of all CloudFormation stacks that have a "FAILED" status.

3. Check the status of a specific CloudFormation stack: If you want to check the status of a specific CloudFormation stack, you can use the following AWS CLI command:

   ```
   aws cloudformation describe-stacks --stack-name <your-stack-name> --query 'Stacks[0].StackStatus' --output text
   ```
   Replace `<your-stack-name>` with the name of your CloudFormation stack. This command will return the status of the specified CloudFormation stack.

4. Analyze the output: If the output of the above commands includes any CloudFormation stacks with a "FAILED" status, it means that those stacks have a misconfiguration. You should investigate further to determine the cause of the failure.
</Accordion>

<Accordion title='Using Python'>
1. Install the necessary Python libraries: To interact with AWS services, you need to install the AWS SDK for Python (Boto3). You can install it using pip:

   ```bash
   pip install boto3
   ```

2. Configure AWS Credentials: Before you can begin using Boto3, you need to set up authentication credentials for your AWS account using the AWS CLI. Credentials for your AWS account can be found in the IAM Console. You can create or use an existing user. 

   ```bash
   aws configure
   AWS Access Key ID [None]: YOUR_ACCESS_KEY
   AWS Secret Access Key [None]: YOUR_SECRET_KEY
   Default region name [None]: YOUR_REGION_NAME
   Default output format [None]: json
   ```

3. Write a Python script to list all CloudFormation stacks and check their status:

   ```python
   import boto3

   # Create a client for the AWS service
   client = boto3.client('cloudformation')

   # Call the AWS service to get a list of stacks
   response = client.describe_stacks()

   # Iterate over the stacks and check their status
   for stack in response['Stacks']:
       if stack['StackStatus'] == 'ROLLBACK_FAILED' or stack['StackStatus'] == 'CREATE_FAILED' or stack['StackStatus'] == 'DELETE_FAILED' or stack['StackStatus'] == 'UPDATE_ROLLBACK_FAILED':
           print(f"Stack {stack['StackName']} has a failed status: {stack['StackStatus']}")
   ```

4. Run the Python script: Save the script to a file, for example `check_stack_status.py`, and then run it using Python:

   ```bash
   python check_stack_status.py
   ```

This script will print out the names of all CloudFormation stacks that have a failed status. If no such stacks are found, it will not print anything.
</Accordion>

</AccordionGroup>
</Tab>

<Tab title='Remediation'>
### Remediation

<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
To remediate a CloudFormation stack with a failed status in AWS using the AWS console, follow these steps:

1. Open the AWS Management Console and navigate to the CloudFormation service.
2. Select the CloudFormation stack with a failed status that you want to remediate.
3. Click on the "Events" tab to view the events associated with the stack.
4. Review the events to identify the root cause of the failure. The event details will provide information on the resource that failed and the reason for the failure.
5. Once you have identified the root cause of the failure, take the appropriate action to remediate the issue. This may involve updating the CloudFormation template, modifying the resource configuration, or resolving any dependencies or permissions issues.
6. After making the necessary changes, update the CloudFormation stack by clicking on the "Update Stack" button.
7. Follow the prompts to upload the updated template and apply the changes to the stack.
8. Monitor the stack events to ensure that the update is successful and the stack status changes to "CREATE_COMPLETE" or "UPDATE_COMPLETE".

By following these steps, you can remediate a CloudFormation stack with a failed status in AWS using the AWS console.

#
</Accordion>

<Accordion title='Using CLI'>
To remediate the CloudFormation Stack failed status in AWS using AWS CLI, follow these steps:

1. Identify the CloudFormation Stack that has a failed status by running the following command in the AWS CLI:

```
aws cloudformation describe-stacks --stack-name <stack-name>
```

Replace `<stack-name>` with the name of the CloudFormation Stack that has a failed status.

2. Check the events associated with the failed stack by running the following command:

```
aws cloudformation describe-stack-events --stack-name <stack-name>
```

This will provide information on the events that led to the failed status of the stack.

3. Fix any issues that caused the stack to fail. This may involve updating the CloudFormation template or fixing any resource dependencies.

4. Once the issues are fixed, update the stack by running the following command:

```
aws cloudformation update-stack --stack-name <stack-name> --template-body file://<path-to-template-file> --parameters file://<path-to-parameters-file>
```

Replace `<stack-name>` with the name of the CloudFormation Stack that has a failed status, `<path-to-template-file>` with the path to the updated CloudFormation template file, and `<path-to-parameters-file>` with the path to the updated parameters file.

5. Wait for the stack to update and check its status by running the following command:

```
aws cloudformation describe-stacks --stack-name <stack-name>
```

If the stack status is CREATE_COMPLETE or UPDATE_COMPLETE, then the remediation is successful. If not, repeat steps 2 to 4 until the stack status is successful.
</Accordion>

<Accordion title='Using Python'>
To remediate the "CloudFormation Stacks Should Not Have A Failed Status" misconfiguration in AWS using Python, you can follow the below steps:

1. First, you need to identify the CloudFormation stack(s) that have a failed status. You can use the `boto3` library in Python to get the list of CloudFormation stacks and their status.

```python
import boto3

# Create a CloudFormation client
cf_client = boto3.client('cloudformation')

# Get the list of stacks
stacks = cf_client.list_stacks(StackStatusFilter=['CREATE_FAILED', 'ROLLBACK_COMPLETE'])

# Print the stack names and their status
for stack in stacks['StackSummaries']:
    print(stack['StackName'], stack['StackStatus'])
```

2. Once you have identified the failed stacks, you can delete them using the `delete_stack` method of the CloudFormation client.

```python
# Delete the failed stacks
for stack in stacks['StackSummaries']:
    cf_client.delete_stack(StackName=stack['StackName'])
```

3. You can also update the CloudFormation template and re-create the stack(s) to remediate the misconfiguration.

```python
# Update the CloudFormation template
with open('template.yaml', 'r') as file:
    template_body = file.read()

# Update the stack(s)
for stack in stacks['StackSummaries']:
    cf_client.update_stack(StackName=stack['StackName'], TemplateBody=template_body)
```

Note: Make sure to test the CloudFormation template before updating the stack(s) to avoid any further misconfigurations.
</Accordion>

</AccordionGroup>
</Tab>
</Tabs>
### Additional Reading:

- [https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/troubleshooting.html](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/troubleshooting.html) 

