---
slug: elb_app_tier_health_check
title: Right Health Check Configurations Should Be Used For App-Tier ELBs
sidebar_label: Right Health Check Configurations Should Be Used For App-Tier ELBs
---

### More Info:

Improve the reliability of the applications behind your app-tier ELBs by using the appropriate health check configuration.

### Risk Level

Low

### Address

Reliability

### Compliance Standards

HIPAA



### Triage and Remediation
<Tabs>


<Tab title='Prevention'>
### How to Prevent
<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
To prevent misconfigurations related to health check configurations for App-Tier Elastic Load Balancers (ELBs) using the AWS Management Console, follow these steps:

1. **Navigate to the Load Balancers Section:**
   - Open the AWS Management Console.
   - In the navigation pane, choose "EC2" to open the Amazon EC2 dashboard.
   - In the left-hand menu, under "Load Balancing," choose "Load Balancers."

2. **Select the Target Load Balancer:**
   - From the list of load balancers, select the App-Tier ELB for which you want to configure the health checks.
   - Click on the "Description" tab to view the details of the selected load balancer.

3. **Configure Health Checks:**
   - In the lower pane, select the "Health Checks" tab.
   - Click the "Edit health check" button to modify the health check settings.
   - Ensure that the health check path, protocol, port, and other parameters are correctly configured to match the requirements of your application. For example, set the appropriate HTTP/HTTPS path, response timeout, interval, and thresholds for healthy/unhealthy states.

4. **Save and Verify:**
   - After configuring the health check parameters, click the "Save" button to apply the changes.
   - Monitor the health check status to ensure that the instances behind the ELB are being correctly evaluated and that the health check configuration is functioning as expected.

By following these steps, you can ensure that the health check configurations for your App-Tier ELBs are correctly set up, helping to prevent misconfigurations and maintain the availability and reliability of your application.
</Accordion>

<Accordion title='Using CLI'>
To ensure that the right health check configurations are used for App-Tier Elastic Load Balancers (ELBs) in AWS using the AWS CLI, follow these steps:

1. **Describe the Load Balancer:**
   First, identify the ELB for which you want to configure the health check. Use the following command to list all your load balancers and get the necessary details:
   ```sh
   aws elb describe-load-balancers
   ```

2. **Configure Health Check:**
   Use the `configure-health-check` command to set the appropriate health check parameters for your ELB. Replace `<load-balancer-name>` with your ELB's name and adjust the parameters (`Target`, `Interval`, `Timeout`, `UnhealthyThreshold`, `HealthyThreshold`) as needed.
   ```sh
   aws elb configure-health-check --load-balancer-name <load-balancer-name> --health-check Target=HTTP:80/index.html,Interval=30,Timeout=5,UnhealthyThreshold=2,HealthyThreshold=2
   ```

3. **Verify Health Check Configuration:**
   After configuring the health check, verify the settings to ensure they are correctly applied. Use the following command to describe the health check configuration of your ELB:
   ```sh
   aws elb describe-load-balancers --load-balancer-names <load-balancer-name>
   ```

4. **Monitor Health Check Status:**
   Regularly monitor the health check status of your instances to ensure they are healthy and the configurations are working as expected. Use the following command to describe the instance health:
   ```sh
   aws elb describe-instance-health --load-balancer-name <load-balancer-name>
   ```

By following these steps, you can ensure that the right health check configurations are used for your App-Tier ELBs in AWS using the AWS CLI.
</Accordion>

<Accordion title='Using Python'>
To prevent misconfigurations related to health check configurations for App-Tier Elastic Load Balancers (ELBs) in AWS using Python scripts, you can use the Boto3 library, which is the AWS SDK for Python. Here are the steps to ensure the right health check configurations are used:

1. **Install Boto3 Library**:
   Ensure you have the Boto3 library installed. If not, you can install it using pip:
   ```bash
   pip install boto3
   ```

2. **Set Up AWS Credentials**:
   Make sure your AWS credentials are configured. You can set them up using the AWS CLI or by creating a `~/.aws/credentials` file.

3. **Create a Python Script to Configure Health Checks**:
   Write a Python script to configure the health checks for your ELBs. Below is an example script:

   ```python
   import boto3

   # Initialize a session using Amazon EC2
   session = boto3.Session(
       aws_access_key_id='YOUR_ACCESS_KEY',
       aws_secret_access_key='YOUR_SECRET_KEY',
       region_name='YOUR_REGION'
   )

   # Initialize the ELB client
   elb_client = session.client('elb')

   # Function to configure health check for a given ELB
   def configure_health_check(elb_name, target, interval, timeout, healthy_threshold, unhealthy_threshold):
       response = elb_client.configure_health_check(
           LoadBalancerName=elb_name,
           HealthCheck={
               'Target': target,
               'Interval': interval,
               'Timeout': timeout,
               'UnhealthyThreshold': unhealthy_threshold,
               'HealthyThreshold': healthy_threshold
           }
       )
       return response

   # Example usage
   elb_name = 'your-elb-name'
   target = 'HTTP:80/ping'  # Example target
   interval = 30  # Interval in seconds
   timeout = 5  # Timeout in seconds
   healthy_threshold = 3  # Number of consecutive health check successes
   unhealthy_threshold = 2  # Number of consecutive health check failures

   response = configure_health_check(elb_name, target, interval, timeout, healthy_threshold, unhealthy_threshold)
   print(response)
   ```

4. **Run the Script**:
   Execute the script to apply the health check configurations to your ELB:
   ```bash
   python configure_health_check.py
   ```

This script ensures that the health check configurations for your App-Tier ELBs are set correctly, preventing misconfigurations. Adjust the parameters (`target`, `interval`, `timeout`, `healthy_threshold`, and `unhealthy_threshold`) as per your application's requirements.
</Accordion>

</AccordionGroup>
</Tab>

<Tab title='Cause'>
### Check Cause
<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
1. Log in to the AWS Management Console and navigate to the EC2 dashboard.
2. In the navigation pane, under "Load Balancing", click on "Load Balancers".
3. Select the load balancer you want to check. In the "Description" tab below, find the "Health Check" section.
4. Check the "Ping Target" field. It should be set to the correct protocol (HTTP or HTTPS) and the path should point to a valid URL that returns a 200 response code. The "Timeout" and "Interval" fields should be set to appropriate values. The "Unhealthy Threshold" should be set to a value that allows enough time for the instance to recover before it is taken out of service, and the "Healthy Threshold" should be set to a value that ensures the instance is stable before it is put back into service.
</Accordion>

<Accordion title='Using CLI'>
1. First, you need to install and configure AWS CLI on your local machine. You can download it from the official AWS website and configure it using the "aws configure" command. You will be prompted to provide your AWS Access Key ID, Secret Access Key, Default region name, and Default output format.

2. Once the AWS CLI is set up, you can list all the available Elastic Load Balancers (ELBs) in your account using the following command:

   ```
   aws elbv2 describe-load-balancers
   ```

   This command will return a JSON output with details of all the ELBs.

3. To check the health check configurations of a specific ELB, you need to know its ARN (Amazon Resource Name). You can find the ARN in the output of the previous command. Once you have the ARN, use the following command to get the health check configurations:

   ```
   aws elbv2 describe-target-health --target-group-arn <your-target-group-arn>
   ```

   Replace `<your-target-group-arn>` with the ARN of your ELB. This command will return a JSON output with the health check configurations of the specified ELB.

4. Now, you need to manually inspect the output and check if the right health check configurations are used. The important fields to look for are "HealthCheckPath", "HealthCheckIntervalSeconds", "HealthCheckTimeoutSeconds", and "HealthyThresholdCount". The exact values for these fields depend on your specific requirements, but generally, the health check path should be set to a URL that returns a 200 status code when the application is healthy, the health check interval should be set to a reasonable value to avoid unnecessary health checks, the health check timeout should be less than the health check interval, and the healthy threshold count should be set to a value that avoids flapping.
</Accordion>

<Accordion title='Using Python'>
1. Install the necessary Python libraries: Before you can start writing the script, you need to install the necessary Python libraries. The AWS SDK for Python (Boto3) allows Python developers to write software that makes use of services like Amazon S3, Amazon EC2, etc. You can install it using pip:

   ```bash
   pip install boto3
   ```

2. Configure AWS Credentials: Boto3 needs your AWS credentials (access key and secret access key) to call AWS services on your behalf. These can be configured in several ways, but the simplest is to use the AWS CLI:

   ```bash
   aws configure
   ```

   You'll be prompted to enter your AWS Access Key ID, Secret Access Key, default region name, and default output format.

3. Write the Python script: The following Python script uses Boto3 to retrieve all the ELBs in your AWS account and checks if the right health check configurations are used for App-Tier ELBs.

   ```python
   import boto3

   def check_health_check_configurations():
       elb = boto3.client('elbv2')
       elbs = elb.describe_load_balancers()
       for elb in elbs['LoadBalancers']:
           if 'app-tier' in elb['LoadBalancerName']:
               health_check = elb.describe_target_health(TargetGroupArn=elb['TargetGroups'][0]['TargetGroupArn'])
               if health_check['HealthCheckIntervalSeconds'] != 30 or health_check['HealthCheckPath'] != '/healthcheck':
                   print(f"ELB {elb['LoadBalancerName']} does not have the right health check configurations.")

   if __name__ == "__main__":
       check_health_check_configurations()
   ```

   This script retrieves all the ELBs and checks if the name contains 'app-tier'. If it does, it retrieves the health check configurations and checks if the interval is set to 30 seconds and the path is set to '/healthcheck'. If not, it prints a message indicating that the ELB does not have the right health check configurations.

4. Run the Python script: You can run the Python script using the following command:

   ```bash
   python check_health_check_configurations.py
   ```

   This will print out the names of all the App-Tier ELBs that do not have the right health check configurations.
</Accordion>

</AccordionGroup>
</Tab>

<Tab title='Remediation'>
### Remediation

<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
To remediate the misconfiguration "Right Health Check Configurations Should Be Used For App-Tier ELBs" for AWS using AWS console, follow these steps:

1. Open the AWS Management Console and navigate to the EC2 dashboard.
2. Click on the "Load Balancers" option in the left-hand menu.
3. Select the Application Load Balancer that you want to remediate.
4. In the "Listeners" tab, click on the "View/edit rules" button for the listener that you want to remediate.
5. In the "Rules" section, click on the "Edit" button for the rule that you want to remediate.
6. In the "Edit Rule" dialog box, click on the "Add Condition" button.
7. In the "Add Condition" dialog box, select "Health Check" from the drop-down menu and configure the health check settings as per your requirements.
8. Click on the "Save" button to save the changes.
9. Repeat steps 5-8 for all the rules that you want to remediate.
10. Once you have remediated all the rules, click on the "Save" button in the "Listeners" tab to save the changes.

By following these steps, you can remediate the misconfiguration "Right Health Check Configurations Should Be Used For App-Tier ELBs" for AWS using AWS console.

#
</Accordion>

<Accordion title='Using CLI'>
To remediate the misconfiguration "Right Health Check Configurations Should Be Used For App-Tier ELBs" for AWS using AWS CLI, follow these steps:

1. Open the AWS CLI and run the following command to describe the load balancer:

```
aws elb describe-load-balancers --load-balancer-name <load_balancer_name>
```

Note: Replace `<load_balancer_name>` with the name of your load balancer.

2. Check the health check configuration of the load balancer. The health check configuration includes the ping target, port, and timeout settings. To view the health check configuration, run the following command:

```
aws elb describe-instance-health --load-balancer-name <load_balancer_name>
```

3. If the health check configuration is incorrect, update it by running the following command:

```
aws elb configure-health-check --load-balancer-name <load_balancer_name> --health-check Target=HTTP:80/index.html,Interval=30,UnhealthyThreshold=2,HealthyThreshold=2,Timeout=3
```

Note: Replace `<load_balancer_name>` with the name of your load balancer. You can modify the health check configuration as per your requirement.

4. After updating the health check configuration, verify it by running the following command:

```
aws elb describe-instance-health --load-balancer-name <load_balancer_name>
```

This will ensure that the right health check configurations are used for the app-tier ELBs in AWS.
</Accordion>

<Accordion title='Using Python'>
To remediate the misconfiguration "Right Health Check Configurations Should Be Used For App-Tier ELBs" in AWS using Python, you can follow these steps:

1. Import the necessary libraries: boto3 and json.

```
import boto3
import json
```

2. Create a connection to the AWS resource using the boto3 library.

```
elbv2 = boto3.client('elbv2')
```

3. Get all the load balancers in the region.

```
response = elbv2.describe_load_balancers()
```

4. Loop through the load balancers and check if they are application load balancers.

```
for lb in response['LoadBalancers']:
    if lb['Type'] == 'application':
```

5. If the load balancer is an application load balancer, get its ARN.

```
arn = lb['LoadBalancerArn']
```

6. Get the current health check configuration for the load balancer.

```
health_check = elbv2.describe_target_group_health(TargetGroupArn=arn)
```

7. Check if the health check configuration is correct.

```
if health_check['HealthChecks'][0]['IntervalSeconds'] != 30 or health_check['HealthChecks'][0]['TimeoutSeconds'] != 5 or health_check['HealthChecks'][0]['HealthyThresholdCount'] != 5 or health_check['HealthChecks'][0]['UnhealthyThresholdCount'] != 2:
```

8. If the health check configuration is incorrect, update it to the correct configuration.

```
elbv2.modify_target_group_attributes(
    TargetGroupArn=arn,
    Attributes=[
        {
            'Key': 'deregistration_delay.timeout_seconds',
            'Value': '30'
        },
        {
            'Key': 'health_check.interval_seconds',
            'Value': '30'
        },
        {
            'Key': 'health_check.timeout_seconds',
            'Value': '5'
        },
        {
            'Key': 'health_check.healthy_threshold_count',
            'Value': '5'
        },
        {
            'Key': 'health_check.unhealthy_threshold_count',
            'Value': '2'
        },
    ]
)
```

9. Print a message to confirm that the health check configuration has been updated.

```
print(f"Health check configuration updated for load balancer {lb['LoadBalancerName']}")
```

10. The final code will look like this:

```
import boto3
import json

elbv2 = boto3.client('elbv2')

response = elbv2.describe_load_balancers()

for lb in response['LoadBalancers']:
    if lb['Type'] == 'application':
        arn = lb['LoadBalancerArn']
        health_check = elbv2.describe_target_group_health(TargetGroupArn=arn)

        if health_check['HealthChecks'][0]['IntervalSeconds'] != 30 or health_check['HealthChecks'][0]['TimeoutSeconds'] != 5 or health_check['HealthChecks'][0]['HealthyThresholdCount'] != 5 or health_check['HealthChecks'][0]['UnhealthyThresholdCount'] != 2:
            elbv2.modify_target_group_attributes(
                TargetGroupArn=arn,
                Attributes=[
                    {
                        'Key': 'deregistration_delay.timeout_seconds',
                        'Value': '30'
                    },
                    {
                        'Key': 'health_check.interval_seconds',
                        'Value': '30'
                    },
                    {
                        'Key': 'health_check.timeout_seconds',
                        'Value': '5'
                    },
                    {
                        'Key': 'health_check.healthy_threshold_count',
                        'Value': '5'
                    },
                    {
                        'Key': 'health_check.unhealthy_threshold_count',
                        'Value': '2'
                    },
                ]
            )
            print(f"Health check configuration updated for load balancer {lb['LoadBalancerName']}")
```
</Accordion>

</AccordionGroup>
</Tab>
</Tabs>
### Additional Reading:

- [https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/elb-healthchecks.html](https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/elb-healthchecks.html) 

