---
slug: elb_app_tier_health_check
title: Right Health Check Configurations Should Be Used For App-Tier ELBs
sidebar_label: Right Health Check Configurations Should Be Used For App-Tier ELBs
---

### More Info:

Improve the reliability of the applications behind your app-tier ELBs by using the appropriate health check configuration.

### Risk Level

Low

### Address

Reliability

### Compliance Standards

HIPAA




<Tabs><Tab title='Cause'>
### Check Cause

#### Using Console

1. Log in to the AWS Management Console and navigate to the EC2 dashboard.
2. In the navigation pane, under LOAD BALANCING, click on Load Balancers.
3. Select the Elastic Load Balancer (ELB) that you want to check.
4. In the bottom panel, click on the Health Check tab. Here, you can see the health check configurations for the selected ELB. Check if the configurations are set correctly according to your application's requirements.

#### Using CLI

1. First, you need to install and configure AWS CLI on your local machine. You can do this by following the instructions provided by AWS. Make sure you have the necessary permissions to access the Elastic Load Balancer (ELB) resources.

2. Once the AWS CLI is set up, you can list all the available load balancers in your account using the following command:

   ```
   aws elbv2 describe-load-balancers
   ```

   This command will return a JSON output with details about all the load balancers.

3. To check the health check configurations for a specific load balancer, you need to know the ARN (Amazon Resource Name) of the load balancer. You can find this in the output of the previous command. Once you have the ARN, use the following command to get the health check configurations:

   ```
   aws elbv2 describe-target-health --target-group-arn <target-group-arn>
   ```

   Replace `<target-group-arn>` with the ARN of your target group. This command will return a JSON output with details about the health checks for the specified load balancer.

4. To check if the right health check configurations are used, you need to analyze the output of the previous command. Look for the `HealthCheck` field in the output. This field contains the configurations for the health checks. The right configurations depend on your specific use case, but generally, you should check for the following:

   - `HealthCheckIntervalSeconds`: This should be set to a reasonable value to ensure that health checks are performed frequently enough to detect any issues.
   - `HealthCheckPath`: This should be set to the path of a resource that can be used to check the health of the application.
   - `HealthyThresholdCount` and `UnhealthyThresholdCount`: These should be set to values that ensure that the load balancer doesn't switch to a different target too quickly or too slowly in case of a failure.

#### Using Python

1. Install the necessary Python libraries: Before you can start writing the script, you need to install the necessary Python libraries. The AWS SDK for Python (Boto3) allows Python developers to write software that makes use of services like Amazon S3, Amazon EC2, etc. You can install it using pip:

```python
pip install boto3
```

2. Set up AWS credentials: You need to configure your AWS credentials. You can do this in several ways, but the simplest is to use the AWS CLI. Run `aws configure` and then enter your access key, secret access key, and default region when prompted.

3. Write the Python script: The following Python script uses Boto3 to check the health check configurations for all ELBs in the specified region.

```python
import boto3

def check_health_check_configurations():
    client = boto3.client('elbv2')
    elbs = client.describe_load_balancers()
    for elb in elbs['LoadBalancers']:
        response = client.describe_target_health(
            TargetGroupArn=elb['TargetGroupArn']
        )
        for health_desc in response['TargetHealthDescriptions']:
            if health_desc['TargetHealth']['State'] != 'healthy':
                print(f"ELB {elb['LoadBalancerName']} has unhealthy target: {health_desc['Target']['Id']}")

check_health_check_configurations()
```

4. Run the script: You can run the script using any Python environment. The script will print the names of any ELBs that have unhealthy targets. If no such ELBs are found, the script will not print anything. This can help you identify any misconfigurations that might be causing health check failures.

Remember to replace `'elbv2'` with the appropriate service identifier if you're not using Application Load Balancers. For Classic Load Balancers, use `'elb'`, and for Network Load Balancers, use `'elbv2'`.

</Tab>


<Tab title='Remediation'>
### Remediation

#### Using Console

To remediate the misconfiguration "Right Health Check Configurations Should Be Used For App-Tier ELBs" for AWS using AWS console, follow these steps:

1. Open the AWS Management Console and navigate to the EC2 dashboard.
2. Click on the "Load Balancers" option in the left-hand menu.
3. Select the Application Load Balancer that you want to remediate.
4. In the "Listeners" tab, click on the "View/edit rules" button for the listener that you want to remediate.
5. In the "Rules" section, click on the "Edit" button for the rule that you want to remediate.
6. In the "Edit Rule" dialog box, click on the "Add Condition" button.
7. In the "Add Condition" dialog box, select "Health Check" from the drop-down menu and configure the health check settings as per your requirements.
8. Click on the "Save" button to save the changes.
9. Repeat steps 5-8 for all the rules that you want to remediate.
10. Once you have remediated all the rules, click on the "Save" button in the "Listeners" tab to save the changes.

By following these steps, you can remediate the misconfiguration "Right Health Check Configurations Should Be Used For App-Tier ELBs" for AWS using AWS console.

#### Using CLI

To remediate the misconfiguration "Right Health Check Configurations Should Be Used For App-Tier ELBs" for AWS using AWS CLI, follow these steps:

1. Open the AWS CLI and run the following command to describe the load balancer:

```
aws elb describe-load-balancers --load-balancer-name <load_balancer_name>
```

Note: Replace `<load_balancer_name>` with the name of your load balancer.

2. Check the health check configuration of the load balancer. The health check configuration includes the ping target, port, and timeout settings. To view the health check configuration, run the following command:

```
aws elb describe-instance-health --load-balancer-name <load_balancer_name>
```

3. If the health check configuration is incorrect, update it by running the following command:

```
aws elb configure-health-check --load-balancer-name <load_balancer_name> --health-check Target=HTTP:80/index.html,Interval=30,UnhealthyThreshold=2,HealthyThreshold=2,Timeout=3
```

Note: Replace `<load_balancer_name>` with the name of your load balancer. You can modify the health check configuration as per your requirement.

4. After updating the health check configuration, verify it by running the following command:

```
aws elb describe-instance-health --load-balancer-name <load_balancer_name>
```

This will ensure that the right health check configurations are used for the app-tier ELBs in AWS.

#### Using Python

To remediate the misconfiguration "Right Health Check Configurations Should Be Used For App-Tier ELBs" in AWS using Python, you can follow these steps:

1. Import the necessary libraries: boto3 and json.

```
import boto3
import json
```

2. Create a connection to the AWS resource using the boto3 library.

```
elbv2 = boto3.client('elbv2')
```

3. Get all the load balancers in the region.

```
response = elbv2.describe_load_balancers()
```

4. Loop through the load balancers and check if they are application load balancers.

```
for lb in response['LoadBalancers']:
    if lb['Type'] == 'application':
```

5. If the load balancer is an application load balancer, get its ARN.

```
arn = lb['LoadBalancerArn']
```

6. Get the current health check configuration for the load balancer.

```
health_check = elbv2.describe_target_group_health(TargetGroupArn=arn)
```

7. Check if the health check configuration is correct.

```
if health_check['HealthChecks'][0]['IntervalSeconds'] != 30 or health_check['HealthChecks'][0]['TimeoutSeconds'] != 5 or health_check['HealthChecks'][0]['HealthyThresholdCount'] != 5 or health_check['HealthChecks'][0]['UnhealthyThresholdCount'] != 2:
```

8. If the health check configuration is incorrect, update it to the correct configuration.

```
elbv2.modify_target_group_attributes(
    TargetGroupArn=arn,
    Attributes=[
        {
            'Key': 'deregistration_delay.timeout_seconds',
            'Value': '30'
        },
        {
            'Key': 'health_check.interval_seconds',
            'Value': '30'
        },
        {
            'Key': 'health_check.timeout_seconds',
            'Value': '5'
        },
        {
            'Key': 'health_check.healthy_threshold_count',
            'Value': '5'
        },
        {
            'Key': 'health_check.unhealthy_threshold_count',
            'Value': '2'
        },
    ]
)
```

9. Print a message to confirm that the health check configuration has been updated.

```
print(f"Health check configuration updated for load balancer {lb['LoadBalancerName']}")
```

10. The final code will look like this:

```
import boto3
import json

elbv2 = boto3.client('elbv2')

response = elbv2.describe_load_balancers()

for lb in response['LoadBalancers']:
    if lb['Type'] == 'application':
        arn = lb['LoadBalancerArn']
        health_check = elbv2.describe_target_group_health(TargetGroupArn=arn)

        if health_check['HealthChecks'][0]['IntervalSeconds'] != 30 or health_check['HealthChecks'][0]['TimeoutSeconds'] != 5 or health_check['HealthChecks'][0]['HealthyThresholdCount'] != 5 or health_check['HealthChecks'][0]['UnhealthyThresholdCount'] != 2:
            elbv2.modify_target_group_attributes(
                TargetGroupArn=arn,
                Attributes=[
                    {
                        'Key': 'deregistration_delay.timeout_seconds',
                        'Value': '30'
                    },
                    {
                        'Key': 'health_check.interval_seconds',
                        'Value': '30'
                    },
                    {
                        'Key': 'health_check.timeout_seconds',
                        'Value': '5'
                    },
                    {
                        'Key': 'health_check.healthy_threshold_count',
                        'Value': '5'
                    },
                    {
                        'Key': 'health_check.unhealthy_threshold_count',
                        'Value': '2'
                    },
                ]
            )
            print(f"Health check configuration updated for load balancer {lb['LoadBalancerName']}")
```


### Additional Reading:

- [https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/elb-healthchecks.html](https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/elb-healthchecks.html) 


</Tab>
</Tabs>