---
slug: fms_webacl_resource_policy
title: FMS Policy Owner Specifies WebACLId
sidebar_label: FMS Policy Owner Specifies WebACLId
---

### More Info:

Checks if the web ACL is associated with an Application Load Balancer, API Gateway stage, or Amazon CloudFront distributions. When AWS Firewall Manager creates this rule, the FMS policy owner specifies the WebACLId in the FMS policy and can optionally enable remediation.

### Risk Level

Medium

### Address

Security

### Compliance Standards

CBP



<Tabs><Tab title='Cause'>
### Check Cause

#### Using Console

1. Log in to the AWS Management Console and open the AWS Firewall Manager service.

2. In the Firewall Manager dashboard, select the "Security policies" option from the left-hand side menu.

3. In the Security Policies page, you will see a list of all the Firewall Manager policies. Look for the policy that is associated with your Elastic Load Balancer.

4. Click on the policy name to open its details. In the policy details, check the "AWS WAF web ACL" field. If the field is specified with a WebACLId, then the FMS Policy Owner has specified the WebACLId in the Elastic Load Balancer. If the field is empty or not specified, then the FMS Policy Owner has not specified the WebACLId in the Elastic Load Balancer.

#### Using CLI

1. First, you need to list all the Elastic Load Balancers (ELBs) in your AWS environment. You can do this by using the AWS CLI command:

```bash
aws elbv2 describe-load-balancers
```

This command will return a JSON output with details of all the ELBs in your environment.

2. Next, you need to list all the Firewall Manager (FMS) policies in your AWS environment. You can do this by using the AWS CLI command:

```bash
aws fms list-policies
```

This command will return a JSON output with details of all the FMS policies in your environment.

3. Now, you need to check if the WebACLId is specified in the FMS policy for each ELB. You can do this by using the AWS CLI command:

```bash
aws fms get-policy --policy-id <PolicyId>
```

Replace `<PolicyId>` with the ID of the FMS policy you want to check. This command will return a JSON output with the details of the FMS policy.

4. Finally, you need to parse the JSON output from the previous step to check if the WebACLId is specified. You can do this by using a tool like `jq`:

```bash
aws fms get-policy --policy-id <PolicyId> | jq '.Policy.WebACLId'
```

If the WebACLId is specified, this command will return the ID of the WebACL. If the WebACLId is not specified, this command will return `null`.

#### Using Python

To check if the Firewall Manager (FMS) policy owner specifies the WebACLId in the Elastic Load Balancer (ELB), you can use the AWS SDK for Python (Boto3). Here are the steps:

1. **Setup AWS SDK for Python (Boto3):**
   First, you need to install and configure Boto3. You can install it using pip:

   ```bash
   pip install boto3
   ```

   Then, configure your AWS credentials. You can do this by setting the following environment variables:

   ```bash
   export AWS_ACCESS_KEY_ID='your_access_key'
   export AWS_SECRET_ACCESS_KEY='your_secret_key'
   ```

2. **Get the list of all FMS policies:**
   Use the `list_policies` method to get all the FMS policies. Here is a sample script:

   ```python
   import boto3

   client = boto3.client('fms')

   response = client.list_policies()

   for policy in response['PolicyList']:
       print(policy['PolicyId'])
   ```

3. **Get the details of each FMS policy:**
   Use the `get_policy` method to get the details of each FMS policy. Here is a sample script:

   ```python
   import boto3

   client = boto3.client('fms')

   response = client.get_policy(
       PolicyId='policy_id'  # replace with your policy id
   )

   print(response['Policy'])
   ```

4. **Check if the WebACLId is specified in the ELB:**
   In the policy details, check if the `WebACLId` is specified in the `SecurityServicePolicyData` field. Here is a sample script:

   ```python
   import boto3

   client = boto3.client('fms')

   response = client.get_policy(
       PolicyId='policy_id'  # replace with your policy id
   )

   policy_data = response['Policy']['SecurityServicePolicyData']

   if 'WebACLId' in policy_data:
       print('WebACLId is specified.')
   else:
       print('WebACLId is not specified.')
   ```

This script will print 'WebACLId is specified.' if the WebACLId is specified in the ELB, otherwise it will print 'WebACLId is not specified.'

</Tab>


<Tab title='Remediation'>
### Remediation

#### Using Console

To remediate the misconfiguration where FMS Policy Owner Specifies WebACLId for an AWS Elastic Load Balancer using the AWS console, follow these step-by-step instructions:

1. **Login to AWS Console**: Go to the AWS Management Console (https://aws.amazon.com/console/) and login with your credentials.

2. **Navigate to AWS WAF & Shield**: In the AWS Management Console, navigate to the AWS WAF & Shield service by typing "WAF" in the search bar and selecting the "AWS WAF & Shield" service.

3. **Select the Web ACL**: In the AWS WAF & Shield dashboard, select the Web ACL that is associated with the FMS Policy Owner Specifies WebACLId misconfiguration.

4. **Remove Web ACL from FMS Policy**: Find the option to remove the Web ACL from the FMS Policy Owner in the Web ACL settings. This will ensure that the FMS Policy Owner no longer specifies the WebACLId for the Elastic Load Balancer.

5. **Save Changes**: Once you have removed the Web ACL from the FMS Policy Owner, save the changes to update the configuration.

6. **Verify Configuration**: Verify that the Web ACL is no longer specified by the FMS Policy Owner for the Elastic Load Balancer. You can do this by checking the configuration settings or testing the Elastic Load Balancer to ensure it is functioning correctly.

By following these steps, you should be able to successfully remediate the misconfiguration where the FMS Policy Owner Specifies WebACLId for an AWS Elastic Load Balancer using the AWS console.

#### Using CLI

To remediate the misconfiguration where the FMS Policy Owner Specifies WebACLId for an AWS Elastic Load Balancer using AWS CLI, you can follow these steps:

1. **Identify the FMS Policy Owner**: First, identify the FMS Policy Owner who has specified the WebACLId for the Elastic Load Balancer.

2. **Revoke Permissions**: The FMS Policy Owner needs to revoke the permission to specify the WebACLId for the Elastic Load Balancer. This can be done by updating the FMS policy or removing the specific permission from the IAM policy attached to the FMS Policy Owner.

3. **Update WebACLId for the Elastic Load Balancer**: If the WebACLId specified by the FMS Policy Owner is incorrect or causing issues, you can update the WebACLId for the Elastic Load Balancer. Use the following AWS CLI command to update the WebACLId for the Elastic Load Balancer:
   
   ```bash
   aws elbv2 modify-load-balancer-attributes --load-balancer-arn <your-load-balancer-arn> --attributes Key=aws.wafv2.webacl,Value=<new-webacl-id>
   ```

   Replace `<your-load-balancer-arn>` with the ARN of your Elastic Load Balancer and `<new-webacl-id>` with the correct WebACLId that you want to associate with the Elastic Load Balancer.

4. **Verify the Changes**: After updating the WebACLId for the Elastic Load Balancer, verify the changes by checking the attributes of the Elastic Load Balancer using the following AWS CLI command:
   
   ```bash
   aws elbv2 describe-load-balancer-attributes --load-balancer-arn <your-load-balancer-arn>
   ```

   Ensure that the correct WebACLId is now associated with the Elastic Load Balancer.

By following these steps, you can remediate the misconfiguration where the FMS Policy Owner Specifies WebACLId for an AWS Elastic Load Balancer using AWS CLI.

#### Using Python

To remediate the misconfiguration where the FMS Policy Owner Specifies WebACLId for AWS Elastic Load Balancer using Python, you can follow these steps:

1. **Identify the Misconfiguration**: Check the AWS WAF WebACL associated with the Elastic Load Balancer to ensure that the FMS Policy Owner does not specify a specific WebACLId.

2. **Update the WebACL**: Use the AWS SDK for Python (Boto3) to update the WebACL associated with the Elastic Load Balancer. You can remove the FMS Policy Owner's specified WebACLId and revert to the default WebACL or a more appropriate one.

3. **Install Boto3**: If you haven't already, install the Boto3 library by running the following command:
   ```
   pip install boto3
   ```

4. **Python Script to Remediate**:
   Here is a sample Python script that demonstrates how to update the WebACL associated with an Elastic Load Balancer:
   ```python
   import boto3

   elb_client = boto3.client('elbv2')

   # Specify the ARN of the Elastic Load Balancer
   elb_arn = 'YOUR_ELB_ARN_HERE'

   # Specify the ARN of the desired WebACL
   web_acl_arn = 'YOUR_WEBACL_ARN_HERE'

   # Update the WebACL associated with the Elastic Load Balancer
   response = elb_client.modify_load_balancer_attributes(
       LoadBalancerArn=elb_arn,
       Attributes=[
           {
               'Key': 'aws:waf:webacl',
               'Value': web_acl_arn
           }
       ]
   )

   print('WebACL updated successfully.')
   ```

5. **Replace the placeholders**:
   - Replace `YOUR_ELB_ARN_HERE` with the ARN of your Elastic Load Balancer.
   - Replace `YOUR_WEBACL_ARN_HERE` with the ARN of the desired WebACL that you want to associate with the Elastic Load Balancer.

6. **Run the Script**: Execute the Python script to update the WebACL associated with the Elastic Load Balancer. This will remediate the misconfiguration where the FMS Policy Owner Specifies WebACLId.

By following these steps and running the Python script, you can successfully remediate the misconfiguration for an AWS Elastic Load Balancer where the FMS Policy Owner Specifies WebACLId.


</Tab>
</Tabs>