---
slug: pubsublite_topic_retention_set
title: PubSub Lite Topic Retention Should Be Set
sidebar_label: PubSub Lite Topic Retention Should Be Set
---

### More Info:

Ensure that topic retention is set for PubSub Lite

### Risk Level

Low

### Address

Operational Maturity, Reliability

### Compliance Standards

CBP


<Tabs><Tab title='Cause'>
### Check Cause

#### Using Console

1. Log in to the Google Cloud Platform Console: Open your web browser, navigate to the Google Cloud Platform web console and sign in with your Google account that has the necessary permissions to access and manage your GCP resources.

2. Navigate to Pub/Sub Lite: From the main dashboard of the Google Cloud Console, click on the navigation menu (three horizontal lines) located at the top left corner of the page. Scroll down and click on "Pub/Sub Lite" under the "Big Data" section.

3. Check Topic Retention: In the Pub/Sub Lite page, you will see a list of all your Pub/Sub Lite topics. Click on the name of the topic you want to check. This will open the topic details page. Here, look for the "Retention" section. This section will show you the current retention settings for the topic.

4. Verify Retention Settings: If the retention is set to "default" or if there is no specific retention period set, then it means that the Pub/Sub Lite Topic Retention is not set. If a specific retention period is mentioned, then the Pub/Sub Lite Topic Retention is set.

#### Using CLI

1. Install and configure the Google Cloud SDK CLI if you haven't done so already. You can do this by following the instructions provided by Google Cloud at https://cloud.google.com/sdk/docs/install.

2. Once the SDK is installed and configured, open your terminal and use the following command to list all the PubSub Lite topics in your project:

   ```
   gcloud pubsub lite-topics list --location=<location>
   ```
   Replace `<location>` with the location of your PubSub Lite topics.

3. For each topic listed, use the following command to get the details of the topic:

   ```
   gcloud pubsub lite-topics describe <topic-name> --location=<location>
   ```
   Replace `<topic-name>` with the name of the topic and `<location>` with the location of the topic.

4. In the output of the above command, look for the `perPartitionBytes` field under `retentionConfig`. This field indicates the retention of the topic. If this field is not set or is set to a low value, it indicates that the PubSub Lite Topic Retention is not properly set.

#### Using Python

To check PubSub Lite Topic Retention in Google Cloud Platform (GCP) using Python scripts, you can use the Google Cloud Pub/Sub Lite API. Here are the steps:

1. **Set up Google Cloud SDK and Python Environment:**
   Before you start, make sure you have the Google Cloud SDK installed and initialized, and also have Python and pip (Python package installer) installed on your machine. You can install the Google Cloud Pub/Sub Lite Python client library using pip:

   ```bash
   pip install google-cloud-pubsublite
   ```

2. **Import Required Libraries:**
   Import the necessary libraries in your Python script:

   ```python
   from google.cloud.pubsublite import AdminClient
   from google.cloud.pubsublite.types import CloudRegion, CloudZone, TopicPath
   ```

3. **Initialize the PubSub Lite Admin Client:**
   Initialize the PubSub Lite Admin Client with the appropriate region and project ID:

   ```python
   region = "your-region"
   project_number = "your-project-number"
   client = AdminClient(CloudRegion.of(region))
   ```

4. **Check Topic Retention Settings:**
   Use the `get_topic` method to retrieve the topic details and check the retention settings:

   ```python
   topic_path = TopicPath(project_number, CloudZone(CloudRegion.of(region), "zone-id"), "topic-id")
   topic = client.get_topic(topic_path)
   print(f"Topic retention settings: {topic.retention_config}")
   ```

This script will print the retention settings for the specified PubSub Lite topic. If the retention settings are not set or are set incorrectly, you have detected a misconfiguration.

</Tab>


<Tab title='Remediation'>
### Remediation

#### Using Console

To remediate the misconfiguration of PubSub Lite Topic Retention using GCP console, you can follow the below steps:

1. Login to the Google Cloud Console (https://console.cloud.google.com/).
2. Select the project that contains the PubSub Lite topic that needs to be remediated.
3. Navigate to the PubSub Lite Topics page by selecting "Pub/Sub Lite" from the left-hand menu and then selecting "Topics".
4. Select the PubSub Lite topic that needs to be remediated.
5. Click on the "Edit" button at the top of the page.
6. Under the "Retention" section, set the retention period for the topic. You can set the retention period to any value between 10 minutes and 7 days.
7. Click on the "Save" button to save the changes.

After following these steps, the PubSub Lite Topic Retention will be set for the selected topic.

#### Using CLI

To remediate the misconfiguration of PubSub Lite Topic Retention for GCP using GCP CLI, follow these steps:

1. Open the GCP Cloud Shell in your GCP console.

2. Run the following command to set the retention period for a PubSub Lite topic:

   ```
   gcloud pubsub lite-topics update TOPIC-ID --retention-period PERIOD
   ```

   Replace TOPIC-ID with the ID of the PubSub Lite topic that you want to update, and replace PERIOD with the number of seconds that you want to retain messages in the topic.

   For example, to set the retention period to 7 days (or 604800 seconds), run the following command:

   ```
   gcloud pubsub lite-topics update my-topic --retention-period 604800
   ```

3. Verify that the retention period has been set correctly by running the following command:

   ```
   gcloud pubsub lite-topics describe TOPIC-ID
   ```

   Replace TOPIC-ID with the ID of the PubSub Lite topic that you updated. The output should include a "retentionConfig" section that shows the retention period that you set.

   For example, the output might look like this:

   ```
   name: projects/my-project/locations/us-central1-a/topics/my-topic
   partitionConfig:
     count: 1
     ...
   retentionConfig:
     perPartitionBytes: 0
     period: 604800s
   state: READY
   ```

   This indicates that the retention period for the topic has been set to 7 days.

4. Repeat these steps for any other PubSub Lite topics that need to have their retention period set.

By following these steps, you can remediate the misconfiguration of PubSub Lite Topic Retention for GCP using GCP CLI.

#### Using Python

To remediate the PubSub Lite Topic Retention misconfiguration in GCP using Python, you can follow the below steps:

1. Install the Google Cloud Pub/Sub Lite client library for Python using pip:

```
pip install google-cloud-pubsublite
```

2. Import the required libraries:

```python
from google.api_core.exceptions import NotFound
from google.cloud.pubsublite import AdminClient, SubscriptionPath, TopicPath
from google.cloud.pubsublite.types import Duration
```

3. Set the project ID and location of the Pub/Sub Lite topic:

```python
project_id = "your-project-id"
location = "your-location"
topic_id = "your-topic-id"
```

4. Create an instance of the `AdminClient` class:

```python
admin_client = AdminClient()
```

5. Use the `TopicPath` class to create a `topic_path` object:

```python
topic_path = TopicPath(project_id, location, topic_id)
```

6. Use a try-except block to check if the topic exists:

```python
try:
    topic = admin_client.get_topic(topic_path)
except NotFound:
    print(f"Topic '{topic_path}' not found.")
    return
```

7. Set the retention duration for the topic using the `update_topic` method:

```python
retention_duration = Duration(seconds=604800) # 1 week in seconds
topic.retention_config.period = retention_duration
admin_client.update_topic(topic)
```

8. Verify that the retention duration has been set:

```python
print(f"Retention duration set to {topic.retention_config.period.seconds} seconds.")
```

The complete code to remediate the PubSub Lite Topic Retention misconfiguration in GCP using Python is as follows:

```python
from google.api_core.exceptions import NotFound
from google.cloud.pubsublite import AdminClient, SubscriptionPath, TopicPath
from google.cloud.pubsublite.types import Duration

project_id = "your-project-id"
location = "your-location"
topic_id = "your-topic-id"

admin_client = AdminClient()
topic_path = TopicPath(project_id, location, topic_id)

try:
    topic = admin_client.get_topic(topic_path)
except NotFound:
    print(f"Topic '{topic_path}' not found.")
    return

retention_duration = Duration(seconds=604800) # 1 week in seconds
topic.retention_config.period = retention_duration
admin_client.update_topic(topic)

print(f"Retention duration set to {topic.retention_config.period.seconds} seconds.")
```




</Tab>
</Tabs>