---
slug: dns_compliant_s3_bucket_names
title: S3 Bucket Names Should Be DNS-compliant
sidebar_label: S3 Bucket Names Should Be DNS-compliant
---

### More Info:

S3 buckets should use DNS-compliant bucket names in order to adhere to AWS best practices and to benefit from the new S3 features such as S3 Transfer Acceleration, to benefit from operational improvements and to receive support for virtual-host style access to buckets.

### Risk Level

Low

### Address

Operational Maturity, Security

### Compliance Standards

CBP


### Triage and Remediation
<Tabs>


<Tab title='Prevention'>
### How to Prevent
<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
To ensure that S3 bucket names are DNS-compliant using the AWS Management Console, follow these steps:

1. **Navigate to S3 Service:**
   - Open the AWS Management Console.
   - In the search bar at the top, type "S3" and select "S3" from the dropdown list to navigate to the Amazon S3 dashboard.

2. **Create a New Bucket:**
   - Click on the "Create bucket" button.
   - In the "Bucket name" field, enter a name that adheres to DNS-compliant naming conventions. Ensure the name:
     - Is between 3 and 63 characters long.
     - Only contains lowercase letters, numbers, hyphens, and periods.
     - Does not start or end with a period or hyphen.
     - Does not contain consecutive periods or hyphens.

3. **Review and Confirm:**
   - Continue through the bucket creation steps, configuring settings as needed.
   - Review the settings and ensure the bucket name is DNS-compliant before finalizing the creation.

4. **Verify Existing Buckets:**
   - For existing buckets, navigate to the S3 dashboard and review the list of bucket names.
   - Ensure all existing bucket names follow DNS-compliant naming conventions. If any bucket names are not compliant, consider renaming them (note that renaming a bucket requires creating a new bucket and migrating data).

By following these steps, you can ensure that your S3 bucket names are DNS-compliant using the AWS Management Console.
</Accordion>

<Accordion title='Using CLI'>
To ensure that S3 bucket names are DNS-compliant using the AWS CLI, you can follow these steps:

1. **Understand DNS-compliant Naming Rules**:
   - Bucket names must be between 3 and 63 characters long.
   - Bucket names can contain lowercase letters, numbers, hyphens, and periods.
   - Bucket names must start and end with a lowercase letter or number.
   - Bucket names cannot be formatted as IP addresses (e.g., 192.168.1.1).

2. **Create a DNS-compliant Bucket**:
   Use the `aws s3api create-bucket` command to create a bucket with a DNS-compliant name.
   ```sh
   aws s3api create-bucket --bucket my-dns-compliant-bucket --region us-east-1
   ```

3. **Validate Bucket Name Before Creation**:
   Before creating a bucket, you can use a script to validate the bucket name against DNS-compliant rules. Here is an example using a simple shell script:
   ```sh
   BUCKET_NAME="my-dns-compliant-bucket"
   if [[ $BUCKET_NAME =~ ^[a-z0-9][a-z0-9.-]{1,61}[a-z0-9]$ ]] && ! [[ $BUCKET_NAME =~ ^[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
       aws s3api create-bucket --bucket $BUCKET_NAME --region us-east-1
   else
       echo "Bucket name is not DNS-compliant."
   fi
   ```

4. **Use AWS CLI to List and Verify Existing Buckets**:
   To ensure existing buckets are DNS-compliant, list all buckets and manually verify their names.
   ```sh
   aws s3api list-buckets --query "Buckets[].Name"
   ```

By following these steps, you can ensure that your S3 bucket names are DNS-compliant using the AWS CLI.
</Accordion>

<Accordion title='Using Python'>
To ensure that S3 bucket names are DNS-compliant using Python scripts, you can follow these steps:

1. **Install Boto3 Library**:
   Ensure you have the Boto3 library installed, which is the AWS SDK for Python. You can install it using pip if you haven't already.

   ```bash
   pip install boto3
   ```

2. **Define DNS-compliant Bucket Name Function**:
   Create a function to validate that the bucket name is DNS-compliant. AWS S3 bucket names must follow specific rules to be DNS-compliant, such as being between 3 and 63 characters long, only containing lowercase letters, numbers, hyphens, and not starting or ending with a hyphen.

   ```python
   import re

   def is_dns_compliant(bucket_name):
       pattern = re.compile(r'^[a-z0-9]([a-z0-9\-]{1,61}[a-z0-9])?$')
       return bool(pattern.match(bucket_name))
   ```

3. **Create S3 Bucket with Validation**:
   Use the Boto3 library to create an S3 bucket only if the bucket name is DNS-compliant. This script will check the bucket name and create it if it passes the validation.

   ```python
   import boto3
   from botocore.exceptions import ClientError

   def create_s3_bucket(bucket_name, region=None):
       if not is_dns_compliant(bucket_name):
           raise ValueError("Bucket name is not DNS-compliant")

       s3_client = boto3.client('s3', region_name=region)
       try:
           if region is None:
               s3_client.create_bucket(Bucket=bucket_name)
           else:
               s3_client.create_bucket(
                   Bucket=bucket_name,
                   CreateBucketConfiguration={'LocationConstraint': region}
               )
           print(f"Bucket {bucket_name} created successfully.")
       except ClientError as e:
           print(f"Error creating bucket: {e}")

   # Example usage
   create_s3_bucket('my-valid-bucket-name', 'us-west-1')
   ```

4. **Automate Bucket Creation with Compliance Check**:
   Integrate the compliance check into your workflow to ensure that any bucket creation request is validated before proceeding. This can be part of a larger automation script or CI/CD pipeline.

   ```python
   def automated_bucket_creation(bucket_name, region=None):
       try:
           create_s3_bucket(bucket_name, region)
       except ValueError as ve:
           print(f"Validation Error: {ve}")
       except ClientError as ce:
           print(f"AWS Error: {ce}")

   # Example usage in an automated workflow
   bucket_names = ['valid-bucket-1', 'Invalid_Bucket_2', 'another-valid-bucket']
   for name in bucket_names:
       automated_bucket_creation(name, 'us-west-1')
   ```

By following these steps, you can ensure that all S3 bucket names are DNS-compliant before they are created, thus preventing misconfigurations.
</Accordion>

</AccordionGroup>
</Tab>

<Tab title='Cause'>
### Check Cause
<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
1. Log in to the AWS Management Console and open the Amazon S3 console at https://console.aws.amazon.com/s3/.

2. In the S3 console, you will see a list of all your S3 buckets. The bucket names are displayed under the 'Name' column.

3. Check each bucket name individually. According to AWS naming conventions, bucket names must be DNS-compliant and should not contain underscores, end with a dash, have consecutive periods, or use dashes adjacent to periods. 

4. If any of your bucket names do not follow these rules, they are not DNS-compliant. You can also use the AWS CLI or SDKs to automate this process by listing all the bucket names and checking them against the naming rules.
</Accordion>

<Accordion title='Using CLI'>
1. Install and configure AWS CLI: Before you can start using AWS CLI, you need to install it on your local machine and configure it with your AWS account credentials. You can do this by running the following commands:

   Installation:
   ```
   pip install awscli
   ```
   Configuration:
   ```
   aws configure
   ```
   You will be prompted to provide your AWS Access Key ID, Secret Access Key, Default region name, and Default output format.

2. List all S3 buckets: Once your AWS CLI is set up, you can list all your S3 buckets by running the following command:
   ```
   aws s3api list-buckets --query "Buckets[].Name"
   ```
   This command will return a list of all your S3 bucket names.

3. Check for DNS-compliance: Now, you can check each bucket name for DNS-compliance. According to AWS, a bucket name must be a series of one or more labels, separated by a single period (.). Each label must start and end with a lowercase letter or a number. Bucket names can contain lowercase letters, numbers, and hyphens. 

   You can use a simple Python script to check for DNS-compliance. Here's an example:

   ```python
   import re

   def is_dns_compliant(bucket_name):
       pattern = r'^(?!-)[a-z0-9-]{1,63}(?<!-)$'
       labels = bucket_name.split('.')
       for label in labels:
           if not re.match(pattern, label):
               return False
       return True

   bucket_names = [...]  # replace with your bucket names
   for bucket_name in bucket_names:
       if not is_dns_compliant(bucket_name):
           print(f'Bucket name "{bucket_name}" is not DNS-compliant.')
   ```

4. Run the script: Finally, you can run the script to check all your S3 bucket names for DNS-compliance. Any non-compliant bucket names will be printed out.
</Accordion>

<Accordion title='Using Python'>
1. Install the necessary Python libraries: To interact with AWS services, you need to install the Boto3 library. You can install it using pip:

   ```bash
   pip install boto3
   ```

2. Configure AWS credentials: Before you can interact with AWS services, you need to set up your AWS credentials. You can do this by creating a file at ~/.aws/credentials. At the very least, the contents of the file should be:

   ```bash
   [default]
   aws_access_key_id = YOUR_ACCESS_KEY
   aws_secret_access_key = YOUR_SECRET_KEY
   ```

3. Create a Python script: Now, you can create a Python script that uses the Boto3 library to interact with AWS services. Here's a simple script that checks if S3 bucket names are DNS-compliant:

   ```python
   import boto3

   def is_dns_compliant(bucket_name):
       if len(bucket_name) < 3 or len(bucket_name) > 63:
           return False
       if '..' in bucket_name or '.-' in bucket_name or '-.' in bucket_name:
           return False
       if bucket_name[-1] == '.' or bucket_name[0] == '.':
           return False
       for char in bucket_name:
           if char.isalpha():
               continue
           if char.isdigit():
               continue
           if char != '.' and char != '-':
               return False
       return True

   s3 = boto3.resource('s3')
   for bucket in s3.buckets.all():
       if not is_dns_compliant(bucket.name):
           print(f'Bucket {bucket.name} is not DNS-compliant')
   ```

4. Run the Python script: Finally, you can run the Python script from the command line:

   ```bash
   python check_s3_buckets.py
   ```

This script will print out the names of all S3 buckets that are not DNS-compliant.
</Accordion>

</AccordionGroup>
</Tab>

<Tab title='Remediation'>
### Remediation

<AccordionGroup>
<Accordion title='Using Console' defaultOpen='true'>
Sure, here are the step-by-step instructions to remediate the misconfiguration in AWS:

1. Log in to your AWS Management Console.
2. Navigate to the S3 service.
3. Select the bucket that is non-compliant with DNS naming conventions.
4. Click on the "Properties" tab.
5. Scroll down to the "Static Website Hosting" section.
6. In the "Static website hosting" section, click on the "Edit" button.
7. In the "Edit static website hosting" dialog box, select the "Use this bucket to host a website" checkbox.
8. In the "Index document" field, enter a valid index document name (e.g. index.html).
9. In the "Error document" field, enter a valid error document name (e.g. error.html).
10. Click on the "Save changes" button.

Once you have completed the above steps, your S3 bucket will be compliant with DNS naming conventions. It is important to note that you should choose a bucket name that is unique, easy to remember and relevant to your organization.

#
</Accordion>

<Accordion title='Using CLI'>
To remediate the S3 bucket name non-compliance issue, you need to follow the below steps using AWS CLI:

1. Open your terminal or command prompt and install the AWS CLI (if not already installed) by following the instructions provided in the AWS documentation.

2. After installing AWS CLI, open your terminal or command prompt and enter the following command:

```
aws s3api put-bucket-acl --bucket BUCKET_NAME --acl public-read
```

Note: Replace BUCKET_NAME with the name of your S3 bucket.

3. Once the above command is executed successfully, enter the following command to rename the non-compliant bucket name:

```
aws s3api copy-object --copy-source BUCKET_NAME/ --bucket NEW_BUCKET_NAME --metadata-directive REPLACE --metadata x-amz-meta-original-name=BUCKET_NAME
```

Note: Replace BUCKET_NAME with the name of your non-compliant S3 bucket and NEW_BUCKET_NAME with the new DNS-compliant name you want to give to your S3 bucket.

4. After executing the above command successfully, enter the following command to delete the non-compliant bucket:

```
aws s3api delete-bucket --bucket BUCKET_NAME
```

Note: Replace BUCKET_NAME with the name of your non-compliant S3 bucket.

5. Finally, verify that the new bucket name is DNS-compliant by entering the following command:

```
aws s3api head-bucket --bucket NEW_BUCKET_NAME
```

If the above command returns no error, it means that your S3 bucket name is now DNS-compliant.
</Accordion>

<Accordion title='Using Python'>
To remediate this misconfiguration in AWS using Python, you can follow these steps:

1. Install the AWS SDK for Python (boto3) using pip:
```
pip install boto3
```

2. Create an AWS S3 client object:
```python
import boto3

s3 = boto3.client('s3')
```

3. List all the S3 buckets in your AWS account:
```python
response = s3.list_buckets()

for bucket in response['Buckets']:
    print(bucket['Name'])
```

4. Check if the bucket names are DNS-compliant:
```python
import re

for bucket in response['Buckets']:
    bucket_name = bucket['Name']
    if not re.match("^[a-z0-9][a-z0-9\-]{1,61}[a-z0-9]$", bucket_name):
        print(f"Bucket name {bucket_name} is not DNS-compliant")
```

5. If a bucket name is not DNS-compliant, rename the bucket:
```python
new_bucket_name = "new-dns-compliant-bucket-name"
s3.copy_object(Bucket=new_bucket_name, CopySource={"Bucket": bucket_name, "Key": ""})
s3.delete_object(Bucket=bucket_name, Key="")
s3.delete_bucket(Bucket=bucket_name)
```

Note: Make sure to replace "new-dns-compliant-bucket-name" with a DNS-compliant bucket name of your choice. Also, be aware that renaming a bucket can have implications for any applications or services that rely on the original bucket name.
</Accordion>

</AccordionGroup>
</Tab>
</Tabs>
### Additional Reading:

- [https://docs.aws.amazon.com/AmazonS3/latest/userguide/BucketRestrictions.html](https://docs.aws.amazon.com/AmazonS3/latest/userguide/BucketRestrictions.html) 

