---
slug: monitor_database_up_check
title: Cloud Monitoring Should Monitor Database Up Check
sidebar_label: Cloud Monitoring Should Monitor Database Up Check
---

### More Info:

Ensure Cloud Monitoring monitors database up check.

### Risk Level

Medium

### Address

Operational Maturity, Performance Efficiency

### Compliance Standards

CBP



<Tabs><Tab title='Cause'>
### Check Cause

#### Using Console

1. Log in to your Google Cloud Platform (GCP) console: Navigate to the GCP console (console.cloud.google.com) and sign in with your Google account that has the necessary permissions to access the GCP resources.

2. Navigate to Cloud Monitoring: From the GCP console dashboard, click on the navigation menu (three horizontal lines on the top left corner), scroll down and click on "Monitoring" under the "Operations" section.

3. Check for Database Up Check: In the Cloud Monitoring dashboard, click on "Uptime checks" under the "Alerting" section. Here, you should see a list of all the uptime checks configured for your GCP resources. Look for the Database Up Check in this list. If it's not present, it means it's not configured.

4. Verify the Configuration: If the Database Up Check is present, click on it to view its details. Ensure that it's configured correctly to monitor the database's uptime. The details should include the monitored resource type (e.g., Cloud SQL Database), the check type (e.g., TCP, HTTP, HTTPS), and the check frequency. If any of these details are incorrect or missing, it indicates a misconfiguration.

#### Using CLI

1. Install and authenticate the Google Cloud SDK: The Google Cloud SDK includes the gcloud command-line tool, which is necessary to interact with GCP services. You can install it by following the instructions provided by Google Cloud. After installation, authenticate the SDK using the command `gcloud auth login`.

2. List all the monitored resources: Use the following command to list all the monitored resources in your project. This will include databases and other resources.
   ```
   gcloud monitoring monitored-resource-descriptors list
   ```
   This command will return a list of all monitored resources, including their type and display name.

3. Check the monitoring status of the database: To check the monitoring status of a specific database, you need to know its resource type. For example, if you're monitoring a Cloud SQL database, the resource type would be `gcp_cloudsql_database`. Use the following command to get the details of the monitoring configuration for this resource type:
   ```
   gcloud monitoring monitored-resource-descriptors describe gcp_cloudsql_database
   ```
   This command will return the details of the monitoring configuration for the Cloud SQL database, including whether it's being monitored for uptime.

4. Check the uptime checks: To see the uptime checks associated with your project, use the following command:
   ```
   gcloud monitoring uptime-checks list
   ```
   This command will return a list of all uptime checks in your project. If there's an uptime check for your database, it should appear in this list. If it doesn't, then the database is not being monitored for uptime.

#### Using Python

1. **Import Required Libraries**: The first step is to import the required libraries. For this task, you will need the `google-cloud-monitoring` library. If you don't have it installed, you can do so using pip: `pip install google-cloud-monitoring`. Then, in your Python script, you would import it as follows:

```python
from google.cloud import monitoring_v3
```

2. **Initialize the Client**: The next step is to initialize the client. You will need to authenticate with your GCP account. You can do this by setting the `GOOGLE_APPLICATION_CREDENTIALS` environment variable to the path of your service account key file. Then, initialize the client as follows:

```python
client = monitoring_v3.MetricServiceClient()
```

3. **Specify the Project and Metric**: Now, you need to specify the project and the metric that you want to monitor. The project is identified by its project ID, and the metric is identified by its type. For a database up check, the metric type would be `uptime_check/check_passed`. Here is how you can specify these:

```python
project_id = "your-project-id"
metric_type = "uptime_check/check_passed"
project_name = f"projects/{project_id}"
```

4. **Retrieve and Check the Metric Data**: Finally, you can retrieve the metric data and check whether the database is up. You can do this by calling the `list_time_series` method on the client, passing in the project name and the metric type. Then, you can iterate over the returned time series and check the value of the metric:

```python
interval = monitoring_v3.types.TimeInterval()
now = time.time()
interval.end_time.seconds = int(now)
interval.start_time.seconds = int(now - 3600)
results = client.list_time_series(
    project_name,
    f'metric.type = "{metric_type}"',
    interval,
    monitoring_v3.enums.ListTimeSeriesRequest.TimeSeriesView.FULL
)
for result in results:
    print(result.metric.labels)
    print(result.resource.labels)
    for point in result.points:
        print(point.value)
```

This script will print the labels and values of the metric for the past hour. If the database is up, the value of the metric should be `True`.

</Tab>


<Tab title='Remediation'>
### Remediation

#### Using Console

To remediate the misconfiguration of not monitoring the database up check in GCP using GCP console, follow the below steps:

1. Login to your GCP console.
2. Navigate to the Monitoring page by clicking on the hamburger menu on the top left corner of the console and selecting "Monitoring" under the "Operations" section.
3. Click on the "Uptime Checks" tab on the left-hand side of the Monitoring page.
4. Click on the "Create Uptime Check" button on the top right corner of the Uptime Checks page.
5. In the "Create Uptime Check" page, fill in the required details such as the name of the check, the resource type, and the resource name.
6. In the "Check Type" section, select the "TCP" option.
7. In the "TCP" section, enter the IP address of your database instance and the port number on which the database is running.
8. In the "Advanced Settings" section, you can configure additional settings such as the check frequency, the notification settings, and the authentication settings.
9. Once you have configured all the required settings, click on the "Create" button to create the uptime check.

With these steps, you have successfully remediated the misconfiguration of not monitoring the database up check in GCP using GCP console. Now you will receive notifications whenever your database is down, and you can take necessary actions to bring it back up.

#### Using CLI

To remediate the misconfiguration of not monitoring database up check in GCP using GCP CLI, you can follow these steps:

1. Open the Google Cloud Console and navigate to the Cloud SQL instances page.
2. Select the instance that you want to monitor and click on the "Edit" button.
3. Scroll down to the "Monitoring" section and click on the "Configure monitoring" button.
4. In the "Metrics" tab, select the "Database" category and check the "Database uptime" metric.
5. In the "Alerting" tab, click on the "Add condition" button.
6. Set the "Condition name", "Condition trigger", and "Condition threshold" according to your requirements.
7. In the "Notification channels" section, add the email addresses or phone numbers of the people who should be notified in case of an alert.
8. Click on the "Save" button to save the changes.

Once you have completed these steps, you will be notified if the database goes down or if it is unavailable for any reason. You can also customize the alerting settings further by adding additional conditions, notification channels, or by adjusting the thresholds for the existing conditions.

#### Using Python

To remediate the misconfiguration "Cloud Monitoring should monitor database up check" for GCP using Python, follow these steps:

1. First, ensure that you have the necessary permissions to create a Cloud Monitoring uptime check. You will need the `monitoring.editor` or `monitoring.admin` role.

2. Import the necessary libraries and authenticate to the GCP project:

```
from google.cloud import monitoring_v3
from google.oauth2 import service_account

credentials = service_account.Credentials.from_service_account_file('<path_to_service_account_file>')
client = monitoring_v3.UptimeCheckServiceClient(credentials=credentials)
project_name = 'projects/<project_id>'
```

3. Next, create a new uptime check using the `create_uptime_check_config` method. In this example, we will create an HTTP check for a database running on port 3306:

```
config = monitoring_v3.types.UptimeCheckConfig(
    display_name='Database Up Check',
    http_check=monitoring_v3.types.UptimeCheckHttpCheck(
        port=3306,
        path='/',
        request_method='GET',
        use_ssl=False,
    ),
    timeout=5.0,
    period=60,
    content_matchers=[],
    selected_regions=['us-central1'],
    is_internal=False,
)
response = client.create_uptime_check_config(request={'parent': project_name, 'uptime_check_config': config})
print('Created uptime check: {}'.format(response.name))
```

4. The `create_uptime_check_config` method returns the name of the newly created uptime check. You can use this name to update or delete the check later.

That's it! You have now remediated the misconfiguration by creating a Cloud Monitoring uptime check to monitor the database's availability.


### Additional Reading:

- [https://cloud.google.com/monitoring/api/metrics_gcp](https://cloud.google.com/monitoring/api/metrics_gcp) 


</Tab>
</Tabs>