--- 
slug: Microsoft.Compute.cloudServices.roles.write
eventname: Microsoft.Compute.cloudServices.roles.write
title: Microsoft.Compute.cloudServices.roles.write
sidebar_label: Microsoft.Compute.cloudServices.roles.write
---
                       
### Event Information

- The Microsoft.Compute.cloudServices.roles.write event in Azure for AzureVirtualMachines refers to a write operation performed on the roles associated with a cloud service in Azure.
- This event indicates that a change has been made to the roles configuration of a cloud service, specifically related to Azure Virtual Machines.
- It could involve actions such as adding, modifying, or deleting roles within the cloud service, which can impact the deployment and management of virtual machines within that service.


### Examples

1. Unauthorized users gaining write access to the Microsoft.Compute.cloudServices.roles.write permission can potentially create new roles or modify existing roles within Azure Virtual Machines. This can lead to privilege escalation attacks, where attackers can grant themselves elevated permissions and gain unauthorized access to sensitive resources.

2. Misconfiguration or misuse of the Microsoft.Compute.cloudServices.roles.write permission can result in unintended changes to the roles assigned to Azure Virtual Machines. This can lead to disruption of services, as incorrect or inappropriate roles may be assigned, causing issues with access control and resource management.

3. If the Microsoft.Compute.cloudServices.roles.write permission is compromised, an attacker could potentially modify the roles assigned to Azure Virtual Machines to gain unauthorized access to sensitive data or perform malicious actions. This can result in data breaches, data loss, or unauthorized modifications to critical infrastructure components. It is crucial to closely monitor and restrict access to this permission to prevent security incidents.

### Remediation

#### Using Console

1. To remediate unauthorized users gaining write access to the Microsoft.Compute.cloudServices.roles.write permission in Azure Virtual Machines, follow these steps:

- Step 1: Sign in to the Azure portal (https://portal.azure.com) using your Azure account credentials.
- Step 2: Navigate to the Azure Virtual Machines service.
- Step 3: Select the specific virtual machine for which you want to manage access.
- Step 4: In the left-hand menu, click on "Access control (IAM)".
- Step 5: Review the existing role assignments and identify any unauthorized users or roles with write access to the Microsoft.Compute.cloudServices.roles.write permission.
- Step 6: Remove the unauthorized users or roles by selecting them and clicking on the "Remove" button.
- Step 7: Regularly review and audit the role assignments to ensure that only authorized users have write access to this permission.

2. To remediate misconfiguration or misuse of the Microsoft.Compute.cloudServices.roles.write permission in Azure Virtual Machines, follow these steps:

- Step 1: Sign in to the Azure portal using your Azure account credentials.
- Step 2: Navigate to the Azure Virtual Machines service.
- Step 3: Select the specific virtual machine for which you want to manage access.
- Step 4: In the left-hand menu, click on "Access control (IAM)".
- Step 5: Review the existing role assignments and identify any incorrect or inappropriate roles assigned to the virtual machine.
- Step 6: Modify the role assignments by selecting the incorrect or inappropriate roles and clicking on the "Edit" button.
- Step 7: Assign the correct roles or remove the inappropriate roles to ensure proper access control and resource management.

3. To remediate a compromised Microsoft.Compute.cloudServices.roles.write permission in Azure Virtual Machines, follow these steps:

- Step 1: Sign in to the Azure portal using your Azure account credentials.
- Step 2: Navigate to the Azure Virtual Machines service.
- Step 3: Select the specific virtual machine that may have been compromised.
- Step 4: In the left-hand menu, click on "Access control (IAM)".
- Step 5: Review the existing role assignments and identify any unauthorized modifications to the roles assigned to the virtual machine.
- Step 6: Remove the unauthorized role assignments by selecting them and clicking on the "Remove" button.
- Step 7: Monitor the virtual machine for any suspicious activities and consider implementing additional security measures, such as enabling Azure Security Center and implementing network security groups, to prevent further unauthorized access or malicious actions.

#### Using CLI

To remediate unauthorized users gaining write access to the Microsoft.Compute.cloudServices.roles.write permission in Azure Virtual Machines, you can follow these steps using Azure CLI commands:

1. Identify the affected Azure Virtual Machines:
   - Run the following command to list all the Azure Virtual Machines in your subscription:
     ```
     az vm list
     ```

2. Remove the write permission from unauthorized users:
   - Run the following command to remove the write permission for a specific user or group from a Virtual Machine:
     ```
     az role assignment delete --assignee <user or group ID> --scope <Virtual Machine ID> --role "Microsoft.Compute/cloudServices.roles/write"
     ```
     Replace `<user or group ID>` with the ID of the unauthorized user or group, and `<Virtual Machine ID>` with the ID of the affected Virtual Machine.

3. Implement proper access controls and monitoring:
   - Regularly review and audit role assignments for Azure Virtual Machines to ensure that only authorized users have the necessary permissions.
   - Enable Azure Monitor to track and alert on any changes to role assignments or suspicious activities related to Azure Virtual Machines.

Remember to replace `<user or group ID>` and `<Virtual Machine ID>` with the appropriate values specific to your Azure environment.

#### Using Python

To remediate unauthorized users gaining write access to the Microsoft.Compute.cloudServices.roles.write permission in Azure Virtual Machines, you can follow these steps:

1. Regularly review and audit role assignments: Use Azure PowerShell or Azure CLI to retrieve the current role assignments for Azure Virtual Machines. Identify any unauthorized or suspicious role assignments and remove them immediately.

```python
import os
from azure.identity import DefaultAzureCredential
from azure.mgmt.authorization import AuthorizationManagementClient

# Authenticate using default credentials
credential = DefaultAzureCredential()

# Create the authorization management client
authorization_client = AuthorizationManagementClient(credential, os.environ['AZURE_SUBSCRIPTION_ID'])

# Get the role assignments for Azure Virtual Machines
role_assignments = authorization_client.role_assignments.list()

# Iterate through the role assignments and remove any unauthorized or suspicious assignments
for assignment in role_assignments:
    if assignment.role_definition_id == 'Microsoft.Compute/cloudServices.roles/write' and assignment.principal_type == 'User':
        authorization_client.role_assignments.delete_by_id(assignment.id)
```

2. Implement just-in-time access control: Enable just-in-time (JIT) access control for Azure Virtual Machines. This feature allows you to restrict access to virtual machines by only granting temporary access when needed. By limiting the time window for access, you can reduce the risk of unauthorized users gaining write access to roles.

```python
import os
from azure.identity import DefaultAzureCredential
from azure.mgmt.network import NetworkManagementClient

# Authenticate using default credentials
credential = DefaultAzureCredential()

# Create the network management client
network_client = NetworkManagementClient(credential, os.environ['AZURE_SUBSCRIPTION_ID'])

# Enable just-in-time access control for Azure Virtual Machines
network_client.virtual_machine_scale_set_extensions.create_or_update(
    resource_group_name='your-resource-group',
    vm_scale_set_name='your-vm-scale-set',
    extension_name='JitExtension',
    extension_parameters={
        'publisher': 'Microsoft.Azure.Security',
        'type': 'AzureSecurityCenter',
        'type_handler_version': '1.0',
        'auto_upgrade_minor_version': True,
        'settings': {
            'protectedSettings': {
                'storageAccountName': 'your-storage-account',
                'storageAccountSasToken': 'your-storage-account-sas-token'
            },
            'publicSettings': {
                'Jit': {
                    'enabled': True
                }
            }
        }
    }
)
```

3. Implement strong access control policies: Ensure that only authorized users have the necessary permissions to modify role assignments for Azure Virtual Machines. Use Azure RBAC (Role-Based Access Control) to define granular access control policies and assign roles to users based on their responsibilities. Regularly review and update these policies to maintain a secure environment.

```python
import os
from azure.identity import DefaultAzureCredential
from azure.mgmt.authorization import AuthorizationManagementClient

# Authenticate using default credentials
credential = DefaultAzureCredential()

# Create the authorization management client
authorization_client = AuthorizationManagementClient(credential, os.environ['AZURE_SUBSCRIPTION_ID'])

# Create a custom role definition with restricted permissions for modifying role assignments
role_definition = authorization_client.role_definitions.create_or_update(
    scope='/subscriptions/your-subscription-id',
    role_definition_id='CustomRoleDefinitionId',
    role_definition={
        'role_name': 'Custom Role',
        'description': 'Custom role with restricted permissions for modifying role assignments',
        'assignable_scopes': ['/subscriptions/your-subscription-id'],
        'permissions': [
            {
                'actions': [
                    'Microsoft.Authorization/roleAssignments/write',
                    'Microsoft.Authorization/roleAssignments/delete'
                ],
                'notActions': [],
                'dataActions': [],
                'notDataActions': []
            }
        ]
    }
)

# Assign the custom role to authorized users
authorization_client.role_assignments.create(
    scope='/subscriptions/your-subscription-id',
    role_assignment_name='CustomRoleAssignment',
    parameters={
        'role_definition_id': role_definition.id,
        'principal_id': 'your-user-or-group-object-id'
    }
)
```

These steps will help remediate the unauthorized write access to the Microsoft.Compute.cloudServices.roles.write permission in Azure Virtual Machines and enhance the security of your environment.


 